// railway/dist/iac/index.js
import { parse, print, Kind } from 'graphql';
import path2 from 'path';
import { pathToFileURL } from 'url';
import { createHash } from 'crypto';
import fs from 'fs';

// src/iac/graph.ts
var RAILWAY_GRAPH_VERSION = 1;
function resourceAddress(type, name) {
  return `${type}.${name}`;
}
function indexGraph(graph) {
  const byAddress = /* @__PURE__ */ new Map();
  const byTypeAndName = /* @__PURE__ */ new Map();
  for (const resource of graph.resources) {
    byAddress.set(resource.address, resource);
    byTypeAndName.set(`${resource.type}:${resource.name}`, resource);
  }
  return { byAddress, byTypeAndName };
}
function validateGraph(graph) {
  const errors = [];
  if (graph.version !== RAILWAY_GRAPH_VERSION) errors.push(`Unsupported graph version: ${graph.version}`);
  const addresses = /* @__PURE__ */ new Set();
  for (const resource of graph.resources) {
    if (addresses.has(resource.address)) errors.push(`Duplicate resource address: ${resource.address}`);
    addresses.add(resource.address);
  }
  for (const edge of graph.edges) {
    if (!addresses.has(edge.from)) errors.push(`Edge references missing source: ${edge.from}`);
    if (!addresses.has(edge.to)) errors.push(`Edge references missing target: ${edge.to}`);
  }
  return errors;
}

// src/iac/compiler.ts
function projectDefinitionToGraph(definition) {
  const resources = [...(definition.resources ?? definition.services ?? []).flat()];
  const resourcesByAddress = new Set(resources.map((resource) => resource.address));
  for (const resource of [...resources]) {
    if (resource.type !== "service" && resource.type !== "database") continue;
    for (const attachment of Object.values(resource.volumeAttachments ?? {})) {
      if (resourcesByAddress.has(attachment.volume)) continue;
      const volumeName = attachment.volume.split(".").slice(1).join(".");
      resources.push({ address: attachment.volume, type: "volume", name: volumeName });
      resourcesByAddress.add(attachment.volume);
    }
  }
  const edges = [];
  for (const resource of resources) {
    if (resource.type !== "service" && resource.type !== "database") continue;
    for (const attachment of Object.values(resource.volumeAttachments ?? {})) {
      edges.push({ from: resource.address, to: attachment.volume, type: "mount", key: attachment.mountPath });
    }
    for (const [key, value] of Object.entries(resource.variables ?? {})) {
      if (value.type !== "reference") continue;
      edges.push({ from: resource.address, to: value.resource, type: "variable", key });
    }
  }
  return {
    version: RAILWAY_GRAPH_VERSION,
    project: { name: definition.name },
    environments: (definition.environments ?? []).map((name) => ({ name })),
    resources: resources.map(stripRuntimeHelpers),
    edges
  };
}
function stripRuntimeHelpers(value) {
  if (Array.isArray(value)) return value.map(stripRuntimeHelpers);
  if (value == null || typeof value !== "object") return value;
  return Object.fromEntries(
    Object.entries(value).filter(([, child]) => typeof child !== "function").map(([key, child]) => [key, stripRuntimeHelpers(child)])
  );
}
function graphToEnvironmentConfig(graph, options = {}) {
  const config = { services: {} };
  const resourceNamesById = Object.fromEntries(graph.resources.map((resource) => [resource.address, resource.name]));
  const existingServiceIds = new Set(options.existingServiceIds ?? []);
  for (const resource of graph.resources) {
    if (resource.type === "service" || resource.type === "database") {
      const serviceKey = options.serviceIdsByName?.[resource.name] ?? resource.name;
      config.services[serviceKey] = resource.type === "database" ? databaseToEnvironmentConfig(resource, {
        isNew: !existingServiceIds.has(serviceKey),
        ...options.volumeIdsByServiceName?.[resource.name] ? { volumeId: options.volumeIdsByServiceName[resource.name] } : {}
      }) : serviceToEnvironmentConfig(resource, resourceNamesById, {
        isNew: !existingServiceIds.has(serviceKey),
        volumeIdsByName: options.volumeIdsByName ?? {}
      });
      const volumeId = options.volumeIdsByServiceName?.[resource.name];
      if (resource.type === "database" && volumeId != null) {
        const region = databaseRegion(resource);
        config.volumes = config.volumes ?? {};
        config.volumes[volumeId] = { isCreated: true, ...region ? { region } : {} };
      }
      continue;
    }
    if (resource.type === "volume") {
      config.volumes = config.volumes ?? {};
      const existingVolumeId = options.volumeIdsByName?.[resource.name];
      config.volumes[existingVolumeId ?? resource.name] = { ...existingVolumeId ? {} : { isCreated: true }, ...resource.config };
      continue;
    }
    if (resource.type === "bucket") {
      const existingBucketId = options.bucketIdsByName?.[resource.name];
      const bucketKey = existingBucketId ?? resource.name;
      config.buckets = config.buckets ?? {};
      config.buckets[bucketKey] = { ...existingBucketId ? {} : { isCreated: true }, ...resource.config, ...resource.groupId ? { groupId: resource.groupId } : {} };
      continue;
    }
    if (resource.type === "group") {
      config.groups = config.groups ?? {};
      config.groups[resource.name] = pruneEmpty({
        isCreated: true,
        name: resource.name,
        color: resource.color,
        icon: resource.icon,
        isCollapsed: resource.isCollapsed
      });
    }
  }
  return pruneEmpty(config);
}
function environmentConfigToGraph(config, options = {}) {
  const resources = [];
  const groupNamesById = Object.fromEntries(
    Object.entries(config.groups ?? {}).map(([groupId, groupConfig]) => [groupId, groupConfig?.name ?? groupId])
  );
  for (const [groupId, groupConfig] of Object.entries(config.groups ?? {})) {
    if (groupConfig == null || groupConfig.isDeleted) continue;
    const name = groupConfig.name ?? groupId;
    const parentGroupId = groupConfig.groupId;
    resources.push(pruneEmpty({
      address: resourceAddress("group", name),
      type: "group",
      name,
      color: groupConfig.color,
      icon: groupConfig.icon,
      isCollapsed: groupConfig.isCollapsed,
      ...parentGroupId ? { groupId: groupNamesById[parentGroupId] ?? parentGroupId } : {}
    }));
  }
  for (const [serviceId, serviceConfig] of Object.entries(config.services ?? {})) {
    if (serviceConfig == null || serviceConfig.isDeleted) continue;
    const name = options.serviceNamesById?.[serviceId] ?? serviceId;
    const imageName = serviceConfig.source?.image;
    const looksLikeDatabase = imageName?.includes("postgres") || imageName?.includes("mysql") || imageName?.includes("redis") || imageName?.includes("mongo");
    if (looksLikeDatabase) {
      const engine = imageName?.includes("mysql") ? "mysql" : imageName?.includes("redis") ? "redis" : imageName?.includes("mongo") ? "mongo" : "postgres";
      resources.push(pruneEmpty({
        address: resourceAddress("database", name),
        type: "database",
        kind: "database",
        engine,
        name,
        image: imageName ?? "postgres:16",
        output: engine === "redis" ? "REDIS_URL" : engine === "mysql" ? "MYSQL_URL" : engine === "mongo" ? "MONGO_URL" : "DATABASE_URL",
        defaultMountPath: Object.keys(serviceConfig.volumeMounts ?? {}).length > 0 ? serviceConfig.deploy?.requiredMountPath : void 0,
        ...serviceConfig.deploy ? { deploy: serviceConfig.deploy } : {},
        ...serviceConfig.volumeMounts ? { volumeMounts: serviceConfig.volumeMounts } : {},
        ...serviceConfig.groupId ? { groupId: groupNamesById[serviceConfig.groupId] ?? serviceConfig.groupId } : {}
      }));
      continue;
    }
    resources.push({
      address: resourceAddress("service", name),
      type: "service",
      kind: serviceConfig.source?.repo ? "github" : serviceConfig.source?.image ? "docker-image" : serviceConfig.deploy?.cronSchedule ? "function" : "empty",
      name,
      ...serviceConfig.source ? { source: { type: serviceConfig.source.image ? "image" : "github", ...serviceConfig.source } } : {},
      ...serviceConfig.build ? { build: serviceConfig.build } : {},
      ...serviceConfig.deploy ? { deploy: serviceConfig.deploy } : {},
      ...serviceConfig.variables ? { variables: variablesFromEnvironmentConfig(serviceConfig.variables) } : {},
      ...serviceConfig.networking || options.customDomainsByServiceId?.[serviceId] ? { networking: pruneEmpty({ ...serviceConfig.networking, customDomains: options.customDomainsByServiceId?.[serviceId] ?? serviceConfig.networking?.customDomains }) } : {},
      ...volumeAttachmentsFromEnvironmentConfig(serviceConfig.volumeMounts, options.volumeNamesById),
      ...serviceConfig.configFile ? { configFile: serviceConfig.configFile } : {},
      ...serviceConfig.groupId ? { groupId: groupNamesById[serviceConfig.groupId] ?? serviceConfig.groupId } : {}
    });
  }
  for (const [volumeId, volumeConfig] of Object.entries(config.volumes ?? {})) {
    if (volumeConfig == null || volumeConfig.isDeleted) continue;
    const name = options.volumeNamesById?.[volumeId] ?? volumeId;
    resources.push({
      address: resourceAddress("volume", name),
      type: "volume",
      name,
      config: volumeConfig
    });
  }
  for (const [bucketId, bucketConfig] of Object.entries(config.buckets ?? {})) {
    if (bucketConfig == null || bucketConfig.isDeleted) continue;
    const name = options.bucketNamesById?.[bucketId] ?? bucketId;
    const groupId = bucketConfig.groupId;
    resources.push({
      address: resourceAddress("bucket", name),
      type: "bucket",
      name,
      config: bucketConfig,
      ...groupId ? { groupId: groupNamesById[groupId] ?? groupId } : {}
    });
  }
  return projectDefinitionToGraph({
    name: options.projectName ?? "imported-project",
    resources
  });
}
function composePatch({ currentConfig, desiredConfig }) {
  return pruneEmpty(addDeletionMarkers({ currentConfig, desiredConfig }));
}
function addDeletionMarkers({ currentConfig, desiredConfig }) {
  const next = structuredClone(desiredConfig);
  for (const [volumeId, currentVolume] of Object.entries(currentConfig.volumes ?? {})) {
    if (currentVolume == null || currentVolume.isDeleted) continue;
    if (desiredConfig.volumes?.[volumeId] != null) continue;
    next.volumes = next.volumes ?? {};
    next.volumes[volumeId] = { isDeleted: true };
  }
  for (const [serviceId, currentService] of Object.entries(currentConfig.services ?? {})) {
    if (currentService == null || currentService.isDeleted) continue;
    const desiredService = desiredConfig.services?.[serviceId];
    if (desiredService == null) {
      next.services = next.services ?? {};
      next.services[serviceId] = { isDeleted: true };
      continue;
    }
    for (const [variableName, currentVariable] of Object.entries(currentService.variables ?? {})) {
      if (currentVariable == null) continue;
      if (desiredService.variables?.[variableName] != null) continue;
      next.services = next.services ?? {};
      next.services[serviceId] = next.services[serviceId] ?? {};
      next.services[serviceId].variables = next.services[serviceId].variables ?? {};
      next.services[serviceId].variables[variableName] = null;
    }
  }
  return next;
}
function databaseRegion(database2) {
  const regions = Object.entries(database2.deploy?.multiRegionConfig ?? {}).filter(([, config]) => config != null);
  if (regions.length !== 1) return void 0;
  return regions[0]?.[0];
}
function databaseDeploy(database2, requiredMountPath) {
  return {
    requiredMountPath,
    ...database2.deploy?.multiRegionConfig ? { multiRegionConfig: database2.deploy.multiRegionConfig } : {}
  };
}
function databaseToEnvironmentConfig(database2, options) {
  if (database2.engine !== "postgres") {
    return pruneEmpty({
      ...options.isNew ? { isCreated: true } : {},
      source: { image: database2.image },
      ...database2.defaultMountPath ? { deploy: databaseDeploy(database2, database2.defaultMountPath) } : {},
      ...options.volumeId && database2.defaultMountPath ? { volumeMounts: { [options.volumeId]: { mountPath: database2.defaultMountPath } } } : {}
    });
  }
  return pruneEmpty({
    ...options.isNew ? { isCreated: true } : {},
    source: { image: database2.image },
    deploy: databaseDeploy(database2, "/var/lib/postgresql/data"),
    variables: {
      PGDATA: { value: "/var/lib/postgresql/data/pgdata" },
      PGHOST: { value: "${{RAILWAY_PRIVATE_DOMAIN}}" },
      PGPORT: { value: "5432" },
      PGUSER: { value: "${{POSTGRES_USER}}" },
      PGDATABASE: { value: "${{POSTGRES_DB}}" },
      PGPASSWORD: { value: "${{POSTGRES_PASSWORD}}" },
      POSTGRES_DB: { value: "railway" },
      DATABASE_URL: { value: "postgresql://${{PGUSER}}:${{POSTGRES_PASSWORD}}@${{RAILWAY_PRIVATE_DOMAIN}}:5432/${{PGDATABASE}}" },
      POSTGRES_USER: { value: "postgres" },
      SSL_CERT_DAYS: { value: "820" },
      POSTGRES_PASSWORD: { generator: 'secret(32, "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")' },
      DATABASE_PUBLIC_URL: { value: "postgresql://${{PGUSER}}:${{POSTGRES_PASSWORD}}@${{RAILWAY_TCP_PROXY_DOMAIN}}:${{RAILWAY_TCP_PROXY_PORT}}/${{PGDATABASE}}" },
      RAILWAY_DEPLOYMENT_DRAINING_SECONDS: { value: "60" }
    },
    networking: { tcpProxies: { "5432": {} } },
    ...options.volumeId ? { volumeMounts: { [options.volumeId]: { mountPath: "/var/lib/postgresql/data" } } } : {}
  });
}
function serviceToEnvironmentConfig(service2, resourceNamesById, options) {
  const config = { ...options.isNew ? { isCreated: true } : {} };
  if (service2.source) {
    const source = pruneEmpty({
      repo: service2.source.type === "github" ? service2.source.repo : void 0,
      branch: service2.source.type === "github" ? service2.source.branch : void 0,
      image: service2.source.type === "image" ? service2.source.image : void 0,
      rootDirectory: service2.source.rootDirectory,
      commitSha: service2.source.commitSha,
      upstreamUrl: service2.source.upstreamUrl,
      checkSuites: service2.source.checkSuites,
      autoUpdates: service2.source.autoUpdates
    });
    if (source) config.source = source;
  }
  if (service2.build) config.build = service2.build;
  if (service2.deploy) config.deploy = service2.deploy;
  if (service2.variables) config.variables = variablesToEnvironmentConfig(service2.variables, resourceNamesById);
  if (service2.networking) config.networking = service2.networking;
  const attachedVolumeMounts = Object.fromEntries(
    Object.entries(service2.volumeAttachments ?? {}).map(([, attachment]) => {
      const volumeName = resourceNamesById[attachment.volume] ?? attachment.volume.split(".").slice(1).join(".");
      const volumeId = options.volumeIdsByName?.[volumeName] ?? volumeName;
      return [volumeId, pruneEmpty({ mountPath: attachment.mountPath, ...attachment.backupSchedules ? { backupSchedules: attachment.backupSchedules } : {} })];
    })
  );
  if (service2.volumeMounts || Object.keys(attachedVolumeMounts).length > 0) config.volumeMounts = { ...service2.volumeMounts ?? {}, ...attachedVolumeMounts };
  if (service2.configFile) config.configFile = service2.configFile;
  if (service2.parentServiceId) config.parentServiceId = service2.parentServiceId;
  if (service2.groupId) config.groupId = service2.groupId;
  if (service2.clusterRole) config.clusterRole = service2.clusterRole;
  if (service2.replicaConfig) config.replicaConfig = service2.replicaConfig;
  if (service2.clusterDisplay) config.clusterDisplay = service2.clusterDisplay;
  return pruneEmpty(config);
}
function variablesToEnvironmentConfig(variables, resourceNamesById) {
  return Object.fromEntries(
    Object.entries(variables).filter((entry) => entry[1].type !== "preserve").map(([key, value]) => [
      key,
      value.type === "literal" ? literalVariable(value) : value.type === "raw" ? value.value : value.type === "sharedReference" ? sharedReferenceVariable(value) : referenceVariable(value, resourceNamesById)
    ])
  );
}
function referenceVariable(value, resourceNamesById) {
  return { value: `\${{${resourceNamesById[value.resource] ?? value.resource}.${value.output}}}` };
}
function sharedReferenceVariable(value) {
  return { value: `\${{shared.${value.name}}}` };
}
function literalVariable(value) {
  const { type: _type, ...variable } = value;
  return variable;
}
function volumeAttachmentsFromEnvironmentConfig(volumeMounts, volumeNamesById = {}) {
  const attachments = Object.fromEntries(
    Object.entries(volumeMounts ?? {}).flatMap(([volumeId, mount]) => {
      if (!mount?.mountPath) return [];
      const name = volumeNamesById[volumeId] ?? volumeId;
      return [[name, { volume: resourceAddress("volume", name), mountPath: mount.mountPath, ...mount.backupSchedules ? { backupSchedules: mount.backupSchedules } : {} }]];
    })
  );
  return Object.keys(attachments).length > 0 ? { volumeAttachments: attachments } : {};
}
function variablesFromEnvironmentConfig(variables) {
  return Object.fromEntries(Object.entries(variables).filter(([, value]) => value != null).map(([key, value]) => [key, value?.value == null || value.value === "" ? { type: "preserve" } : { type: "literal", value: value.value }]));
}
function pruneEmpty(value, path3 = []) {
  if (Array.isArray(value)) return value.map((child) => pruneEmpty(child, path3));
  if (value == null || typeof value !== "object") return value;
  return Object.fromEntries(
    Object.entries(value).map(([key, child]) => [key, pruneEmpty(child, [...path3, key])]).filter(([, child]) => {
      if (child == null) return false;
      if (typeof child === "object" && !Array.isArray(child)) {
        if (["customDomains", "serviceDomains", "tcpProxies"].includes(path3[path3.length - 1] ?? "")) return true;
        return Object.keys(child).length > 0;
      }
      return true;
    })
  );
}

// src/iac/change-set.ts
var RAILWAY_CHANGE_SET_VERSION = 1;
var SUPPORTED_CHANGE_SET_VERSIONS = [0, 1];
function diffGraphs({ current, desired, revealValues = false }) {
  const changes = [];
  const diagnostics = [];
  const currentByAddress = new Map(current.resources.map((resource) => [resource.address, resource]));
  const desiredByAddress = new Map(desired.resources.map((resource) => [resource.address, resource]));
  for (const resource of desired.resources) {
    const previous = currentByAddress.get(resource.address);
    if (previous && isManagedByRepoConfig(previous)) {
      diagnostics.push({
        severity: "error",
        path: `resources.${resource.address}.configFile`,
        message: `${previous.name} is already managed by ${String(previous.configFile)}. Remove or migrate the repo config before managing this service from .railway/railway.ts.`
      });
      continue;
    }
    if (!previous) {
      changes.push({
        kind: "resource.create",
        address: resource.address,
        resource,
        path: `resources.${resource.address}`,
        summary: `Create ${resource.type} ${resource.name}`,
        severity: "safe",
        deployEffect: resource.type === "service" || resource.type === "database" ? "deploy" : "none"
      });
      continue;
    }
    if (previous.name !== resource.name) {
      changes.push(update(resource.address, "name", previous.name, resource.name, `Rename ${resource.type} ${previous.name} to ${resource.name}`));
    }
    diffVariables({ previous, resource, changes, resourcesByAddress: desiredByAddress, revealValues });
    diffTopLevelField({ previous, resource, field: "source", changes });
    diffTopLevelField({ previous, resource, field: "build", changes });
    if (previous.type === "database" && resource.type === "database") {
      diffDatabaseDeploy({ previous, resource, changes });
    } else {
      diffTopLevelField({ previous, resource, field: "deploy", changes });
    }
    diffTopLevelField({ previous, resource, field: "groupId", changes });
    diffNetworking({ previous, resource, changes });
    if (previous.type === "bucket" && resource.type === "bucket" && bucketRegion(previous) !== bucketRegion(resource)) {
      diagnostics.push({
        severity: "error",
        path: `resources.${resource.address}.config.region`,
        message: `Bucket region cannot be changed after creation. Create a new bucket in ${bucketRegion(resource) ?? "the desired region"} and migrate data instead.`
      });
    } else if (previous.type === "volume" && resource.type === "volume") {
      diffVolumeConfig({ previous, resource, changes });
    } else {
      diffTopLevelField({ previous, resource, field: "config", changes });
    }
    diffVolumeAttachments({ previous, resource, changes });
  }
  for (const resource of current.resources) {
    if (desiredByAddress.has(resource.address)) continue;
    changes.push({
      kind: "resource.delete",
      address: resource.address,
      previous: resource,
      path: `resources.${resource.address}`,
      summary: `Delete ${resource.type} ${resource.name}`,
      severity: "destructive",
      deployEffect: resource.type === "service" || resource.type === "database" ? "deploy" : "none"
    });
  }
  return { version: RAILWAY_CHANGE_SET_VERSION, changes, diagnostics };
}
function changeSetToGraph({ current, changeSet }) {
  const next = structuredClone(current);
  const resources = new Map(next.resources.map((resource) => [resource.address, resource]));
  for (const change of changeSet.changes) {
    if (change.kind === "resource.create") {
      resources.set(change.address, structuredClone(change.resource));
      continue;
    }
    if (change.kind === "resource.delete") {
      resources.delete(change.address);
      continue;
    }
    const resource = resources.get(change.address);
    if (!resource) continue;
    if (change.kind === "domain.create") {
      continue;
    }
    if (change.kind === "resource.update") {
      resource[change.field] = structuredClone(change.after);
      continue;
    }
    if (change.kind === "variable.set") {
      const serviceLike = resource;
      serviceLike.variables = serviceLike.variables ?? {};
      serviceLike.variables[change.variable] = structuredClone(change.after);
      continue;
    }
    if (change.kind === "variable.delete" && "variables" in resource) {
      delete resource.variables?.[change.variable];
    }
  }
  next.resources = [...resources.values()];
  next.edges = next.edges.filter((edge) => resources.has(edge.from) && resources.has(edge.to));
  return next;
}
function changeSetToEnvironmentPatch({
  currentGraph,
  currentConfig,
  changeSet,
  compileOptions = {}
}) {
  const desiredGraph = changeSetToGraph({ current: currentGraph, changeSet });
  const desiredConfig = graphToEnvironmentConfig(desiredGraph, compileOptions);
  return composePatch({ currentConfig, desiredConfig });
}
function validateChangeSet(changeSet) {
  const diagnostics = [];
  if (changeSet.version !== RAILWAY_CHANGE_SET_VERSION) {
    diagnostics.push({ severity: "error", path: "version", message: `Unsupported change set version: ${changeSet.version}` });
  }
  for (const [index, change] of changeSet.changes.entries()) {
    if (!change.path) diagnostics.push({ severity: "error", path: `changes.${index}.path`, message: "Change path is required" });
    if (!change.summary) diagnostics.push({ severity: "error", path: `changes.${index}.summary`, message: "Change summary is required" });
    if (change.kind === "resource.delete" && change.severity !== "destructive") {
      diagnostics.push({ severity: "warning", path: `changes.${index}.severity`, message: "Resource deletion should be marked destructive" });
    }
  }
  return diagnostics;
}
function renderChangeSet(changeSet) {
  if (changeSet.changes.length === 0) return "No changes.";
  return changeSet.changes.map((change) => `${marker(change)} ${change.summary}`).join("\n");
}
function diffVariables({ previous, resource, changes, resourcesByAddress, revealValues }) {
  if (!("variables" in previous) && !("variables" in resource)) return;
  const before = "variables" in previous ? previous.variables ?? {} : {};
  const after = "variables" in resource ? resource.variables ?? {} : {};
  for (const [key, value] of Object.entries(after)) {
    if (isPreservedVariable(value)) continue;
    if (!isUnknownCurrentVariable(before[key]) && stableStringify(normalizeVariableForDiff(before[key], resourcesByAddress)) === stableStringify(normalizeVariableForDiff(value, resourcesByAddress))) continue;
    changes.push({
      kind: "variable.set",
      address: resource.address,
      variable: key,
      before: before[key],
      after: value,
      path: `resources.${resource.address}.variables.${key}`,
      summary: `${before[key] ? "Update" : "Set"} variable ${resource.name}.${key}`,
      details: [`${resource.name}.${key} (${formatVariableDiffValue(before[key], resourcesByAddress, revealValues)} \u2192 ${formatVariableDiffValue(value, resourcesByAddress, revealValues)})`],
      severity: "safe",
      deployEffect: "deploy"
    });
  }
  for (const [key, value] of Object.entries(before)) {
    if (key in after) continue;
    changes.push({
      kind: "variable.delete",
      address: resource.address,
      variable: key,
      previous: value,
      path: `resources.${resource.address}.variables.${key}`,
      summary: `Delete variable ${resource.name}.${key}`,
      severity: "destructive",
      deployEffect: "deploy"
    });
  }
}
function diffNetworking({ previous, resource, changes }) {
  const before = "networking" in previous ? previous.networking : void 0;
  const after = "networking" in resource ? resource.networking : void 0;
  const beforeDomains = before?.customDomains ?? {};
  const afterDomains = after?.customDomains ?? {};
  for (const [domain, config] of Object.entries(afterDomains)) {
    if (beforeDomains[domain]) continue;
    changes.push({
      kind: "domain.create",
      address: resource.address,
      domain,
      ...config?.port !== void 0 && config.port !== null ? { targetPort: config.port } : {},
      path: `resources.${resource.address}.domains.${domain}`,
      summary: `Create custom domain ${resource.name}.${domain}`,
      severity: "safe",
      deployEffect: "none"
    });
  }
  const normalizedBefore = normalizeForDiff("networking", { ...before, customDomains: void 0, serviceDomains: void 0 });
  const normalizedAfter = normalizeForDiff("networking", { ...after, customDomains: void 0, serviceDomains: void 0 });
  if (stableStringify(normalizedBefore) !== stableStringify(normalizedAfter)) {
    changes.push(update(resource.address, "networking", before, after, `Update ${resource.name} networking`));
  }
}
function diffVolumeAttachments({ previous, resource, changes }) {
  const before = previous.volumeAttachments;
  const after = resource.volumeAttachments;
  const normalizedBefore = normalizeForDiff("volumeAttachments", before);
  const normalizedAfter = normalizeForDiff("volumeAttachments", after);
  if (stableStringify(normalizedBefore) === stableStringify(normalizedAfter)) return;
  const destructive = Object.keys(before ?? {}).some((key) => !(key in (after ?? {})));
  changes.push(update(resource.address, "volumeAttachments", before, after, summaryForField(resource, "volumeAttachments", normalizedBefore, normalizedAfter), changedLeafPaths(normalizedBefore, normalizedAfter, "volumeAttachments"), destructive ? "destructive" : "safe"));
}
function diffTopLevelField({ previous, resource, field, changes }) {
  const before = previous[field];
  const after = resource[field];
  if (field === "source" && previous.type === "database" && isEquivalentDatabaseSource(previous, after)) return;
  const normalizedBefore = normalizeForDiff(field, before);
  const normalizedAfter = normalizeForDiff(field, after);
  if (stableStringify(normalizedBefore) === stableStringify(normalizedAfter)) return;
  changes.push(update(resource.address, field, before, after, summaryForField(resource, field, normalizedBefore, normalizedAfter), changedLeafPaths(normalizedBefore, normalizedAfter, field)));
}
function diffVolumeConfig({ previous, resource, changes }) {
  const before = normalizeForDiff("config", previous.config);
  const after = normalizeForDiff("config", resource.config);
  if (stableStringify(before) === stableStringify(after)) return;
  const previousSize = previous.config?.sizeMB;
  const desiredSize = resource.config?.sizeMB;
  const severity = previous.config?.region !== resource.config?.region || typeof previousSize === "number" && typeof desiredSize === "number" && desiredSize < previousSize ? "destructive" : "safe";
  const summary = typeof previousSize === "number" && typeof desiredSize === "number" && desiredSize > previousSize ? `Resize volume ${resource.name} from ${previousSize}MB to ${desiredSize}MB` : summaryForField(resource, "config", before, after);
  changes.push(update(resource.address, "config", previous.config, resource.config, summary, changedLeafPaths(before, after, "config"), severity));
}
function diffDatabaseDeploy({ previous, resource, changes }) {
  const previousRegion = databaseRegion2(previous);
  const desiredRegion = databaseRegion2(resource);
  if (desiredRegion !== void 0 && desiredRegion !== previousRegion) {
    changes.push(update(
      resource.address,
      "deploy",
      previous.deploy,
      resource.deploy,
      `Move database ${resource.name} to ${desiredRegion}`,
      changedLeafPaths(normalizeDatabaseDeploy(previous.deploy), normalizeDatabaseDeploy(resource.deploy), "deploy"),
      "destructive"
    ));
    return;
  }
  const before = normalizeDatabaseDeploy(previous.deploy);
  const after = normalizeDatabaseDeploy(resource.deploy);
  if (stableStringify(before) === stableStringify(after)) return;
  changes.push(update(resource.address, "deploy", previous.deploy, resource.deploy, summaryForField(resource, "deploy", before, after), changedLeafPaths(before, after, "deploy")));
}
function normalizeDatabaseDeploy(value) {
  const normalized = normalizeForDiff("deploy", value);
  if (normalized == null || typeof normalized !== "object") return normalized;
  const copy = { ...normalized };
  delete copy.requiredMountPath;
  return Object.keys(copy).length === 0 ? void 0 : copy;
}
function summaryForField(resource, field, before, after) {
  const details = changedLeafPaths(before, after, field);
  if (details.length > 0) {
    const shown = details.slice(0, 3).map(detailPathForSummary).join(", ");
    const remaining = details.length - 3;
    return `Update ${resource.name} ${shown}${remaining > 0 ? ` and ${remaining} more` : ""}`;
  }
  if (field === "networking") {
    const beforeDomains = Object.keys(before?.customDomains ?? {});
    const afterDomains = Object.keys(after?.customDomains ?? {});
    const created = afterDomains.filter((domain) => !beforeDomains.includes(domain));
    if (created.length === 1) return `Create custom domain ${resource.name}.${created[0]}`;
    if (created.length > 1) return `Create ${created.length} custom domains for ${resource.name}`;
  }
  return `Update ${resource.name} ${field}`;
}
function detailPathForSummary(detail) {
  return detail.replace(/ \(.+\)$/, "");
}
function changedLeafPaths(before, after, prefix) {
  const beforeFlat = flattenForDiff(before);
  const afterFlat = flattenForDiff(after);
  const changed = [.../* @__PURE__ */ new Set([...Object.keys(beforeFlat), ...Object.keys(afterFlat)])].filter((key) => stableStringify(beforeFlat[key]) !== stableStringify(afterFlat[key])).sort();
  return changed.filter((key) => key !== "" || changed.length === 1).filter((key) => key === "" || !changed.some((other) => other !== key && other.startsWith(`${key}.`))).map((key) => {
    const path3 = key ? `${prefix}.${key}` : prefix;
    const beforeValue = beforeFlat[key];
    const afterValue = afterFlat[key];
    return `${friendlyPath(path3)} (${formatDiffValue(beforeValue)} \u2192 ${formatDiffValue(afterValue)})`;
  });
}
function friendlyPath(path3) {
  const region = /^deploy\.multiRegionConfig\.([^.]+)\.numReplicas$/.exec(path3);
  if (region) return `regions.${region[1]}`;
  return path3;
}
function formatDiffValue(value) {
  if (value === void 0) return "unset";
  if (value === null) return "null";
  if (typeof value === "string") return JSON.stringify(value);
  return JSON.stringify(value);
}
function flattenForDiff(value, prefix = "") {
  if (value == null || typeof value !== "object" || Array.isArray(value)) return { [prefix]: value };
  const entries = Object.entries(value);
  if (entries.length === 0) return { [prefix]: value };
  return Object.fromEntries(entries.flatMap(([key, child]) => Object.entries(flattenForDiff(child, prefix ? `${prefix}.${key}` : key))));
}
function update(address, field, before, after, summary, details, severity = "safe") {
  return {
    kind: "resource.update",
    address,
    field,
    before,
    after,
    ...details && details.length > 0 ? { details } : {},
    path: `resources.${address}.${field}`,
    summary,
    severity,
    deployEffect: field === "config" || field === "groupId" ? "none" : "deploy"
  };
}
function marker(change) {
  if (change.kind === "resource.create" || change.kind === "domain.create") return "+";
  if (change.kind === "resource.delete") return "-";
  return "~";
}
var REDACTED_VARIABLE_VALUE = "\xABhidden\xBB";
function formatVariableDiffValue(value, resourcesByAddress, revealValues) {
  if (value === void 0) return "unset";
  if (value.type === "preserve") return "preserve()";
  if (value.type === "reference") {
    const name = resourcesByAddress.get(value.resource)?.name ?? value.resource.split(".").slice(1).join(".") ?? value.resource;
    return `${name}.${value.output}`;
  }
  if (value.type === "sharedReference") return `shared.${value.name}`;
  if (!revealValues) return REDACTED_VARIABLE_VALUE;
  return value.type === "literal" ? formatDiffValue(value.value) : formatDiffValue(normalizeVariableForDiff(value, resourcesByAddress));
}
function normalizeVariableForDiff(value, resourcesByAddress) {
  if (value?.type === "sharedReference") return { type: "literal", value: `\${{shared.${value.name}}}` };
  if (value?.type !== "reference") return value;
  const name = resourcesByAddress.get(value.resource)?.name ?? value.resource.split(".").slice(1).join(".") ?? value.resource;
  return { type: "literal", value: `\${{${name}.${value.output}}}` };
}
function isManagedByRepoConfig(resource) {
  if (resource.type !== "service" && resource.type !== "database") return false;
  const configFile = resource.configFile;
  return typeof configFile === "string" && /railway\.(json|toml)$/i.test(configFile);
}
function bucketRegion(resource) {
  if (resource.type !== "bucket") return void 0;
  const config = resource.config;
  return config?.region ?? void 0;
}
function databaseRegion2(resource) {
  if (resource.type !== "database") return void 0;
  const regions = Object.entries(resource.deploy?.multiRegionConfig ?? {}).filter(([, config]) => config != null);
  if (regions.length !== 1) return void 0;
  return regions[0]?.[0];
}
function isEquivalentDatabaseSource(previous, after) {
  if (previous.type !== "database" || after == null || typeof after !== "object") return false;
  const source = after;
  return source.type === "image" && normalizeImageTag(String(source.image)) === normalizeImageTag(previous.image);
}
function isPreservedVariable(value) {
  return value?.type === "preserve";
}
function isUnknownCurrentVariable(value) {
  return value?.type === "preserve" || value?.type === "literal" && value.value === "";
}
function normalizeForDiff(field, value) {
  if (value == null || typeof value !== "object") return value;
  const copy = structuredClone(value);
  if (field === "source") {
    if (copy.checkSuites === false) delete copy.checkSuites;
    if (copy.branch === "main") delete copy.branch;
    delete copy.commitSha;
    delete copy.upstreamUrl;
    if (copy.rootDirectory === "") delete copy.rootDirectory;
    if (typeof copy.image === "string") copy.image = normalizeImageTag(copy.image);
  }
  if (field === "build") {
    if (copy.builder === "RAILPACK") delete copy.builder;
    if (copy.buildEnvironment === "V3") delete copy.buildEnvironment;
    if (copy.buildCommand === "") delete copy.buildCommand;
    if (copy.dockerfilePath === "Dockerfile") delete copy.dockerfilePath;
  }
  if (field === "deploy") {
    if (copy.useLegacyStacker === false) delete copy.useLegacyStacker;
    if (copy.ipv6EgressEnabled === false) delete copy.ipv6EgressEnabled;
    if (copy.runtime === "V2") delete copy.runtime;
    copy.multiRegionConfig = normalizeMultiRegionConfig(copy.multiRegionConfig);
    if (isDefaultMultiRegionConfig(copy.multiRegionConfig)) delete copy.multiRegionConfig;
    if (copy.multiRegionConfig != null && typeof copy.multiRegionConfig === "object" && !Array.isArray(copy.multiRegionConfig) && Object.keys(copy.multiRegionConfig).length === 0) delete copy.multiRegionConfig;
  }
  return Object.keys(copy).length === 0 ? void 0 : copy;
}
function normalizeMultiRegionConfig(value) {
  if (value == null || typeof value !== "object" || Array.isArray(value)) return value;
  const entries = Object.entries(value).filter(([, config]) => {
    if (config == null || typeof config !== "object" || Array.isArray(config)) return true;
    const replicas = config.numReplicas;
    return replicas !== null && replicas !== void 0;
  });
  return Object.fromEntries(entries);
}
function isDefaultMultiRegionConfig(value) {
  if (value == null || typeof value !== "object" || Array.isArray(value)) return false;
  const entries = Object.entries(value);
  if (entries.length !== 1) return false;
  const config = entries[0]?.[1];
  if (config == null || typeof config !== "object" || Array.isArray(config)) return false;
  const regionConfig = config;
  return Object.entries(regionConfig).every(
    ([key, child]) => key === "numReplicas" && child === 1 || key === "stackerAssignment" && child == null
  );
}
function normalizeImageTag(image2) {
  const match = /^(?:railwayapp\/|ghcr\.io\/railwayapp-templates\/)?(redis|mysql|mongo|postgres)(?:-ssl)?:(\d+)(?:\.\d+)*$/.exec(image2);
  if (!match) return image2;
  return `${match[1]}:${match[2]}`;
}
function stableStringify(value) {
  return JSON.stringify(sortForJson(value));
}
function sortForJson(value) {
  if (Array.isArray(value)) return value.map(sortForJson);
  if (value == null || typeof value !== "object") return value;
  return Object.fromEntries(
    Object.entries(value).sort(([left], [right]) => left.localeCompare(right)).map(([key, child]) => [key, sortForJson(child)])
  );
}

// src/core/errors.ts
var RailwayError = class extends Error {
  constructor(message, options) {
    super(message, options);
    this.name = new.target.name;
  }
};
var RailwayAuthError = class extends RailwayError {
  variable;
  constructor(variable) {
    super(
      `Missing Railway credential. Set ${variable} or pass it explicitly.`
    );
    this.variable = variable;
  }
};
var STALE_ENVIRONMENT_BASE_CODE = "STALE_ENVIRONMENT_BASE";
var StaleEnvironmentError = class extends RailwayError {
};
var RailwayGraphQLError = class extends RailwayError {
  status;
  errors;
  responseBody;
  constructor(args) {
    super(args.message);
    this.status = args.status;
    this.errors = args.errors ?? [];
    this.responseBody = args.responseBody;
  }
};

// src/core/logger.ts
var NOOP = () => {
};
function createLogger(verbose) {
  if (!verbose) return NOOP;
  return (message) => console.error(`[railway] ${message}`);
}

// src/core/config.ts
var DEFAULT_RAILWAY_GRAPHQL_ENDPOINT = "https://backboard.railway.com/graphql/v2";
var RAILWAY_TOKEN_ENV = "RAILWAY_API_TOKEN";
var RAILWAY_ENDPOINT_ENV = "RAILWAY_GRAPHQL_ENDPOINT";
var RAILWAY_TCP_PROXY_WS_ENV = "RAILWAY_TCP_PROXY_WS_ENDPOINT";
var RAILWAY_VERBOSE_ENV = "RAILWAY_VERBOSE";
var TCP_PROXY_WS_PORT = "2226";
var TCP_PROXY_WS_PATH = "/ws/exec";
function normalizeRailwayClientConfig(config = {}) {
  const token = firstNonEmpty(config.token, readEnv(RAILWAY_TOKEN_ENV));
  if (!token) throw new RailwayAuthError(RAILWAY_TOKEN_ENV);
  const fetchImpl = config.fetch ?? globalThis.fetch;
  if (!fetchImpl) {
    throw new Error("No fetch implementation found. Pass `fetch` in the config.");
  }
  const endpoint = firstNonEmpty(
    config.endpoint,
    config.graphqlEndpoint,
    readEnv(RAILWAY_ENDPOINT_ENV)
  ) ?? DEFAULT_RAILWAY_GRAPHQL_ENDPOINT;
  const tcpProxyWsEndpoint = firstNonEmpty(config.tcpProxyWsEndpoint, readEnv(RAILWAY_TCP_PROXY_WS_ENV)) ?? deriveTcpProxyWsEndpoint(endpoint);
  const verbose = config.verbose ?? isTruthyEnv(readEnv(RAILWAY_VERBOSE_ENV));
  const normalized = {
    token,
    authType: config.authType ?? "bearer",
    endpoint,
    fetch: fetchImpl,
    webSocketImpl: config.webSocketImpl,
    tcpProxyWsEndpoint,
    log: createLogger(verbose)
  };
  normalized.log(
    `config resolved: endpoint=${endpoint} authType=${normalized.authType}`
  );
  return normalized;
}
function deriveTcpProxyWsEndpoint(endpoint) {
  const url = new URL(endpoint);
  const host = url.hostname.startsWith("backboard.") ? `ssh.${url.hostname.slice("backboard.".length)}` : `ssh.${url.hostname}`;
  return `wss://${host}:${TCP_PROXY_WS_PORT}${TCP_PROXY_WS_PATH}`;
}
function readEnv(name) {
  if (typeof process === "undefined") return void 0;
  return process.env?.[name];
}
function isTruthyEnv(value) {
  if (value === void 0) return false;
  const normalized = value.trim().toLowerCase();
  return normalized === "1" || normalized === "true" || normalized === "yes";
}
function firstNonEmpty(...values) {
  for (const value of values) {
    if (value !== void 0 && value.trim().length > 0) return value;
  }
  return void 0;
}

// src/core/version.ts
var SDK_VERSION = "3.4.1" ;
var USER_AGENT = `railway-ts-sdk/${SDK_VERSION}`;

// src/core/graphql-client.ts
async function requestGraphQL(config, document, variables) {
  const operation = operationName(document);
  const start = Date.now();
  config.log(`\u2192 ${operation} POST ${config.endpoint}`);
  const response = await config.fetch(config.endpoint, {
    method: "POST",
    headers: {
      Accept: "application/json",
      ...authHeader(config),
      "Content-Type": "application/json",
      "User-Agent": USER_AGENT
    },
    body: JSON.stringify({
      query: print(document),
      variables
    })
  });
  const body = await parseGraphQLResponse(response);
  const errors = body?.errors ?? [];
  const elapsed = Date.now() - start;
  if (!response.ok || errors.length > 0) {
    config.log(
      `\u2190 ${operation} failed http=${response.status} errors=${errors.length} in ${elapsed}ms`
    );
    throw new RailwayGraphQLError({
      message: errors[0]?.message ?? `Railway GraphQL request failed with HTTP ${response.status}.`,
      status: response.status,
      errors,
      responseBody: body
    });
  }
  if (!body?.data) {
    config.log(`\u2190 ${operation} returned no data`);
    throw new RailwayGraphQLError({
      message: "Railway GraphQL response did not include data.",
      status: response.status,
      responseBody: body
    });
  }
  config.log(`\u2190 ${operation} ok in ${elapsed}ms`);
  return body.data;
}
function operationName(document) {
  for (const definition of document.definitions) {
    if (definition.kind === Kind.OPERATION_DEFINITION) {
      return definition.name?.value ?? "anonymous";
    }
  }
  return "operation";
}
function authHeader(config) {
  if (config.authType === "project-token") return { "project-access-token": config.token };
  return { Authorization: `Bearer ${config.token}` };
}
async function parseGraphQLResponse(response) {
  try {
    return await response.json();
  } catch {
    return void 0;
  }
}

// src/iac/client.ts
var IacClient = class {
  #config;
  constructor(config) {
    this.#config = normalizeRailwayClientConfig(config);
  }
  async getCurrentEnvironment(environmentId, options = {}) {
    const data = await gql(this.#config, `query IacEnvironmentConfig($environmentId: String!, $decryptVariables: Boolean) {
      environment(id: $environmentId) { id name projectId config(decryptVariables: $decryptVariables) configEtag }
    }`, { environmentId, decryptVariables: options.decryptVariables ?? false });
    const projectName = data.environment.projectId ? await this.getProjectName(data.environment.projectId) : void 0;
    const services = data.environment.projectId ? await this.getProjectServices(data.environment.projectId) : [];
    const volumes = data.environment.projectId ? await this.getProjectVolumes(data.environment.projectId) : [];
    const buckets = data.environment.projectId ? await this.getProjectBuckets(data.environment.projectId) : [];
    const customDomainsByServiceId = data.environment.projectId ? await this.getEnvironmentCustomDomains(data.environment.projectId, environmentId, services) : {};
    return {
      projectId: data.environment.projectId,
      projectName,
      environmentId: data.environment.id,
      environmentName: data.environment.name,
      config: data.environment.config ?? {},
      serviceNamesById: Object.fromEntries(services.map((service2) => [service2.id, service2.name])),
      volumeNamesById: Object.fromEntries(volumes.flatMap((volume2) => volume2.name ? [[volume2.id, volume2.name]] : [])),
      bucketNamesById: Object.fromEntries(buckets.map((bucket2) => [bucket2.id, bucket2.name])),
      customDomainsByServiceId,
      ...data.environment.configEtag ? { configEtag: data.environment.configEtag } : {}
    };
  }
  async getStagedPatch(environmentId, options = {}) {
    const data = await gql(this.#config, `query IacStagedPatch($environmentId: String!, $decryptVariables: Boolean) {
      environmentStagedChanges(environmentId: $environmentId) { id status patch(decryptVariables: $decryptVariables) meta }
    }`, { environmentId, decryptVariables: options.decryptVariables ?? false });
    return data.environmentStagedChanges;
  }
  async getProjectName(projectId) {
    const data = await gql(this.#config, `query IacProjectName($projectId: String!) {
      project(id: $projectId) { name }
    }`, { projectId });
    return data.project.name ?? void 0;
  }
  async getProjectServices(projectId) {
    const data = await gql(this.#config, `query IacProjectServices($projectId: String!) {
      project(id: $projectId) { services(first: 1000) { edges { node { id name } } } }
    }`, { projectId });
    return data.project.services.edges.map((edge) => edge.node);
  }
  async getProjectVolumes(projectId) {
    const data = await gql(this.#config, `query IacProjectVolumes($projectId: String!) {
      project(id: $projectId) { volumes(first: 1000) { edges { node { id name } } } }
    }`, { projectId });
    return data.project.volumes?.edges.map((edge) => edge.node) ?? [];
  }
  async getProjectBuckets(projectId) {
    const data = await gql(this.#config, `query IacProjectBuckets($projectId: String!) {
      project(id: $projectId) { buckets(first: 1000) { edges { node { id name } } } }
    }`, { projectId });
    return data.project.buckets.edges.map((edge) => edge.node);
  }
  async getEnvironmentCustomDomains(projectId, environmentId, services) {
    const entries = await Promise.all(services.map(async (service2) => {
      const data = await gql(this.#config, `query IacServiceDomains($projectId: String!, $environmentId: String!, $serviceId: String!) {
        domains(projectId: $projectId, environmentId: $environmentId, serviceId: $serviceId) { customDomains { domain targetPort } }
      }`, { projectId, environmentId, serviceId: service2.id }).catch(() => ({ domains: { customDomains: [] } }));
      return [service2.id, Object.fromEntries(data.domains.customDomains.map((domain) => [domain.domain, domain.targetPort == null ? {} : { port: domain.targetPort }]))];
    }));
    return Object.fromEntries(entries.filter(([, domains]) => Object.keys(domains).length > 0));
  }
  async ensureGraphResources({ projectId, environmentId, graph, currentConfig }) {
    const existing = await this.getProjectServices(projectId);
    const serviceIdsByName = Object.fromEntries(existing.map((service2) => [service2.name, service2.id]));
    const volumeIdsByServiceName = extractVolumeIdsByServiceName({ ...currentConfig ? { currentConfig } : {}, serviceIdsByName });
    for (const resource of graph.resources) {
      if (resource.type !== "service" && resource.type !== "database") continue;
      if (!serviceIdsByName[resource.name]) {
        const service2 = await this.createServiceForResource({ projectId, environmentId, resource });
        serviceIdsByName[resource.name] = service2.id;
      }
      if (resource.type === "database" && resource.defaultMountPath && !volumeIdsByServiceName[resource.name]) {
        const volume2 = await this.createVolumeForService({ projectId, environmentId, serviceId: serviceIdsByName[resource.name] });
        volumeIdsByServiceName[resource.name] = volume2.id;
      }
    }
    const bucketIdsByName = await this.ensureGraphBuckets({ projectId, graph });
    return { serviceIdsByName, volumeIdsByServiceName, bucketIdsByName };
  }
  async stageEnvironmentChanges({ environmentId, patch, merge = true }) {
    const data = await gql(this.#config, `mutation IacStageEnvironmentChanges($environmentId: String!, $payload: EnvironmentConfig!, $merge: Boolean) {
      environmentStageChanges(environmentId: $environmentId, input: $payload, merge: $merge) { id }
    }`, { environmentId, payload: patch, merge });
    return data.environmentStageChanges;
  }
  async previewChangeSet({ environmentId, changeSet }) {
    const data = await gql(this.#config, `mutation IacPreviewChangeSet($environmentId: String!, $input: JSON!) {
      environmentPreviewChangeSet(environmentId: $environmentId, input: $input) { changeSet diagnostics effects }
    }`, { environmentId, input: changeSet });
    return data.environmentPreviewChangeSet;
  }
  async applyChangeSet({ environmentId, changeSet, commitMessage, baseEtag }) {
    const variables = { environmentId, input: changeSet };
    if (commitMessage !== void 0) variables.commitMessage = commitMessage;
    if (baseEtag !== void 0) variables.baseConfigEtag = baseEtag;
    try {
      const data = await gql(this.#config, `mutation IacApplyChangeSet($environmentId: String!, $input: JSON!, $commitMessage: String, $baseConfigEtag: String) {
        environmentApplyChangeSet(environmentId: $environmentId, input: $input, commitMessage: $commitMessage, baseConfigEtag: $baseConfigEtag) { id status deploymentId stagedPatchId diagnostics changes { kind path summary status outputs } }
      }`, variables);
      return data.environmentApplyChangeSet;
    } catch (error) {
      if (isStaleBaseError(error)) {
        throw new StaleEnvironmentError("The environment changed since this plan was computed. Re-run plan and review the changes before applying.", { cause: error });
      }
      throw error;
    }
  }
  async commitStagedPatch({ environmentId, message, skipDeploys }) {
    const variables = { environmentId };
    if (message !== void 0) variables.message = message;
    if (skipDeploys !== void 0) variables.skipDeploys = skipDeploys;
    const data = await gql(this.#config, `mutation IacCommitStagedPatch($environmentId: String!, $message: String, $skipDeploys: Boolean) {
      environmentPatchCommitStaged(environmentId: $environmentId, commitMessage: $message, skipDeploys: $skipDeploys)
    }`, variables);
    return data.environmentPatchCommitStaged;
  }
  async ensureGraphBuckets({ projectId, graph }) {
    const existing = await this.getProjectBuckets(projectId);
    const idsByName = Object.fromEntries(existing.map((bucket2) => [bucket2.name, bucket2.id]));
    for (const resource of graph.resources) {
      if (resource.type !== "bucket" || idsByName[resource.name]) continue;
      const bucket2 = await this.createBucketForResource({ projectId, name: resource.name });
      idsByName[resource.name] = bucket2.id;
    }
    return idsByName;
  }
  async createBucketForResource({ projectId, name }) {
    const input = { projectId, environmentId: null };
    if (name !== void 0) input.name = name;
    const data = await gql(this.#config, `mutation IacBucketCreate($input: BucketCreateInput!) {
      bucketCreate(input: $input) { id name }
    }`, { input });
    return data.bucketCreate;
  }
  async createVolumeForService({ projectId, serviceId }) {
    const input = { projectId, environmentId: null, serviceId, mountPath: "/var/lib/postgresql/data" };
    const data = await gql(this.#config, `mutation IacVolumeCreate($input: VolumeCreateInput!) {
      volumeCreate(input: $input) { id name }
    }`, { input });
    return data.volumeCreate;
  }
  async createServiceForResource({ projectId, resource }) {
    const source = resource.type === "database" ? { image: resource.image } : resource.source?.type === "image" && resource.source.image ? { image: resource.source.image } : resource.source?.type === "github" && resource.source.repo ? { repo: resource.source.repo } : void 0;
    const input = {
      name: resource.name,
      projectId,
      environmentId: null
    };
    if (source) input.source = source;
    if (resource.type === "service" && resource.source?.type === "github" && resource.source.branch) {
      input.branch = resource.source.branch;
    }
    const data = await gql(this.#config, `mutation IacServiceCreate($input: ServiceCreateInput!) {
      serviceCreate(input: $input) { id name }
    }`, { input });
    return data.serviceCreate;
  }
};
async function gql(config, source, variables) {
  return requestGraphQL(config, parse(source), variables);
}
function isStaleBaseError(error) {
  if (!(error instanceof RailwayGraphQLError)) return false;
  return error.errors.some((item) => item.extensions?.code === STALE_ENVIRONMENT_BASE_CODE);
}
function extractVolumeIdsByServiceName({ currentConfig, serviceIdsByName }) {
  const volumeIdsByServiceName = {};
  for (const [serviceName, serviceId] of Object.entries(serviceIdsByName)) {
    const volumeId = Object.keys(currentConfig?.services?.[serviceId]?.volumeMounts ?? {})[0];
    if (volumeId) volumeIdsByServiceName[serviceName] = volumeId;
  }
  return volumeIdsByServiceName;
}

// src/iac/diff.ts
function renderPatchDiff({
  currentConfig,
  patch,
  title = "Railway patch"
}) {
  const lines = [];
  const changes = flattenChanges(currentConfig, patch);
  lines.push(title);
  lines.push("".padEnd(title.length, "\u2500"));
  lines.push("");
  if (changes.length === 0) {
    lines.push("No changes.");
    return lines.join("\n");
  }
  for (const change of changes) lines.push(formatChange(change));
  lines.push("");
  lines.push(`${changes.length} raw change${changes.length === 1 ? "" : "s"}.`);
  return lines.join("\n").trimEnd();
}
function numChangesInEnvironmentConfig(config) {
  return Object.keys(flatten(config)).length;
}
function flattenChanges(currentConfig, patch) {
  const current = flatten(currentConfig);
  const next = flatten(patch);
  return Object.entries(next).map(([path3, value]) => {
    if (value === null || path3.endsWith(".isDeleted")) return { type: "removed", path: path3, previous: current[path3] };
    if (current[path3] == null) return { type: "added", path: path3, next: value };
    return { type: "changed", path: path3, previous: current[path3], next: value };
  });
}
function flatten(value, prefix = "") {
  if (value == null || typeof value !== "object" || Array.isArray(value)) return prefix ? { [prefix]: value } : {};
  const entries = Object.entries(value);
  if (entries.length === 0) return prefix ? { [prefix]: {} } : {};
  return Object.fromEntries(entries.flatMap(([key, child]) => Object.entries(flatten(child, prefix ? `${prefix}.${key}` : key))));
}
function formatChange(change) {
  const prefix = change.type === "added" ? green("+") : change.type === "removed" ? red("-") : amber("~");
  if (change.type === "added") return `  ${prefix} ${change.path}: ${format(change.next)}`;
  if (change.type === "removed") return `  ${prefix} ${change.path}: ${format(change.previous)}`;
  return `  ${prefix} ${change.path}: ${format(change.previous)} -> ${format(change.next)}`;
}
function format(value) {
  return typeof value === "string" ? value : JSON.stringify(value);
}
function green(text) {
  return color(text, "32");
}
function red(text) {
  return color(text, "31");
}
function amber(text) {
  return color(text, "33");
}
function color(text, code) {
  if (process.env.NO_COLOR) return text;
  return `\x1B[${code}m${text}\x1B[0m`;
}
function defineRailway(program) {
  return program;
}
var define = defineRailway;
function project(name, definition) {
  const resources = (definition.resources ?? definition.services ?? []).flat();
  return { name, ...definition, resources };
}
function createRailwayContext(input = {}) {
  const environment = input.environment ?? input.environmentName;
  return {
    ...input,
    ...environment ? { environment, environmentName: environment } : {},
    randomString: (label = "random", bytes = 12) => createHash("sha256").update(`railway-iac:${environment ?? "default"}:${label}`).digest("hex").slice(0, bytes * 2),
    isEnvironment: (name) => environment === name,
    shared: createSharedVariableRefs()
  };
}
function github(repo, options = {}) {
  return { type: "github", repo, branch: options.branch ?? "main", ...options };
}
function image(imageName, options = {}) {
  return { type: "image", image: imageName, ...options };
}
function template(templateName, options = {}) {
  return { type: "template", template: templateName, ...options };
}
function empty() {
  return { type: "empty" };
}
function service(name, config = {}) {
  const source = normalizeSource(config.source, config.root ?? config.rootDirectory);
  const node = pruneEmpty2({
    address: resourceAddress("service", name),
    type: "service",
    kind: source?.type === "github" ? "github" : source?.type === "image" ? "docker-image" : source?.type === "template" ? "template" : "empty",
    name,
    ...source ? { source } : {},
    build: normalizeBuild(config),
    deploy: normalizeDeploy(config),
    networking: normalizeNetworking(config),
    ...config.env || config.variables ? { variables: normalizeVariables({ ...config.variables ?? {}, ...config.env ?? {} }) } : {},
    ...normalizeVolumeMounts(config.volumeMounts),
    ...config.configFile ? { configFile: config.configFile } : {},
    ...config.parentServiceId ? { parentServiceId: config.parentServiceId } : {},
    ...config.groupId ? { groupId: config.groupId } : {},
    ...config.clusterRole ? { clusterRole: config.clusterRole } : {},
    ...config.replicaConfig ? { replicaConfig: config.replicaConfig } : {},
    ...config.clusterDisplay ? { clusterDisplay: config.clusterDisplay } : {}
  });
  return withVariableRefs(node);
}
function fn(name, config = {}) {
  return withVariableRefs({ ...service(name, config), kind: "function" });
}
function postgres(name, config = {}) {
  return database(name, "postgres", {
    image: "ghcr.io/railwayapp-templates/postgres-ssl:18",
    output: "DATABASE_URL",
    defaultMountPath: "/var/lib/postgresql/data",
    ...config
  });
}
function mysql(name, config = {}) {
  return database(name, "mysql", { image: "mysql:9", output: "MYSQL_URL", defaultMountPath: "/var/lib/mysql", ...config });
}
function redis(name, config = {}) {
  return database(name, "redis", { image: "railwayapp/redis:8.2", output: "REDIS_URL", defaultMountPath: "/bitnami", ...config });
}
function mongo(name, config = {}) {
  return database(name, "mongo", { image: "mongo:8", output: "MONGO_URL", defaultMountPath: "/data/db", ...config });
}
function database(name, engine, options) {
  const output = options.output ?? "DATABASE_URL";
  const node = {
    address: resourceAddress("database", name),
    type: "database",
    kind: "database",
    engine,
    name,
    image: options.image,
    output,
    defaultMountPath: options.defaultMountPath,
    source: image(options.image),
    ...options.region ? { deploy: { multiRegionConfig: { [options.region]: { numReplicas: 1 } } } } : {}
  };
  return withVariableRefs(node);
}
function volume(name, config = {}) {
  return { address: resourceAddress("volume", name), type: "volume", name, config };
}
function bucket(name, config = {}) {
  return { address: resourceAddress("bucket", name), type: "bucket", name, config };
}
function group(name, resourcesOrOptions = {}, maybeOptions = {}) {
  const resources = Array.isArray(resourcesOrOptions) ? resourcesOrOptions.flat() : void 0;
  const options = Array.isArray(resourcesOrOptions) ? maybeOptions : resourcesOrOptions;
  const node = { address: resourceAddress("group", name), type: "group", name, ...options };
  if (!resources) return node;
  return [node, ...resources.map((resource) => ({ ...resource, groupId: name }))];
}
function ref(resource, output) {
  return variableReference(resource, output);
}
function preserve() {
  return { type: "preserve" };
}
function variableReference(resource, output) {
  return { type: "reference", resource: resource.address, output };
}
function sharedVariableReference(name) {
  return { type: "sharedReference", name };
}
function createSharedVariableRefs() {
  return new Proxy({}, {
    get(_target, property) {
      if (typeof property !== "string") return void 0;
      return sharedVariableReference(property);
    }
  });
}
function createVariableRefAccessor(resource) {
  return new Proxy({}, {
    get(_target, property) {
      if (typeof property !== "string") return void 0;
      return variableReference(resource, property);
    }
  });
}
function withVariableRefs(node) {
  Object.defineProperty(node, "env", { enumerable: false, value: createVariableRefAccessor(node) });
  return node;
}
function normalizeSource(source, rootDirectory) {
  if (source == null) return rootDirectory ? { type: "empty", rootDirectory } : void 0;
  if ("type" in source && source.type) return pruneEmpty2({ ...source, rootDirectory: source.rootDirectory ?? rootDirectory });
  if ("repo" in source && source.repo) return pruneEmpty2({ type: "github", repo: source.repo, branch: source.branch ?? "main", rootDirectory });
  if ("image" in source && source.image) return pruneEmpty2({ type: "image", image: source.image, rootDirectory });
  return rootDirectory ? { type: "empty", rootDirectory } : void 0;
}
function normalizeBuild(config) {
  if (typeof config.build === "string") return { buildCommand: config.build };
  return pruneEmpty2({ ...config.build });
}
function normalizeDeploy(config) {
  const preDeployCommand = config.preDeploy ?? config.preDeployCommand ?? config.run?.preDeploy;
  return pruneEmpty2({
    ...config.deploy,
    startCommand: config.start ?? config.startCommand ?? config.run?.command ?? config.deploy?.startCommand,
    preDeployCommand: Array.isArray(preDeployCommand) ? preDeployCommand : preDeployCommand ? [preDeployCommand] : config.deploy?.preDeployCommand,
    healthcheckPath: config.healthcheck ?? config.healthcheckPath ?? config.run?.healthcheck ?? config.deploy?.healthcheckPath,
    healthcheckTimeout: config.healthcheckTimeout ?? config.run?.healthcheckTimeout ?? config.deploy?.healthcheckTimeout,
    multiRegionConfig: normalizeReplicas(config.replicas, config.regions) ?? config.deploy?.multiRegionConfig
  });
}
function normalizeReplicas(replicas, regions) {
  if (typeof replicas === "number") return normalizeRegions({ "us-west2": replicas });
  if (replicas) return normalizeRegions(replicas);
  if (regions) return normalizeRegions(regions);
  return void 0;
}
function normalizeRegions(regions) {
  return Object.fromEntries(
    Object.entries(regions).map(([region, value]) => [
      region,
      typeof value === "number" ? { numReplicas: value } : pruneEmpty2({ numReplicas: value.count ?? value.replicas, stackerAssignment: value.stacker }) ?? {}
    ])
  );
}
function normalizeNetworking(config) {
  const customDomains = config.domains ? Object.fromEntries(config.domains.map((domain) => typeof domain === "string" ? [domain, { port: 8080 }] : [domain.domain, { port: domain.port ?? 8080 }])) : void 0;
  const tcpProxies = config.tcp ? Object.fromEntries(config.tcp.map((port) => [String(port), {}])) : config.tcpProxies ? Object.fromEntries(config.tcpProxies.map((port) => [port, {}])) : void 0;
  return pruneEmpty2({ ...config.networking, customDomains, tcpProxies });
}
function normalizeVariables(variables) {
  return Object.fromEntries(
    Object.entries(variables).map(([key, value]) => {
      if (typeof value === "string") return [key, { type: "literal", value }];
      if ("type" in value) return [key, value];
      return [key, { type: "raw", value }];
    })
  );
}
function normalizeVolumeMounts(volumeMounts) {
  if (!volumeMounts) return {};
  const rawMounts = {};
  const attachments = {};
  for (const [key, value] of Object.entries(volumeMounts)) {
    if (value && typeof value === "object" && "type" in value && value.type === "volume") {
      attachments[value.name] = { volume: value.address, mountPath: key };
      continue;
    }
    rawMounts[key] = value;
  }
  return pruneEmpty2({ volumeMounts: rawMounts, volumeAttachments: attachments }) ?? {};
}
function pruneEmpty2(value) {
  if (value == null || typeof value !== "object") return value;
  const entries = Object.entries(value).filter(([, child]) => child != null);
  if (entries.length === 0) return void 0;
  return Object.fromEntries(entries);
}

// src/iac/evaluator.ts
async function evaluateRailwayFile(filePath, options = {}) {
  const absolutePath = path2.resolve(filePath);
  const mod = await importRailwayFile(absolutePath);
  const exported = mod.default ?? mod;
  const definition = await resolveDefinition(exported, options.context);
  const graph = projectDefinitionToGraph(definition);
  const desiredConfig = graphToEnvironmentConfig(graph, options);
  return { graph, desiredConfig };
}
async function importRailwayFile(absolutePath) {
  const url = `${pathToFileURL(absolutePath).toString()}?t=${Date.now()}`;
  if (/\.[cm]?tsx?$/.test(absolutePath)) {
    const { tsImport } = await import('tsx/esm/api');
    return tsImport(url, import.meta.url);
  }
  return import(url);
}
async function resolveDefinition(exported, context = {}) {
  if (typeof exported === "function") return exported(createRailwayContext(context), project);
  return exported;
}
async function evaluateRailwayProject(options = {}) {
  const file = options.file ? path2.resolve(options.file) : findRailwayFile(options.cwd);
  const { graph, desiredConfig } = await evaluateRailwayFile(file, options);
  return new EvaluatedRailwayProject({ file, graph, desiredConfig });
}
function findRailwayFile(cwd = process.cwd()) {
  let current = path2.resolve(cwd);
  while (true) {
    const candidate = path2.join(current, ".railway", "railway.ts");
    if (fs.existsSync(candidate)) return candidate;
    const parent = path2.dirname(current);
    if (parent === current) {
      throw new Error(
        `Could not find .railway/railway.ts from ${path2.resolve(cwd)}. Pass { file } explicitly.`
      );
    }
    current = parent;
  }
}
var EvaluatedRailwayProject = class {
  file;
  graph;
  desiredConfig;
  constructor(snapshot) {
    this.file = snapshot.file;
    this.graph = snapshot.graph;
    this.desiredConfig = snapshot.desiredConfig;
  }
  get name() {
    return this.graph.project.name;
  }
  get resources() {
    return this.graph.resources;
  }
  get services() {
    return this.graph.resources.filter(
      (resource) => resource.type === "service" || resource.type === "database"
    );
  }
  resource(name) {
    const resource = this.graph.resources.find((candidate) => candidate.name === name);
    if (!resource) throw new Error(`Railway resource not found: ${name}`);
    return resource;
  }
  service(name) {
    const resource = this.graph.resources.find(
      (candidate) => candidate.name === name && (candidate.type === "service" || candidate.type === "database")
    );
    if (!resource) throw new Error(`Railway service not found: ${name}`);
    return resource;
  }
  database(name) {
    const resource = this.graph.resources.find(
      (candidate) => candidate.name === name && candidate.type === "database"
    );
    if (!resource) throw new Error(`Railway database not found: ${name}`);
    return resource;
  }
  bucket(name) {
    const resource = this.graph.resources.find(
      (candidate) => candidate.name === name && candidate.type === "bucket"
    );
    if (!resource) throw new Error(`Railway bucket not found: ${name}`);
    return resource;
  }
  volume(name) {
    const resource = this.graph.resources.find(
      (candidate) => candidate.name === name && candidate.type === "volume"
    );
    if (!resource) throw new Error(`Railway volume not found: ${name}`);
    return resource;
  }
  config(options = {}) {
    return graphToEnvironmentConfig(this.graph, options);
  }
  toJSON() {
    return {
      file: this.file,
      graph: this.graph,
      desiredConfig: this.desiredConfig
    };
  }
};

// src/iac/typegen.ts
function renderRailwayGraphTypes(graph) {
  const services = names(
    graph.resources.filter(
      (resource) => resource.type === "service" || resource.type === "database"
    )
  );
  const databases = names(graph.resources.filter((resource) => resource.type === "database"));
  const buckets = names(graph.resources.filter((resource) => resource.type === "bucket"));
  const volumes = names(graph.resources.filter((resource) => resource.type === "volume"));
  const resources = names(graph.resources);
  return `// Generated by Railway IaC. Do not edit by hand.
// Provides autocomplete for evaluated graph resource lookups.

import "railway/iac";

declare module "railway/iac" {
  interface RailwayIacServiceMap ${mapBody(services)}
  interface RailwayIacDatabaseMap ${mapBody(databases)}
  interface RailwayIacBucketMap ${mapBody(buckets)}
  interface RailwayIacVolumeMap ${mapBody(volumes)}
  interface RailwayIacResourceMap ${mapBody(resources)}
}
`;
}
function names(resources) {
  return [...new Set(resources.map((resource) => resource.name))].sort();
}
function mapBody(names2) {
  if (names2.length === 0) return "{}";
  return `{
${names2.map((name) => `    ${JSON.stringify(name)}: true;`).join("\n")}
  }`;
}

// src/iac/runner.ts
async function runRailwayIac(request = {}) {
  const command = request.command ?? "evaluate";
  const cwd = request.cwd ?? process.cwd();
  const file = request.file ?? await findRailwayFile(cwd);
  try {
    const evaluated = await evaluateRailwayProject({ file, context: runnerContext(request, command) });
    const input = {
      request,
      file: evaluated.file,
      graph: evaluated.graph,
      diagnostics: graphDiagnostics(evaluated.graph)
    };
    return await runEvaluatedCommand(command, input);
  } catch (error) {
    return errorResponse(command, file, error);
  }
}
async function runEvaluatedCommand(command, input) {
  switch (command) {
    case "typegen":
      return typegenRailwayIac(input);
    case "evaluate":
      return evaluateRailwayIac(input);
    case "current":
      return currentRailwayIac(input);
    case "plan":
      return planRailwayIac(input);
    case "stage":
      return stageRailwayIac(input);
    case "apply":
      return applyRailwayIac(input);
  }
}
function typegenRailwayIac({ file, graph, diagnostics }) {
  return {
    ok: hasNoErrors(diagnostics),
    command: "typegen",
    file,
    graphTypes: renderRailwayGraphTypes(graph),
    diagnostics
  };
}
function evaluateRailwayIac({ request, file, graph, diagnostics }) {
  return {
    ok: hasNoErrors(diagnostics),
    command: "evaluate",
    file,
    graph,
    ...request.includeTypes ? { graphTypes: renderRailwayGraphTypes(graph) } : {},
    diagnostics
  };
}
async function currentRailwayIac({ file, graph: desiredGraph, request, diagnostics }) {
  const context = requireBackboardContext(request.backboard, "plan");
  const client = new IacClient(clientConfig(context));
  const current = await readCurrentEnvironment(client, context);
  const currentGraph = graphFromCurrentEnvironment(current, desiredGraph);
  const { config: _config, ...currentEnvironment } = current;
  return {
    ok: hasNoErrors(diagnostics),
    command: "current",
    file,
    mode: "real",
    currentGraph,
    currentConfig: current.config,
    currentEnvironment,
    ...request.includeTypes ? { graphTypes: renderRailwayGraphTypes(currentGraph) } : {},
    diagnostics
  };
}
async function planRailwayIac({ file, graph: desiredGraph, request, diagnostics }) {
  const context = requireBackboardContext(request.backboard, "plan");
  const client = new IacClient(clientConfig(context));
  const current = await readCurrentEnvironment(client, context);
  const currentGraph = graphFromCurrentEnvironment(current, desiredGraph);
  const changeSet = diffGraphs({ current: currentGraph, desired: desiredGraph, revealValues: request.revealValues ?? false });
  const allDiagnostics = [...diagnostics, ...changeSetDiagnostics(changeSet)];
  const { config: _config, ...currentEnvironment } = current;
  const hasErrors = !hasNoErrors(allDiagnostics);
  const preview = !hasErrors && changeSet.changes.length > 0 ? await client.previewChangeSet({ environmentId: context.environmentId, changeSet }) : void 0;
  const previewedChangeSet = preview?.changeSet ?? changeSet;
  return {
    ok: !hasErrors,
    command: "plan",
    file,
    mode: "real",
    currentGraph,
    desiredGraph,
    currentConfig: current.config,
    currentEnvironment,
    changeSet: previewedChangeSet,
    ...preview ? { preview } : {},
    diff: renderChangeSet(previewedChangeSet),
    ...request.includeTypes ? { graphTypes: renderRailwayGraphTypes(desiredGraph) } : {},
    diagnostics: allDiagnostics
  };
}
async function stageRailwayIac(input) {
  const planned = await planRailwayIac(input);
  if (!planned.ok || !planned.changeSet || planned.changeSet.changes.length === 0) {
    return { ...planned, command: "stage" };
  }
  requireBackboardContext(input.request.backboard, "stage");
  return { ...planned, command: "stage" };
}
async function applyRailwayIac(input) {
  const planned = await planRailwayIac(input);
  if (!planned.ok || !planned.changeSet || planned.changeSet.changes.length === 0) {
    return { ...planned, command: "apply" };
  }
  const context = requireBackboardContext(input.request.backboard, "apply");
  const client = new IacClient(clientConfig(context));
  const applyResult = await client.applyChangeSet({
    environmentId: context.environmentId,
    changeSet: planned.changeSet,
    commitMessage: "Apply Railway configuration",
    // Base snapshot the plan was diffed against; backboard rejects a stale apply.
    ...planned.currentEnvironment?.configEtag ? { baseEtag: planned.currentEnvironment.configEtag } : {}
  });
  return {
    ...planned,
    command: "apply",
    applyResult,
    ...applyResult.deploymentId ? { deploymentId: applyResult.deploymentId } : {},
    ...applyResult.stagedPatchId ? { stagedPatchId: applyResult.stagedPatchId } : {}
  };
}
async function readCurrentEnvironment(client, context) {
  return client.getCurrentEnvironment(
    context.environmentId,
    context.decryptVariables === void 0 ? {} : { decryptVariables: context.decryptVariables }
  );
}
function graphFromCurrentEnvironment(current, desiredGraph) {
  return environmentConfigToGraph(current.config, {
    projectName: current.projectName ?? desiredGraph.project.name,
    serviceNamesById: current.serviceNamesById,
    volumeNamesById: current.volumeNamesById,
    bucketNamesById: current.bucketNamesById,
    customDomainsByServiceId: current.customDomainsByServiceId
  });
}
function changeSetDiagnostics(changeSet) {
  return changeSet.diagnostics.map((diagnostic) => ({
    severity: diagnostic.severity,
    path: diagnostic.path,
    message: diagnostic.message
  }));
}
function runnerContext(request, command) {
  const environment = request.context?.environment ?? request.context?.environmentName;
  const projectId = request.context?.projectId ?? request.backboard?.projectId;
  const environmentId = request.context?.environmentId ?? request.backboard?.environmentId;
  return {
    ...request.context,
    command,
    ...projectId ? { projectId } : {},
    ...environmentId ? { environmentId } : {},
    ...environment ? { environment, environmentName: environment } : {}
  };
}
function requireBackboardContext(context, command) {
  if (!context?.token) throw new Error(`Backboard token is required for ${command}.`);
  if (!context.environmentId) throw new Error(`Backboard environmentId is required for ${command}.`);
  return context;
}
function clientConfig(context) {
  return { token: context.token, ...context.authType ? { authType: context.authType } : {}, ...context.endpoint ? { graphqlEndpoint: context.endpoint } : {} };
}
function graphDiagnostics(graph) {
  return validateGraph(graph).map((message) => ({ severity: "error", path: "graph", message }));
}
function hasNoErrors(diagnostics) {
  return diagnostics.every((diagnostic) => diagnostic.severity !== "error");
}
function errorResponse(command, file, error) {
  return {
    ok: false,
    command,
    file,
    ...command === "plan" || command === "stage" || command === "apply" || command === "current" ? { mode: "real" } : {},
    diagnostics: [{ severity: "error", path: "", message: error instanceof Error ? error.message : String(error) }]
  };
}

export { EvaluatedRailwayProject, IacClient, RAILWAY_CHANGE_SET_VERSION, RAILWAY_GRAPH_VERSION, SUPPORTED_CHANGE_SET_VERSIONS, bucket, changeSetToEnvironmentPatch, changeSetToGraph, composePatch, createRailwayContext, database, define, defineRailway, diffGraphs, empty, environmentConfigToGraph, evaluateRailwayFile, evaluateRailwayProject, findRailwayFile, fn, github, graphToEnvironmentConfig, group, image, indexGraph, mongo, mysql, numChangesInEnvironmentConfig, postgres, preserve, project, projectDefinitionToGraph, redis, ref, renderChangeSet, renderPatchDiff, renderRailwayGraphTypes, resourceAddress, runRailwayIac, service, template, validateChangeSet, validateGraph, volume };
