// railway/dist/index.cjs
'use strict';

var graphql = require('graphql');
var path2 = require('path');
var url = require('url');
var crypto = require('crypto');
var fs = require('fs');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var path2__default = /*#__PURE__*/_interopDefault(path2);
var fs__default = /*#__PURE__*/_interopDefault(fs);

var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/iac/index.ts
var iac_exports = {};
__export(iac_exports, {
  EvaluatedRailwayProject: () => EvaluatedRailwayProject,
  IacClient: () => IacClient,
  RAILWAY_CHANGE_SET_VERSION: () => RAILWAY_CHANGE_SET_VERSION,
  RAILWAY_GRAPH_VERSION: () => RAILWAY_GRAPH_VERSION,
  SUPPORTED_CHANGE_SET_VERSIONS: () => SUPPORTED_CHANGE_SET_VERSIONS,
  bucket: () => bucket,
  changeSetToEnvironmentPatch: () => changeSetToEnvironmentPatch,
  changeSetToGraph: () => changeSetToGraph,
  composePatch: () => composePatch,
  createRailwayContext: () => createRailwayContext,
  database: () => database,
  define: () => define,
  defineRailway: () => defineRailway,
  diffGraphs: () => diffGraphs,
  empty: () => empty,
  environmentConfigToGraph: () => environmentConfigToGraph,
  evaluateRailwayFile: () => evaluateRailwayFile,
  evaluateRailwayProject: () => evaluateRailwayProject,
  findRailwayFile: () => findRailwayFile,
  fn: () => fn,
  github: () => github,
  graphToEnvironmentConfig: () => graphToEnvironmentConfig,
  group: () => group,
  image: () => image,
  indexGraph: () => indexGraph,
  mongo: () => mongo,
  mysql: () => mysql,
  numChangesInEnvironmentConfig: () => numChangesInEnvironmentConfig,
  postgres: () => postgres,
  preserve: () => preserve,
  project: () => project,
  projectDefinitionToGraph: () => projectDefinitionToGraph,
  redis: () => redis,
  ref: () => ref,
  renderChangeSet: () => renderChangeSet,
  renderPatchDiff: () => renderPatchDiff,
  renderRailwayGraphTypes: () => renderRailwayGraphTypes,
  resourceAddress: () => resourceAddress,
  runRailwayIac: () => runRailwayIac,
  service: () => service,
  template: () => template,
  validateChangeSet: () => validateChangeSet,
  validateGraph: () => validateGraph,
  volume: () => volume
});

// src/iac/graph.ts
var RAILWAY_GRAPH_VERSION = 1;
function resourceAddress(type, name) {
  return `${type}.${name}`;
}
function indexGraph(graph) {
  const byAddress = /* @__PURE__ */ new Map();
  const byTypeAndName = /* @__PURE__ */ new Map();
  for (const resource of graph.resources) {
    byAddress.set(resource.address, resource);
    byTypeAndName.set(`${resource.type}:${resource.name}`, resource);
  }
  return { byAddress, byTypeAndName };
}
function validateGraph(graph) {
  const errors = [];
  if (graph.version !== RAILWAY_GRAPH_VERSION) errors.push(`Unsupported graph version: ${graph.version}`);
  const addresses = /* @__PURE__ */ new Set();
  for (const resource of graph.resources) {
    if (addresses.has(resource.address)) errors.push(`Duplicate resource address: ${resource.address}`);
    addresses.add(resource.address);
  }
  for (const edge of graph.edges) {
    if (!addresses.has(edge.from)) errors.push(`Edge references missing source: ${edge.from}`);
    if (!addresses.has(edge.to)) errors.push(`Edge references missing target: ${edge.to}`);
  }
  return errors;
}

// src/iac/compiler.ts
function projectDefinitionToGraph(definition) {
  const resources = [...(definition.resources ?? definition.services ?? []).flat()];
  const resourcesByAddress = new Set(resources.map((resource) => resource.address));
  for (const resource of [...resources]) {
    if (resource.type !== "service" && resource.type !== "database") continue;
    for (const attachment of Object.values(resource.volumeAttachments ?? {})) {
      if (resourcesByAddress.has(attachment.volume)) continue;
      const volumeName = attachment.volume.split(".").slice(1).join(".");
      resources.push({ address: attachment.volume, type: "volume", name: volumeName });
      resourcesByAddress.add(attachment.volume);
    }
  }
  const edges = [];
  for (const resource of resources) {
    if (resource.type !== "service" && resource.type !== "database") continue;
    for (const attachment of Object.values(resource.volumeAttachments ?? {})) {
      edges.push({ from: resource.address, to: attachment.volume, type: "mount", key: attachment.mountPath });
    }
    for (const [key, value] of Object.entries(resource.variables ?? {})) {
      if (value.type !== "reference") continue;
      edges.push({ from: resource.address, to: value.resource, type: "variable", key });
    }
  }
  return {
    version: RAILWAY_GRAPH_VERSION,
    project: { name: definition.name },
    environments: (definition.environments ?? []).map((name) => ({ name })),
    resources: resources.map(stripRuntimeHelpers),
    edges
  };
}
function stripRuntimeHelpers(value) {
  if (Array.isArray(value)) return value.map(stripRuntimeHelpers);
  if (value == null || typeof value !== "object") return value;
  return Object.fromEntries(
    Object.entries(value).filter(([, child]) => typeof child !== "function").map(([key, child]) => [key, stripRuntimeHelpers(child)])
  );
}
function graphToEnvironmentConfig(graph, options = {}) {
  const config = { services: {} };
  const resourceNamesById = Object.fromEntries(graph.resources.map((resource) => [resource.address, resource.name]));
  const existingServiceIds = new Set(options.existingServiceIds ?? []);
  for (const resource of graph.resources) {
    if (resource.type === "service" || resource.type === "database") {
      const serviceKey = options.serviceIdsByName?.[resource.name] ?? resource.name;
      config.services[serviceKey] = resource.type === "database" ? databaseToEnvironmentConfig(resource, {
        isNew: !existingServiceIds.has(serviceKey),
        ...options.volumeIdsByServiceName?.[resource.name] ? { volumeId: options.volumeIdsByServiceName[resource.name] } : {}
      }) : serviceToEnvironmentConfig(resource, resourceNamesById, {
        isNew: !existingServiceIds.has(serviceKey),
        volumeIdsByName: options.volumeIdsByName ?? {}
      });
      const volumeId = options.volumeIdsByServiceName?.[resource.name];
      if (resource.type === "database" && volumeId != null) {
        const region = databaseRegion(resource);
        config.volumes = config.volumes ?? {};
        config.volumes[volumeId] = { isCreated: true, ...region ? { region } : {} };
      }
      continue;
    }
    if (resource.type === "volume") {
      config.volumes = config.volumes ?? {};
      const existingVolumeId = options.volumeIdsByName?.[resource.name];
      config.volumes[existingVolumeId ?? resource.name] = { ...existingVolumeId ? {} : { isCreated: true }, ...resource.config };
      continue;
    }
    if (resource.type === "bucket") {
      const existingBucketId = options.bucketIdsByName?.[resource.name];
      const bucketKey = existingBucketId ?? resource.name;
      config.buckets = config.buckets ?? {};
      config.buckets[bucketKey] = { ...existingBucketId ? {} : { isCreated: true }, ...resource.config, ...resource.groupId ? { groupId: resource.groupId } : {} };
      continue;
    }
    if (resource.type === "group") {
      config.groups = config.groups ?? {};
      config.groups[resource.name] = pruneEmpty({
        isCreated: true,
        name: resource.name,
        color: resource.color,
        icon: resource.icon,
        isCollapsed: resource.isCollapsed
      });
    }
  }
  return pruneEmpty(config);
}
function environmentConfigToGraph(config, options = {}) {
  const resources = [];
  const groupNamesById = Object.fromEntries(
    Object.entries(config.groups ?? {}).map(([groupId, groupConfig]) => [groupId, groupConfig?.name ?? groupId])
  );
  for (const [groupId, groupConfig] of Object.entries(config.groups ?? {})) {
    if (groupConfig == null || groupConfig.isDeleted) continue;
    const name = groupConfig.name ?? groupId;
    const parentGroupId = groupConfig.groupId;
    resources.push(pruneEmpty({
      address: resourceAddress("group", name),
      type: "group",
      name,
      color: groupConfig.color,
      icon: groupConfig.icon,
      isCollapsed: groupConfig.isCollapsed,
      ...parentGroupId ? { groupId: groupNamesById[parentGroupId] ?? parentGroupId } : {}
    }));
  }
  for (const [serviceId, serviceConfig] of Object.entries(config.services ?? {})) {
    if (serviceConfig == null || serviceConfig.isDeleted) continue;
    const name = options.serviceNamesById?.[serviceId] ?? serviceId;
    const imageName = serviceConfig.source?.image;
    const looksLikeDatabase = imageName?.includes("postgres") || imageName?.includes("mysql") || imageName?.includes("redis") || imageName?.includes("mongo");
    if (looksLikeDatabase) {
      const engine = imageName?.includes("mysql") ? "mysql" : imageName?.includes("redis") ? "redis" : imageName?.includes("mongo") ? "mongo" : "postgres";
      resources.push(pruneEmpty({
        address: resourceAddress("database", name),
        type: "database",
        kind: "database",
        engine,
        name,
        image: imageName ?? "postgres:16",
        output: engine === "redis" ? "REDIS_URL" : engine === "mysql" ? "MYSQL_URL" : engine === "mongo" ? "MONGO_URL" : "DATABASE_URL",
        defaultMountPath: Object.keys(serviceConfig.volumeMounts ?? {}).length > 0 ? serviceConfig.deploy?.requiredMountPath : void 0,
        ...serviceConfig.deploy ? { deploy: serviceConfig.deploy } : {},
        ...serviceConfig.volumeMounts ? { volumeMounts: serviceConfig.volumeMounts } : {},
        ...serviceConfig.groupId ? { groupId: groupNamesById[serviceConfig.groupId] ?? serviceConfig.groupId } : {}
      }));
      continue;
    }
    resources.push({
      address: resourceAddress("service", name),
      type: "service",
      kind: serviceConfig.source?.repo ? "github" : serviceConfig.source?.image ? "docker-image" : serviceConfig.deploy?.cronSchedule ? "function" : "empty",
      name,
      ...serviceConfig.source ? { source: { type: serviceConfig.source.image ? "image" : "github", ...serviceConfig.source } } : {},
      ...serviceConfig.build ? { build: serviceConfig.build } : {},
      ...serviceConfig.deploy ? { deploy: serviceConfig.deploy } : {},
      ...serviceConfig.variables ? { variables: variablesFromEnvironmentConfig(serviceConfig.variables) } : {},
      ...serviceConfig.networking || options.customDomainsByServiceId?.[serviceId] ? { networking: pruneEmpty({ ...serviceConfig.networking, customDomains: options.customDomainsByServiceId?.[serviceId] ?? serviceConfig.networking?.customDomains }) } : {},
      ...volumeAttachmentsFromEnvironmentConfig(serviceConfig.volumeMounts, options.volumeNamesById),
      ...serviceConfig.configFile ? { configFile: serviceConfig.configFile } : {},
      ...serviceConfig.groupId ? { groupId: groupNamesById[serviceConfig.groupId] ?? serviceConfig.groupId } : {}
    });
  }
  for (const [volumeId, volumeConfig] of Object.entries(config.volumes ?? {})) {
    if (volumeConfig == null || volumeConfig.isDeleted) continue;
    const name = options.volumeNamesById?.[volumeId] ?? volumeId;
    resources.push({
      address: resourceAddress("volume", name),
      type: "volume",
      name,
      config: volumeConfig
    });
  }
  for (const [bucketId, bucketConfig] of Object.entries(config.buckets ?? {})) {
    if (bucketConfig == null || bucketConfig.isDeleted) continue;
    const name = options.bucketNamesById?.[bucketId] ?? bucketId;
    const groupId = bucketConfig.groupId;
    resources.push({
      address: resourceAddress("bucket", name),
      type: "bucket",
      name,
      config: bucketConfig,
      ...groupId ? { groupId: groupNamesById[groupId] ?? groupId } : {}
    });
  }
  return projectDefinitionToGraph({
    name: options.projectName ?? "imported-project",
    resources
  });
}
function composePatch({ currentConfig, desiredConfig }) {
  return pruneEmpty(addDeletionMarkers({ currentConfig, desiredConfig }));
}
function addDeletionMarkers({ currentConfig, desiredConfig }) {
  const next = structuredClone(desiredConfig);
  for (const [volumeId, currentVolume] of Object.entries(currentConfig.volumes ?? {})) {
    if (currentVolume == null || currentVolume.isDeleted) continue;
    if (desiredConfig.volumes?.[volumeId] != null) continue;
    next.volumes = next.volumes ?? {};
    next.volumes[volumeId] = { isDeleted: true };
  }
  for (const [serviceId, currentService] of Object.entries(currentConfig.services ?? {})) {
    if (currentService == null || currentService.isDeleted) continue;
    const desiredService = desiredConfig.services?.[serviceId];
    if (desiredService == null) {
      next.services = next.services ?? {};
      next.services[serviceId] = { isDeleted: true };
      continue;
    }
    for (const [variableName, currentVariable] of Object.entries(currentService.variables ?? {})) {
      if (currentVariable == null) continue;
      if (desiredService.variables?.[variableName] != null) continue;
      next.services = next.services ?? {};
      next.services[serviceId] = next.services[serviceId] ?? {};
      next.services[serviceId].variables = next.services[serviceId].variables ?? {};
      next.services[serviceId].variables[variableName] = null;
    }
  }
  return next;
}
function databaseRegion(database2) {
  const regions = Object.entries(database2.deploy?.multiRegionConfig ?? {}).filter(([, config]) => config != null);
  if (regions.length !== 1) return void 0;
  return regions[0]?.[0];
}
function databaseDeploy(database2, requiredMountPath) {
  return {
    requiredMountPath,
    ...database2.deploy?.multiRegionConfig ? { multiRegionConfig: database2.deploy.multiRegionConfig } : {}
  };
}
function databaseToEnvironmentConfig(database2, options) {
  if (database2.engine !== "postgres") {
    return pruneEmpty({
      ...options.isNew ? { isCreated: true } : {},
      source: { image: database2.image },
      ...database2.defaultMountPath ? { deploy: databaseDeploy(database2, database2.defaultMountPath) } : {},
      ...options.volumeId && database2.defaultMountPath ? { volumeMounts: { [options.volumeId]: { mountPath: database2.defaultMountPath } } } : {}
    });
  }
  return pruneEmpty({
    ...options.isNew ? { isCreated: true } : {},
    source: { image: database2.image },
    deploy: databaseDeploy(database2, "/var/lib/postgresql/data"),
    variables: {
      PGDATA: { value: "/var/lib/postgresql/data/pgdata" },
      PGHOST: { value: "${{RAILWAY_PRIVATE_DOMAIN}}" },
      PGPORT: { value: "5432" },
      PGUSER: { value: "${{POSTGRES_USER}}" },
      PGDATABASE: { value: "${{POSTGRES_DB}}" },
      PGPASSWORD: { value: "${{POSTGRES_PASSWORD}}" },
      POSTGRES_DB: { value: "railway" },
      DATABASE_URL: { value: "postgresql://${{PGUSER}}:${{POSTGRES_PASSWORD}}@${{RAILWAY_PRIVATE_DOMAIN}}:5432/${{PGDATABASE}}" },
      POSTGRES_USER: { value: "postgres" },
      SSL_CERT_DAYS: { value: "820" },
      POSTGRES_PASSWORD: { generator: 'secret(32, "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")' },
      DATABASE_PUBLIC_URL: { value: "postgresql://${{PGUSER}}:${{POSTGRES_PASSWORD}}@${{RAILWAY_TCP_PROXY_DOMAIN}}:${{RAILWAY_TCP_PROXY_PORT}}/${{PGDATABASE}}" },
      RAILWAY_DEPLOYMENT_DRAINING_SECONDS: { value: "60" }
    },
    networking: { tcpProxies: { "5432": {} } },
    ...options.volumeId ? { volumeMounts: { [options.volumeId]: { mountPath: "/var/lib/postgresql/data" } } } : {}
  });
}
function serviceToEnvironmentConfig(service2, resourceNamesById, options) {
  const config = { ...options.isNew ? { isCreated: true } : {} };
  if (service2.source) {
    const source = pruneEmpty({
      repo: service2.source.type === "github" ? service2.source.repo : void 0,
      branch: service2.source.type === "github" ? service2.source.branch : void 0,
      image: service2.source.type === "image" ? service2.source.image : void 0,
      rootDirectory: service2.source.rootDirectory,
      commitSha: service2.source.commitSha,
      upstreamUrl: service2.source.upstreamUrl,
      checkSuites: service2.source.checkSuites,
      autoUpdates: service2.source.autoUpdates
    });
    if (source) config.source = source;
  }
  if (service2.build) config.build = service2.build;
  if (service2.deploy) config.deploy = service2.deploy;
  if (service2.variables) config.variables = variablesToEnvironmentConfig(service2.variables, resourceNamesById);
  if (service2.networking) config.networking = service2.networking;
  const attachedVolumeMounts = Object.fromEntries(
    Object.entries(service2.volumeAttachments ?? {}).map(([, attachment]) => {
      const volumeName = resourceNamesById[attachment.volume] ?? attachment.volume.split(".").slice(1).join(".");
      const volumeId = options.volumeIdsByName?.[volumeName] ?? volumeName;
      return [volumeId, pruneEmpty({ mountPath: attachment.mountPath, ...attachment.backupSchedules ? { backupSchedules: attachment.backupSchedules } : {} })];
    })
  );
  if (service2.volumeMounts || Object.keys(attachedVolumeMounts).length > 0) config.volumeMounts = { ...service2.volumeMounts ?? {}, ...attachedVolumeMounts };
  if (service2.configFile) config.configFile = service2.configFile;
  if (service2.parentServiceId) config.parentServiceId = service2.parentServiceId;
  if (service2.groupId) config.groupId = service2.groupId;
  if (service2.clusterRole) config.clusterRole = service2.clusterRole;
  if (service2.replicaConfig) config.replicaConfig = service2.replicaConfig;
  if (service2.clusterDisplay) config.clusterDisplay = service2.clusterDisplay;
  return pruneEmpty(config);
}
function variablesToEnvironmentConfig(variables, resourceNamesById) {
  return Object.fromEntries(
    Object.entries(variables).filter((entry) => entry[1].type !== "preserve").map(([key, value]) => [
      key,
      value.type === "literal" ? literalVariable(value) : value.type === "raw" ? value.value : value.type === "sharedReference" ? sharedReferenceVariable(value) : referenceVariable(value, resourceNamesById)
    ])
  );
}
function referenceVariable(value, resourceNamesById) {
  return { value: `\${{${resourceNamesById[value.resource] ?? value.resource}.${value.output}}}` };
}
function sharedReferenceVariable(value) {
  return { value: `\${{shared.${value.name}}}` };
}
function literalVariable(value) {
  const { type: _type, ...variable } = value;
  return variable;
}
function volumeAttachmentsFromEnvironmentConfig(volumeMounts, volumeNamesById = {}) {
  const attachments = Object.fromEntries(
    Object.entries(volumeMounts ?? {}).flatMap(([volumeId, mount]) => {
      if (!mount?.mountPath) return [];
      const name = volumeNamesById[volumeId] ?? volumeId;
      return [[name, { volume: resourceAddress("volume", name), mountPath: mount.mountPath, ...mount.backupSchedules ? { backupSchedules: mount.backupSchedules } : {} }]];
    })
  );
  return Object.keys(attachments).length > 0 ? { volumeAttachments: attachments } : {};
}
function variablesFromEnvironmentConfig(variables) {
  return Object.fromEntries(Object.entries(variables).filter(([, value]) => value != null).map(([key, value]) => [key, value?.value == null || value.value === "" ? { type: "preserve" } : { type: "literal", value: value.value }]));
}
function pruneEmpty(value, path3 = []) {
  if (Array.isArray(value)) return value.map((child) => pruneEmpty(child, path3));
  if (value == null || typeof value !== "object") return value;
  return Object.fromEntries(
    Object.entries(value).map(([key, child]) => [key, pruneEmpty(child, [...path3, key])]).filter(([, child]) => {
      if (child == null) return false;
      if (typeof child === "object" && !Array.isArray(child)) {
        if (["customDomains", "serviceDomains", "tcpProxies"].includes(path3[path3.length - 1] ?? "")) return true;
        return Object.keys(child).length > 0;
      }
      return true;
    })
  );
}

// src/iac/change-set.ts
var RAILWAY_CHANGE_SET_VERSION = 1;
var SUPPORTED_CHANGE_SET_VERSIONS = [0, 1];
function diffGraphs({ current, desired, revealValues = false }) {
  const changes = [];
  const diagnostics = [];
  const currentByAddress = new Map(current.resources.map((resource) => [resource.address, resource]));
  const desiredByAddress = new Map(desired.resources.map((resource) => [resource.address, resource]));
  for (const resource of desired.resources) {
    const previous = currentByAddress.get(resource.address);
    if (previous && isManagedByRepoConfig(previous)) {
      diagnostics.push({
        severity: "error",
        path: `resources.${resource.address}.configFile`,
        message: `${previous.name} is already managed by ${String(previous.configFile)}. Remove or migrate the repo config before managing this service from .railway/railway.ts.`
      });
      continue;
    }
    if (!previous) {
      changes.push({
        kind: "resource.create",
        address: resource.address,
        resource,
        path: `resources.${resource.address}`,
        summary: `Create ${resource.type} ${resource.name}`,
        severity: "safe",
        deployEffect: resource.type === "service" || resource.type === "database" ? "deploy" : "none"
      });
      continue;
    }
    if (previous.name !== resource.name) {
      changes.push(update(resource.address, "name", previous.name, resource.name, `Rename ${resource.type} ${previous.name} to ${resource.name}`));
    }
    diffVariables({ previous, resource, changes, resourcesByAddress: desiredByAddress, revealValues });
    diffTopLevelField({ previous, resource, field: "source", changes });
    diffTopLevelField({ previous, resource, field: "build", changes });
    if (previous.type === "database" && resource.type === "database") {
      diffDatabaseDeploy({ previous, resource, changes });
    } else {
      diffTopLevelField({ previous, resource, field: "deploy", changes });
    }
    diffTopLevelField({ previous, resource, field: "groupId", changes });
    diffNetworking({ previous, resource, changes });
    if (previous.type === "bucket" && resource.type === "bucket" && bucketRegion(previous) !== bucketRegion(resource)) {
      diagnostics.push({
        severity: "error",
        path: `resources.${resource.address}.config.region`,
        message: `Bucket region cannot be changed after creation. Create a new bucket in ${bucketRegion(resource) ?? "the desired region"} and migrate data instead.`
      });
    } else if (previous.type === "volume" && resource.type === "volume") {
      diffVolumeConfig({ previous, resource, changes });
    } else {
      diffTopLevelField({ previous, resource, field: "config", changes });
    }
    diffVolumeAttachments({ previous, resource, changes });
  }
  for (const resource of current.resources) {
    if (desiredByAddress.has(resource.address)) continue;
    changes.push({
      kind: "resource.delete",
      address: resource.address,
      previous: resource,
      path: `resources.${resource.address}`,
      summary: `Delete ${resource.type} ${resource.name}`,
      severity: "destructive",
      deployEffect: resource.type === "service" || resource.type === "database" ? "deploy" : "none"
    });
  }
  return { version: RAILWAY_CHANGE_SET_VERSION, changes, diagnostics };
}
function changeSetToGraph({ current, changeSet }) {
  const next = structuredClone(current);
  const resources = new Map(next.resources.map((resource) => [resource.address, resource]));
  for (const change of changeSet.changes) {
    if (change.kind === "resource.create") {
      resources.set(change.address, structuredClone(change.resource));
      continue;
    }
    if (change.kind === "resource.delete") {
      resources.delete(change.address);
      continue;
    }
    const resource = resources.get(change.address);
    if (!resource) continue;
    if (change.kind === "domain.create") {
      continue;
    }
    if (change.kind === "resource.update") {
      resource[change.field] = structuredClone(change.after);
      continue;
    }
    if (change.kind === "variable.set") {
      const serviceLike = resource;
      serviceLike.variables = serviceLike.variables ?? {};
      serviceLike.variables[change.variable] = structuredClone(change.after);
      continue;
    }
    if (change.kind === "variable.delete" && "variables" in resource) {
      delete resource.variables?.[change.variable];
    }
  }
  next.resources = [...resources.values()];
  next.edges = next.edges.filter((edge) => resources.has(edge.from) && resources.has(edge.to));
  return next;
}
function changeSetToEnvironmentPatch({
  currentGraph,
  currentConfig,
  changeSet,
  compileOptions = {}
}) {
  const desiredGraph = changeSetToGraph({ current: currentGraph, changeSet });
  const desiredConfig = graphToEnvironmentConfig(desiredGraph, compileOptions);
  return composePatch({ currentConfig, desiredConfig });
}
function validateChangeSet(changeSet) {
  const diagnostics = [];
  if (changeSet.version !== RAILWAY_CHANGE_SET_VERSION) {
    diagnostics.push({ severity: "error", path: "version", message: `Unsupported change set version: ${changeSet.version}` });
  }
  for (const [index, change] of changeSet.changes.entries()) {
    if (!change.path) diagnostics.push({ severity: "error", path: `changes.${index}.path`, message: "Change path is required" });
    if (!change.summary) diagnostics.push({ severity: "error", path: `changes.${index}.summary`, message: "Change summary is required" });
    if (change.kind === "resource.delete" && change.severity !== "destructive") {
      diagnostics.push({ severity: "warning", path: `changes.${index}.severity`, message: "Resource deletion should be marked destructive" });
    }
  }
  return diagnostics;
}
function renderChangeSet(changeSet) {
  if (changeSet.changes.length === 0) return "No changes.";
  return changeSet.changes.map((change) => `${marker(change)} ${change.summary}`).join("\n");
}
function diffVariables({ previous, resource, changes, resourcesByAddress, revealValues }) {
  if (!("variables" in previous) && !("variables" in resource)) return;
  const before = "variables" in previous ? previous.variables ?? {} : {};
  const after = "variables" in resource ? resource.variables ?? {} : {};
  for (const [key, value] of Object.entries(after)) {
    if (isPreservedVariable(value)) continue;
    if (!isUnknownCurrentVariable(before[key]) && stableStringify(normalizeVariableForDiff(before[key], resourcesByAddress)) === stableStringify(normalizeVariableForDiff(value, resourcesByAddress))) continue;
    changes.push({
      kind: "variable.set",
      address: resource.address,
      variable: key,
      before: before[key],
      after: value,
      path: `resources.${resource.address}.variables.${key}`,
      summary: `${before[key] ? "Update" : "Set"} variable ${resource.name}.${key}`,
      details: [`${resource.name}.${key} (${formatVariableDiffValue(before[key], resourcesByAddress, revealValues)} \u2192 ${formatVariableDiffValue(value, resourcesByAddress, revealValues)})`],
      severity: "safe",
      deployEffect: "deploy"
    });
  }
  for (const [key, value] of Object.entries(before)) {
    if (key in after) continue;
    changes.push({
      kind: "variable.delete",
      address: resource.address,
      variable: key,
      previous: value,
      path: `resources.${resource.address}.variables.${key}`,
      summary: `Delete variable ${resource.name}.${key}`,
      severity: "destructive",
      deployEffect: "deploy"
    });
  }
}
function diffNetworking({ previous, resource, changes }) {
  const before = "networking" in previous ? previous.networking : void 0;
  const after = "networking" in resource ? resource.networking : void 0;
  const beforeDomains = before?.customDomains ?? {};
  const afterDomains = after?.customDomains ?? {};
  for (const [domain, config] of Object.entries(afterDomains)) {
    if (beforeDomains[domain]) continue;
    changes.push({
      kind: "domain.create",
      address: resource.address,
      domain,
      ...config?.port !== void 0 && config.port !== null ? { targetPort: config.port } : {},
      path: `resources.${resource.address}.domains.${domain}`,
      summary: `Create custom domain ${resource.name}.${domain}`,
      severity: "safe",
      deployEffect: "none"
    });
  }
  const normalizedBefore = normalizeForDiff("networking", { ...before, customDomains: void 0, serviceDomains: void 0 });
  const normalizedAfter = normalizeForDiff("networking", { ...after, customDomains: void 0, serviceDomains: void 0 });
  if (stableStringify(normalizedBefore) !== stableStringify(normalizedAfter)) {
    changes.push(update(resource.address, "networking", before, after, `Update ${resource.name} networking`));
  }
}
function diffVolumeAttachments({ previous, resource, changes }) {
  const before = previous.volumeAttachments;
  const after = resource.volumeAttachments;
  const normalizedBefore = normalizeForDiff("volumeAttachments", before);
  const normalizedAfter = normalizeForDiff("volumeAttachments", after);
  if (stableStringify(normalizedBefore) === stableStringify(normalizedAfter)) return;
  const destructive = Object.keys(before ?? {}).some((key) => !(key in (after ?? {})));
  changes.push(update(resource.address, "volumeAttachments", before, after, summaryForField(resource, "volumeAttachments", normalizedBefore, normalizedAfter), changedLeafPaths(normalizedBefore, normalizedAfter, "volumeAttachments"), destructive ? "destructive" : "safe"));
}
function diffTopLevelField({ previous, resource, field, changes }) {
  const before = previous[field];
  const after = resource[field];
  if (field === "source" && previous.type === "database" && isEquivalentDatabaseSource(previous, after)) return;
  const normalizedBefore = normalizeForDiff(field, before);
  const normalizedAfter = normalizeForDiff(field, after);
  if (stableStringify(normalizedBefore) === stableStringify(normalizedAfter)) return;
  changes.push(update(resource.address, field, before, after, summaryForField(resource, field, normalizedBefore, normalizedAfter), changedLeafPaths(normalizedBefore, normalizedAfter, field)));
}
function diffVolumeConfig({ previous, resource, changes }) {
  const before = normalizeForDiff("config", previous.config);
  const after = normalizeForDiff("config", resource.config);
  if (stableStringify(before) === stableStringify(after)) return;
  const previousSize = previous.config?.sizeMB;
  const desiredSize = resource.config?.sizeMB;
  const severity = previous.config?.region !== resource.config?.region || typeof previousSize === "number" && typeof desiredSize === "number" && desiredSize < previousSize ? "destructive" : "safe";
  const summary = typeof previousSize === "number" && typeof desiredSize === "number" && desiredSize > previousSize ? `Resize volume ${resource.name} from ${previousSize}MB to ${desiredSize}MB` : summaryForField(resource, "config", before, after);
  changes.push(update(resource.address, "config", previous.config, resource.config, summary, changedLeafPaths(before, after, "config"), severity));
}
function diffDatabaseDeploy({ previous, resource, changes }) {
  const previousRegion = databaseRegion2(previous);
  const desiredRegion = databaseRegion2(resource);
  if (desiredRegion !== void 0 && desiredRegion !== previousRegion) {
    changes.push(update(
      resource.address,
      "deploy",
      previous.deploy,
      resource.deploy,
      `Move database ${resource.name} to ${desiredRegion}`,
      changedLeafPaths(normalizeDatabaseDeploy(previous.deploy), normalizeDatabaseDeploy(resource.deploy), "deploy"),
      "destructive"
    ));
    return;
  }
  const before = normalizeDatabaseDeploy(previous.deploy);
  const after = normalizeDatabaseDeploy(resource.deploy);
  if (stableStringify(before) === stableStringify(after)) return;
  changes.push(update(resource.address, "deploy", previous.deploy, resource.deploy, summaryForField(resource, "deploy", before, after), changedLeafPaths(before, after, "deploy")));
}
function normalizeDatabaseDeploy(value) {
  const normalized = normalizeForDiff("deploy", value);
  if (normalized == null || typeof normalized !== "object") return normalized;
  const copy = { ...normalized };
  delete copy.requiredMountPath;
  return Object.keys(copy).length === 0 ? void 0 : copy;
}
function summaryForField(resource, field, before, after) {
  const details = changedLeafPaths(before, after, field);
  if (details.length > 0) {
    const shown = details.slice(0, 3).map(detailPathForSummary).join(", ");
    const remaining = details.length - 3;
    return `Update ${resource.name} ${shown}${remaining > 0 ? ` and ${remaining} more` : ""}`;
  }
  if (field === "networking") {
    const beforeDomains = Object.keys(before?.customDomains ?? {});
    const afterDomains = Object.keys(after?.customDomains ?? {});
    const created = afterDomains.filter((domain) => !beforeDomains.includes(domain));
    if (created.length === 1) return `Create custom domain ${resource.name}.${created[0]}`;
    if (created.length > 1) return `Create ${created.length} custom domains for ${resource.name}`;
  }
  return `Update ${resource.name} ${field}`;
}
function detailPathForSummary(detail) {
  return detail.replace(/ \(.+\)$/, "");
}
function changedLeafPaths(before, after, prefix) {
  const beforeFlat = flattenForDiff(before);
  const afterFlat = flattenForDiff(after);
  const changed = [.../* @__PURE__ */ new Set([...Object.keys(beforeFlat), ...Object.keys(afterFlat)])].filter((key) => stableStringify(beforeFlat[key]) !== stableStringify(afterFlat[key])).sort();
  return changed.filter((key) => key !== "" || changed.length === 1).filter((key) => key === "" || !changed.some((other) => other !== key && other.startsWith(`${key}.`))).map((key) => {
    const path3 = key ? `${prefix}.${key}` : prefix;
    const beforeValue = beforeFlat[key];
    const afterValue = afterFlat[key];
    return `${friendlyPath(path3)} (${formatDiffValue(beforeValue)} \u2192 ${formatDiffValue(afterValue)})`;
  });
}
function friendlyPath(path3) {
  const region = /^deploy\.multiRegionConfig\.([^.]+)\.numReplicas$/.exec(path3);
  if (region) return `regions.${region[1]}`;
  return path3;
}
function formatDiffValue(value) {
  if (value === void 0) return "unset";
  if (value === null) return "null";
  if (typeof value === "string") return JSON.stringify(value);
  return JSON.stringify(value);
}
function flattenForDiff(value, prefix = "") {
  if (value == null || typeof value !== "object" || Array.isArray(value)) return { [prefix]: value };
  const entries = Object.entries(value);
  if (entries.length === 0) return { [prefix]: value };
  return Object.fromEntries(entries.flatMap(([key, child]) => Object.entries(flattenForDiff(child, prefix ? `${prefix}.${key}` : key))));
}
function update(address, field, before, after, summary, details, severity = "safe") {
  return {
    kind: "resource.update",
    address,
    field,
    before,
    after,
    ...details && details.length > 0 ? { details } : {},
    path: `resources.${address}.${field}`,
    summary,
    severity,
    deployEffect: field === "config" || field === "groupId" ? "none" : "deploy"
  };
}
function marker(change) {
  if (change.kind === "resource.create" || change.kind === "domain.create") return "+";
  if (change.kind === "resource.delete") return "-";
  return "~";
}
var REDACTED_VARIABLE_VALUE = "\xABhidden\xBB";
function formatVariableDiffValue(value, resourcesByAddress, revealValues) {
  if (value === void 0) return "unset";
  if (value.type === "preserve") return "preserve()";
  if (value.type === "reference") {
    const name = resourcesByAddress.get(value.resource)?.name ?? value.resource.split(".").slice(1).join(".") ?? value.resource;
    return `${name}.${value.output}`;
  }
  if (value.type === "sharedReference") return `shared.${value.name}`;
  if (!revealValues) return REDACTED_VARIABLE_VALUE;
  return value.type === "literal" ? formatDiffValue(value.value) : formatDiffValue(normalizeVariableForDiff(value, resourcesByAddress));
}
function normalizeVariableForDiff(value, resourcesByAddress) {
  if (value?.type === "sharedReference") return { type: "literal", value: `\${{shared.${value.name}}}` };
  if (value?.type !== "reference") return value;
  const name = resourcesByAddress.get(value.resource)?.name ?? value.resource.split(".").slice(1).join(".") ?? value.resource;
  return { type: "literal", value: `\${{${name}.${value.output}}}` };
}
function isManagedByRepoConfig(resource) {
  if (resource.type !== "service" && resource.type !== "database") return false;
  const configFile = resource.configFile;
  return typeof configFile === "string" && /railway\.(json|toml)$/i.test(configFile);
}
function bucketRegion(resource) {
  if (resource.type !== "bucket") return void 0;
  const config = resource.config;
  return config?.region ?? void 0;
}
function databaseRegion2(resource) {
  if (resource.type !== "database") return void 0;
  const regions = Object.entries(resource.deploy?.multiRegionConfig ?? {}).filter(([, config]) => config != null);
  if (regions.length !== 1) return void 0;
  return regions[0]?.[0];
}
function isEquivalentDatabaseSource(previous, after) {
  if (previous.type !== "database" || after == null || typeof after !== "object") return false;
  const source = after;
  return source.type === "image" && normalizeImageTag(String(source.image)) === normalizeImageTag(previous.image);
}
function isPreservedVariable(value) {
  return value?.type === "preserve";
}
function isUnknownCurrentVariable(value) {
  return value?.type === "preserve" || value?.type === "literal" && value.value === "";
}
function normalizeForDiff(field, value) {
  if (value == null || typeof value !== "object") return value;
  const copy = structuredClone(value);
  if (field === "source") {
    if (copy.checkSuites === false) delete copy.checkSuites;
    if (copy.branch === "main") delete copy.branch;
    delete copy.commitSha;
    delete copy.upstreamUrl;
    if (copy.rootDirectory === "") delete copy.rootDirectory;
    if (typeof copy.image === "string") copy.image = normalizeImageTag(copy.image);
  }
  if (field === "build") {
    if (copy.builder === "RAILPACK") delete copy.builder;
    if (copy.buildEnvironment === "V3") delete copy.buildEnvironment;
    if (copy.buildCommand === "") delete copy.buildCommand;
    if (copy.dockerfilePath === "Dockerfile") delete copy.dockerfilePath;
  }
  if (field === "deploy") {
    if (copy.useLegacyStacker === false) delete copy.useLegacyStacker;
    if (copy.ipv6EgressEnabled === false) delete copy.ipv6EgressEnabled;
    if (copy.runtime === "V2") delete copy.runtime;
    copy.multiRegionConfig = normalizeMultiRegionConfig(copy.multiRegionConfig);
    if (isDefaultMultiRegionConfig(copy.multiRegionConfig)) delete copy.multiRegionConfig;
    if (copy.multiRegionConfig != null && typeof copy.multiRegionConfig === "object" && !Array.isArray(copy.multiRegionConfig) && Object.keys(copy.multiRegionConfig).length === 0) delete copy.multiRegionConfig;
  }
  return Object.keys(copy).length === 0 ? void 0 : copy;
}
function normalizeMultiRegionConfig(value) {
  if (value == null || typeof value !== "object" || Array.isArray(value)) return value;
  const entries = Object.entries(value).filter(([, config]) => {
    if (config == null || typeof config !== "object" || Array.isArray(config)) return true;
    const replicas = config.numReplicas;
    return replicas !== null && replicas !== void 0;
  });
  return Object.fromEntries(entries);
}
function isDefaultMultiRegionConfig(value) {
  if (value == null || typeof value !== "object" || Array.isArray(value)) return false;
  const entries = Object.entries(value);
  if (entries.length !== 1) return false;
  const config = entries[0]?.[1];
  if (config == null || typeof config !== "object" || Array.isArray(config)) return false;
  const regionConfig = config;
  return Object.entries(regionConfig).every(
    ([key, child]) => key === "numReplicas" && child === 1 || key === "stackerAssignment" && child == null
  );
}
function normalizeImageTag(image2) {
  const match = /^(?:railwayapp\/|ghcr\.io\/railwayapp-templates\/)?(redis|mysql|mongo|postgres)(?:-ssl)?:(\d+)(?:\.\d+)*$/.exec(image2);
  if (!match) return image2;
  return `${match[1]}:${match[2]}`;
}
function stableStringify(value) {
  return JSON.stringify(sortForJson(value));
}
function sortForJson(value) {
  if (Array.isArray(value)) return value.map(sortForJson);
  if (value == null || typeof value !== "object") return value;
  return Object.fromEntries(
    Object.entries(value).sort(([left], [right]) => left.localeCompare(right)).map(([key, child]) => [key, sortForJson(child)])
  );
}

// src/core/errors.ts
var RailwayError = class extends Error {
  constructor(message, options) {
    super(message, options);
    this.name = new.target.name;
  }
};
var RailwayAuthError = class extends RailwayError {
  variable;
  constructor(variable) {
    super(
      `Missing Railway credential. Set ${variable} or pass it explicitly.`
    );
    this.variable = variable;
  }
};
var RailwayConnectionError = class extends RailwayError {
  closeCode;
  constructor(args) {
    super(
      args.message,
      args.cause === void 0 ? void 0 : { cause: args.cause }
    );
    this.closeCode = args.closeCode;
  }
};
var STALE_ENVIRONMENT_BASE_CODE = "STALE_ENVIRONMENT_BASE";
var StaleEnvironmentError = class extends RailwayError {
};
var RailwayGraphQLError = class extends RailwayError {
  status;
  errors;
  responseBody;
  constructor(args) {
    super(args.message);
    this.status = args.status;
    this.errors = args.errors ?? [];
    this.responseBody = args.responseBody;
  }
};

// src/core/logger.ts
var NOOP = () => {
};
function createLogger(verbose) {
  if (!verbose) return NOOP;
  return (message) => console.error(`[railway] ${message}`);
}

// src/core/config.ts
var DEFAULT_RAILWAY_GRAPHQL_ENDPOINT = "https://backboard.railway.com/graphql/v2";
var RAILWAY_TOKEN_ENV = "RAILWAY_API_TOKEN";
var RAILWAY_ENVIRONMENT_ENV = "RAILWAY_ENVIRONMENT_ID";
var RAILWAY_ENDPOINT_ENV = "RAILWAY_GRAPHQL_ENDPOINT";
var RAILWAY_TCP_PROXY_WS_ENV = "RAILWAY_TCP_PROXY_WS_ENDPOINT";
var RAILWAY_VERBOSE_ENV = "RAILWAY_VERBOSE";
var TCP_PROXY_WS_PORT = "2226";
var TCP_PROXY_WS_PATH = "/ws/exec";
var TCP_PROXY_FILES_WS_PATH = "/ws/files";
function normalizeRailwayClientConfig(config = {}) {
  const token = firstNonEmpty(config.token, readEnv(RAILWAY_TOKEN_ENV));
  if (!token) throw new RailwayAuthError(RAILWAY_TOKEN_ENV);
  const fetchImpl = config.fetch ?? globalThis.fetch;
  if (!fetchImpl) {
    throw new Error("No fetch implementation found. Pass `fetch` in the config.");
  }
  const endpoint = firstNonEmpty(
    config.endpoint,
    config.graphqlEndpoint,
    readEnv(RAILWAY_ENDPOINT_ENV)
  ) ?? DEFAULT_RAILWAY_GRAPHQL_ENDPOINT;
  const tcpProxyWsEndpoint = firstNonEmpty(config.tcpProxyWsEndpoint, readEnv(RAILWAY_TCP_PROXY_WS_ENV)) ?? deriveTcpProxyWsEndpoint(endpoint);
  const verbose = config.verbose ?? isTruthyEnv(readEnv(RAILWAY_VERBOSE_ENV));
  const normalized = {
    token,
    authType: config.authType ?? "bearer",
    endpoint,
    fetch: fetchImpl,
    webSocketImpl: config.webSocketImpl,
    tcpProxyWsEndpoint,
    log: createLogger(verbose)
  };
  normalized.log(
    `config resolved: endpoint=${endpoint} authType=${normalized.authType}`
  );
  return normalized;
}
function deriveTcpProxyWsEndpoint(endpoint) {
  const url = new URL(endpoint);
  const host = url.hostname.startsWith("backboard.") ? `ssh.${url.hostname.slice("backboard.".length)}` : `ssh.${url.hostname}`;
  return `wss://${host}:${TCP_PROXY_WS_PORT}${TCP_PROXY_WS_PATH}`;
}
function deriveFilesWsEndpoint(tcpProxyWsEndpoint) {
  const url = new URL(tcpProxyWsEndpoint);
  url.pathname = url.pathname.endsWith(TCP_PROXY_WS_PATH) ? `${url.pathname.slice(0, -TCP_PROXY_WS_PATH.length)}${TCP_PROXY_FILES_WS_PATH}` : TCP_PROXY_FILES_WS_PATH;
  return url.toString();
}
function resolveWebSocketImpl(config) {
  const impl = config.webSocketImpl ?? globalThis.WebSocket;
  if (!impl) {
    throw new RailwayConnectionError({
      message: "No WebSocket implementation found. Pass `webSocketImpl` in the config (e.g. the `ws` package) to stream exec output."
    });
  }
  return impl;
}
function resolveEnvironmentId(explicit) {
  const environmentId = firstNonEmpty(
    explicit,
    readEnv(RAILWAY_ENVIRONMENT_ENV)
  );
  if (!environmentId) throw new RailwayAuthError(RAILWAY_ENVIRONMENT_ENV);
  return environmentId;
}
function readEnv(name) {
  if (typeof process === "undefined") return void 0;
  return process.env?.[name];
}
function isTruthyEnv(value) {
  if (value === void 0) return false;
  const normalized = value.trim().toLowerCase();
  return normalized === "1" || normalized === "true" || normalized === "yes";
}
function firstNonEmpty(...values) {
  for (const value of values) {
    if (value !== void 0 && value.trim().length > 0) return value;
  }
  return void 0;
}

// src/core/version.ts
var SDK_VERSION = "3.4.1" ;
var USER_AGENT = `railway-ts-sdk/${SDK_VERSION}`;

// src/core/graphql-client.ts
async function requestGraphQL(config, document, variables) {
  const operation = operationName(document);
  const start = Date.now();
  config.log(`\u2192 ${operation} POST ${config.endpoint}`);
  const response = await config.fetch(config.endpoint, {
    method: "POST",
    headers: {
      Accept: "application/json",
      ...authHeader(config),
      "Content-Type": "application/json",
      "User-Agent": USER_AGENT
    },
    body: JSON.stringify({
      query: graphql.print(document),
      variables
    })
  });
  const body = await parseGraphQLResponse(response);
  const errors = body?.errors ?? [];
  const elapsed = Date.now() - start;
  if (!response.ok || errors.length > 0) {
    config.log(
      `\u2190 ${operation} failed http=${response.status} errors=${errors.length} in ${elapsed}ms`
    );
    throw new RailwayGraphQLError({
      message: errors[0]?.message ?? `Railway GraphQL request failed with HTTP ${response.status}.`,
      status: response.status,
      errors,
      responseBody: body
    });
  }
  if (!body?.data) {
    config.log(`\u2190 ${operation} returned no data`);
    throw new RailwayGraphQLError({
      message: "Railway GraphQL response did not include data.",
      status: response.status,
      responseBody: body
    });
  }
  config.log(`\u2190 ${operation} ok in ${elapsed}ms`);
  return body.data;
}
function operationName(document) {
  for (const definition of document.definitions) {
    if (definition.kind === graphql.Kind.OPERATION_DEFINITION) {
      return definition.name?.value ?? "anonymous";
    }
  }
  return "operation";
}
function authHeader(config) {
  if (config.authType === "project-token") return { "project-access-token": config.token };
  return { Authorization: `Bearer ${config.token}` };
}
async function parseGraphQLResponse(response) {
  try {
    return await response.json();
  } catch {
    return void 0;
  }
}

// src/iac/client.ts
var IacClient = class {
  #config;
  constructor(config) {
    this.#config = normalizeRailwayClientConfig(config);
  }
  async getCurrentEnvironment(environmentId, options = {}) {
    const data = await gql(this.#config, `query IacEnvironmentConfig($environmentId: String!, $decryptVariables: Boolean) {
      environment(id: $environmentId) { id name projectId config(decryptVariables: $decryptVariables) configEtag }
    }`, { environmentId, decryptVariables: options.decryptVariables ?? false });
    const projectName = data.environment.projectId ? await this.getProjectName(data.environment.projectId) : void 0;
    const services = data.environment.projectId ? await this.getProjectServices(data.environment.projectId) : [];
    const volumes = data.environment.projectId ? await this.getProjectVolumes(data.environment.projectId) : [];
    const buckets = data.environment.projectId ? await this.getProjectBuckets(data.environment.projectId) : [];
    const customDomainsByServiceId = data.environment.projectId ? await this.getEnvironmentCustomDomains(data.environment.projectId, environmentId, services) : {};
    return {
      projectId: data.environment.projectId,
      projectName,
      environmentId: data.environment.id,
      environmentName: data.environment.name,
      config: data.environment.config ?? {},
      serviceNamesById: Object.fromEntries(services.map((service2) => [service2.id, service2.name])),
      volumeNamesById: Object.fromEntries(volumes.flatMap((volume2) => volume2.name ? [[volume2.id, volume2.name]] : [])),
      bucketNamesById: Object.fromEntries(buckets.map((bucket2) => [bucket2.id, bucket2.name])),
      customDomainsByServiceId,
      ...data.environment.configEtag ? { configEtag: data.environment.configEtag } : {}
    };
  }
  async getStagedPatch(environmentId, options = {}) {
    const data = await gql(this.#config, `query IacStagedPatch($environmentId: String!, $decryptVariables: Boolean) {
      environmentStagedChanges(environmentId: $environmentId) { id status patch(decryptVariables: $decryptVariables) meta }
    }`, { environmentId, decryptVariables: options.decryptVariables ?? false });
    return data.environmentStagedChanges;
  }
  async getProjectName(projectId) {
    const data = await gql(this.#config, `query IacProjectName($projectId: String!) {
      project(id: $projectId) { name }
    }`, { projectId });
    return data.project.name ?? void 0;
  }
  async getProjectServices(projectId) {
    const data = await gql(this.#config, `query IacProjectServices($projectId: String!) {
      project(id: $projectId) { services(first: 1000) { edges { node { id name } } } }
    }`, { projectId });
    return data.project.services.edges.map((edge) => edge.node);
  }
  async getProjectVolumes(projectId) {
    const data = await gql(this.#config, `query IacProjectVolumes($projectId: String!) {
      project(id: $projectId) { volumes(first: 1000) { edges { node { id name } } } }
    }`, { projectId });
    return data.project.volumes?.edges.map((edge) => edge.node) ?? [];
  }
  async getProjectBuckets(projectId) {
    const data = await gql(this.#config, `query IacProjectBuckets($projectId: String!) {
      project(id: $projectId) { buckets(first: 1000) { edges { node { id name } } } }
    }`, { projectId });
    return data.project.buckets.edges.map((edge) => edge.node);
  }
  async getEnvironmentCustomDomains(projectId, environmentId, services) {
    const entries = await Promise.all(services.map(async (service2) => {
      const data = await gql(this.#config, `query IacServiceDomains($projectId: String!, $environmentId: String!, $serviceId: String!) {
        domains(projectId: $projectId, environmentId: $environmentId, serviceId: $serviceId) { customDomains { domain targetPort } }
      }`, { projectId, environmentId, serviceId: service2.id }).catch(() => ({ domains: { customDomains: [] } }));
      return [service2.id, Object.fromEntries(data.domains.customDomains.map((domain) => [domain.domain, domain.targetPort == null ? {} : { port: domain.targetPort }]))];
    }));
    return Object.fromEntries(entries.filter(([, domains]) => Object.keys(domains).length > 0));
  }
  async ensureGraphResources({ projectId, environmentId, graph, currentConfig }) {
    const existing = await this.getProjectServices(projectId);
    const serviceIdsByName = Object.fromEntries(existing.map((service2) => [service2.name, service2.id]));
    const volumeIdsByServiceName = extractVolumeIdsByServiceName({ ...currentConfig ? { currentConfig } : {}, serviceIdsByName });
    for (const resource of graph.resources) {
      if (resource.type !== "service" && resource.type !== "database") continue;
      if (!serviceIdsByName[resource.name]) {
        const service2 = await this.createServiceForResource({ projectId, environmentId, resource });
        serviceIdsByName[resource.name] = service2.id;
      }
      if (resource.type === "database" && resource.defaultMountPath && !volumeIdsByServiceName[resource.name]) {
        const volume2 = await this.createVolumeForService({ projectId, environmentId, serviceId: serviceIdsByName[resource.name] });
        volumeIdsByServiceName[resource.name] = volume2.id;
      }
    }
    const bucketIdsByName = await this.ensureGraphBuckets({ projectId, graph });
    return { serviceIdsByName, volumeIdsByServiceName, bucketIdsByName };
  }
  async stageEnvironmentChanges({ environmentId, patch, merge = true }) {
    const data = await gql(this.#config, `mutation IacStageEnvironmentChanges($environmentId: String!, $payload: EnvironmentConfig!, $merge: Boolean) {
      environmentStageChanges(environmentId: $environmentId, input: $payload, merge: $merge) { id }
    }`, { environmentId, payload: patch, merge });
    return data.environmentStageChanges;
  }
  async previewChangeSet({ environmentId, changeSet }) {
    const data = await gql(this.#config, `mutation IacPreviewChangeSet($environmentId: String!, $input: JSON!) {
      environmentPreviewChangeSet(environmentId: $environmentId, input: $input) { changeSet diagnostics effects }
    }`, { environmentId, input: changeSet });
    return data.environmentPreviewChangeSet;
  }
  async applyChangeSet({ environmentId, changeSet, commitMessage, baseEtag }) {
    const variables = { environmentId, input: changeSet };
    if (commitMessage !== void 0) variables.commitMessage = commitMessage;
    if (baseEtag !== void 0) variables.baseConfigEtag = baseEtag;
    try {
      const data = await gql(this.#config, `mutation IacApplyChangeSet($environmentId: String!, $input: JSON!, $commitMessage: String, $baseConfigEtag: String) {
        environmentApplyChangeSet(environmentId: $environmentId, input: $input, commitMessage: $commitMessage, baseConfigEtag: $baseConfigEtag) { id status deploymentId stagedPatchId diagnostics changes { kind path summary status outputs } }
      }`, variables);
      return data.environmentApplyChangeSet;
    } catch (error) {
      if (isStaleBaseError(error)) {
        throw new StaleEnvironmentError("The environment changed since this plan was computed. Re-run plan and review the changes before applying.", { cause: error });
      }
      throw error;
    }
  }
  async commitStagedPatch({ environmentId, message, skipDeploys }) {
    const variables = { environmentId };
    if (message !== void 0) variables.message = message;
    if (skipDeploys !== void 0) variables.skipDeploys = skipDeploys;
    const data = await gql(this.#config, `mutation IacCommitStagedPatch($environmentId: String!, $message: String, $skipDeploys: Boolean) {
      environmentPatchCommitStaged(environmentId: $environmentId, commitMessage: $message, skipDeploys: $skipDeploys)
    }`, variables);
    return data.environmentPatchCommitStaged;
  }
  async ensureGraphBuckets({ projectId, graph }) {
    const existing = await this.getProjectBuckets(projectId);
    const idsByName = Object.fromEntries(existing.map((bucket2) => [bucket2.name, bucket2.id]));
    for (const resource of graph.resources) {
      if (resource.type !== "bucket" || idsByName[resource.name]) continue;
      const bucket2 = await this.createBucketForResource({ projectId, name: resource.name });
      idsByName[resource.name] = bucket2.id;
    }
    return idsByName;
  }
  async createBucketForResource({ projectId, name }) {
    const input = { projectId, environmentId: null };
    if (name !== void 0) input.name = name;
    const data = await gql(this.#config, `mutation IacBucketCreate($input: BucketCreateInput!) {
      bucketCreate(input: $input) { id name }
    }`, { input });
    return data.bucketCreate;
  }
  async createVolumeForService({ projectId, serviceId }) {
    const input = { projectId, environmentId: null, serviceId, mountPath: "/var/lib/postgresql/data" };
    const data = await gql(this.#config, `mutation IacVolumeCreate($input: VolumeCreateInput!) {
      volumeCreate(input: $input) { id name }
    }`, { input });
    return data.volumeCreate;
  }
  async createServiceForResource({ projectId, resource }) {
    const source = resource.type === "database" ? { image: resource.image } : resource.source?.type === "image" && resource.source.image ? { image: resource.source.image } : resource.source?.type === "github" && resource.source.repo ? { repo: resource.source.repo } : void 0;
    const input = {
      name: resource.name,
      projectId,
      environmentId: null
    };
    if (source) input.source = source;
    if (resource.type === "service" && resource.source?.type === "github" && resource.source.branch) {
      input.branch = resource.source.branch;
    }
    const data = await gql(this.#config, `mutation IacServiceCreate($input: ServiceCreateInput!) {
      serviceCreate(input: $input) { id name }
    }`, { input });
    return data.serviceCreate;
  }
};
async function gql(config, source, variables) {
  return requestGraphQL(config, graphql.parse(source), variables);
}
function isStaleBaseError(error) {
  if (!(error instanceof RailwayGraphQLError)) return false;
  return error.errors.some((item) => item.extensions?.code === STALE_ENVIRONMENT_BASE_CODE);
}
function extractVolumeIdsByServiceName({ currentConfig, serviceIdsByName }) {
  const volumeIdsByServiceName = {};
  for (const [serviceName, serviceId] of Object.entries(serviceIdsByName)) {
    const volumeId = Object.keys(currentConfig?.services?.[serviceId]?.volumeMounts ?? {})[0];
    if (volumeId) volumeIdsByServiceName[serviceName] = volumeId;
  }
  return volumeIdsByServiceName;
}

// src/iac/diff.ts
function renderPatchDiff({
  currentConfig,
  patch,
  title = "Railway patch"
}) {
  const lines = [];
  const changes = flattenChanges(currentConfig, patch);
  lines.push(title);
  lines.push("".padEnd(title.length, "\u2500"));
  lines.push("");
  if (changes.length === 0) {
    lines.push("No changes.");
    return lines.join("\n");
  }
  for (const change of changes) lines.push(formatChange(change));
  lines.push("");
  lines.push(`${changes.length} raw change${changes.length === 1 ? "" : "s"}.`);
  return lines.join("\n").trimEnd();
}
function numChangesInEnvironmentConfig(config) {
  return Object.keys(flatten(config)).length;
}
function flattenChanges(currentConfig, patch) {
  const current = flatten(currentConfig);
  const next = flatten(patch);
  return Object.entries(next).map(([path3, value]) => {
    if (value === null || path3.endsWith(".isDeleted")) return { type: "removed", path: path3, previous: current[path3] };
    if (current[path3] == null) return { type: "added", path: path3, next: value };
    return { type: "changed", path: path3, previous: current[path3], next: value };
  });
}
function flatten(value, prefix = "") {
  if (value == null || typeof value !== "object" || Array.isArray(value)) return prefix ? { [prefix]: value } : {};
  const entries = Object.entries(value);
  if (entries.length === 0) return prefix ? { [prefix]: {} } : {};
  return Object.fromEntries(entries.flatMap(([key, child]) => Object.entries(flatten(child, prefix ? `${prefix}.${key}` : key))));
}
function formatChange(change) {
  const prefix = change.type === "added" ? green("+") : change.type === "removed" ? red("-") : amber("~");
  if (change.type === "added") return `  ${prefix} ${change.path}: ${format(change.next)}`;
  if (change.type === "removed") return `  ${prefix} ${change.path}: ${format(change.previous)}`;
  return `  ${prefix} ${change.path}: ${format(change.previous)} -> ${format(change.next)}`;
}
function format(value) {
  return typeof value === "string" ? value : JSON.stringify(value);
}
function green(text) {
  return color(text, "32");
}
function red(text) {
  return color(text, "31");
}
function amber(text) {
  return color(text, "33");
}
function color(text, code) {
  if (process.env.NO_COLOR) return text;
  return `\x1B[${code}m${text}\x1B[0m`;
}
function defineRailway(program) {
  return program;
}
var define = defineRailway;
function project(name, definition) {
  const resources = (definition.resources ?? definition.services ?? []).flat();
  return { name, ...definition, resources };
}
function createRailwayContext(input = {}) {
  const environment = input.environment ?? input.environmentName;
  return {
    ...input,
    ...environment ? { environment, environmentName: environment } : {},
    randomString: (label = "random", bytes = 12) => crypto.createHash("sha256").update(`railway-iac:${environment ?? "default"}:${label}`).digest("hex").slice(0, bytes * 2),
    isEnvironment: (name) => environment === name,
    shared: createSharedVariableRefs()
  };
}
function github(repo, options = {}) {
  return { type: "github", repo, branch: options.branch ?? "main", ...options };
}
function image(imageName, options = {}) {
  return { type: "image", image: imageName, ...options };
}
function template(templateName, options = {}) {
  return { type: "template", template: templateName, ...options };
}
function empty() {
  return { type: "empty" };
}
function service(name, config = {}) {
  const source = normalizeSource(config.source, config.root ?? config.rootDirectory);
  const node = pruneEmpty2({
    address: resourceAddress("service", name),
    type: "service",
    kind: source?.type === "github" ? "github" : source?.type === "image" ? "docker-image" : source?.type === "template" ? "template" : "empty",
    name,
    ...source ? { source } : {},
    build: normalizeBuild(config),
    deploy: normalizeDeploy(config),
    networking: normalizeNetworking(config),
    ...config.env || config.variables ? { variables: normalizeVariables({ ...config.variables ?? {}, ...config.env ?? {} }) } : {},
    ...normalizeVolumeMounts(config.volumeMounts),
    ...config.configFile ? { configFile: config.configFile } : {},
    ...config.parentServiceId ? { parentServiceId: config.parentServiceId } : {},
    ...config.groupId ? { groupId: config.groupId } : {},
    ...config.clusterRole ? { clusterRole: config.clusterRole } : {},
    ...config.replicaConfig ? { replicaConfig: config.replicaConfig } : {},
    ...config.clusterDisplay ? { clusterDisplay: config.clusterDisplay } : {}
  });
  return withVariableRefs(node);
}
function fn(name, config = {}) {
  return withVariableRefs({ ...service(name, config), kind: "function" });
}
function postgres(name, config = {}) {
  return database(name, "postgres", {
    image: "ghcr.io/railwayapp-templates/postgres-ssl:18",
    output: "DATABASE_URL",
    defaultMountPath: "/var/lib/postgresql/data",
    ...config
  });
}
function mysql(name, config = {}) {
  return database(name, "mysql", { image: "mysql:9", output: "MYSQL_URL", defaultMountPath: "/var/lib/mysql", ...config });
}
function redis(name, config = {}) {
  return database(name, "redis", { image: "railwayapp/redis:8.2", output: "REDIS_URL", defaultMountPath: "/bitnami", ...config });
}
function mongo(name, config = {}) {
  return database(name, "mongo", { image: "mongo:8", output: "MONGO_URL", defaultMountPath: "/data/db", ...config });
}
function database(name, engine, options) {
  const output = options.output ?? "DATABASE_URL";
  const node = {
    address: resourceAddress("database", name),
    type: "database",
    kind: "database",
    engine,
    name,
    image: options.image,
    output,
    defaultMountPath: options.defaultMountPath,
    source: image(options.image),
    ...options.region ? { deploy: { multiRegionConfig: { [options.region]: { numReplicas: 1 } } } } : {}
  };
  return withVariableRefs(node);
}
function volume(name, config = {}) {
  return { address: resourceAddress("volume", name), type: "volume", name, config };
}
function bucket(name, config = {}) {
  return { address: resourceAddress("bucket", name), type: "bucket", name, config };
}
function group(name, resourcesOrOptions = {}, maybeOptions = {}) {
  const resources = Array.isArray(resourcesOrOptions) ? resourcesOrOptions.flat() : void 0;
  const options = Array.isArray(resourcesOrOptions) ? maybeOptions : resourcesOrOptions;
  const node = { address: resourceAddress("group", name), type: "group", name, ...options };
  if (!resources) return node;
  return [node, ...resources.map((resource) => ({ ...resource, groupId: name }))];
}
function ref(resource, output) {
  return variableReference(resource, output);
}
function preserve() {
  return { type: "preserve" };
}
function variableReference(resource, output) {
  return { type: "reference", resource: resource.address, output };
}
function sharedVariableReference(name) {
  return { type: "sharedReference", name };
}
function createSharedVariableRefs() {
  return new Proxy({}, {
    get(_target, property) {
      if (typeof property !== "string") return void 0;
      return sharedVariableReference(property);
    }
  });
}
function createVariableRefAccessor(resource) {
  return new Proxy({}, {
    get(_target, property) {
      if (typeof property !== "string") return void 0;
      return variableReference(resource, property);
    }
  });
}
function withVariableRefs(node) {
  Object.defineProperty(node, "env", { enumerable: false, value: createVariableRefAccessor(node) });
  return node;
}
function normalizeSource(source, rootDirectory) {
  if (source == null) return rootDirectory ? { type: "empty", rootDirectory } : void 0;
  if ("type" in source && source.type) return pruneEmpty2({ ...source, rootDirectory: source.rootDirectory ?? rootDirectory });
  if ("repo" in source && source.repo) return pruneEmpty2({ type: "github", repo: source.repo, branch: source.branch ?? "main", rootDirectory });
  if ("image" in source && source.image) return pruneEmpty2({ type: "image", image: source.image, rootDirectory });
  return rootDirectory ? { type: "empty", rootDirectory } : void 0;
}
function normalizeBuild(config) {
  if (typeof config.build === "string") return { buildCommand: config.build };
  return pruneEmpty2({ ...config.build });
}
function normalizeDeploy(config) {
  const preDeployCommand = config.preDeploy ?? config.preDeployCommand ?? config.run?.preDeploy;
  return pruneEmpty2({
    ...config.deploy,
    startCommand: config.start ?? config.startCommand ?? config.run?.command ?? config.deploy?.startCommand,
    preDeployCommand: Array.isArray(preDeployCommand) ? preDeployCommand : preDeployCommand ? [preDeployCommand] : config.deploy?.preDeployCommand,
    healthcheckPath: config.healthcheck ?? config.healthcheckPath ?? config.run?.healthcheck ?? config.deploy?.healthcheckPath,
    healthcheckTimeout: config.healthcheckTimeout ?? config.run?.healthcheckTimeout ?? config.deploy?.healthcheckTimeout,
    multiRegionConfig: normalizeReplicas(config.replicas, config.regions) ?? config.deploy?.multiRegionConfig
  });
}
function normalizeReplicas(replicas, regions) {
  if (typeof replicas === "number") return normalizeRegions({ "us-west2": replicas });
  if (replicas) return normalizeRegions(replicas);
  if (regions) return normalizeRegions(regions);
  return void 0;
}
function normalizeRegions(regions) {
  return Object.fromEntries(
    Object.entries(regions).map(([region, value]) => [
      region,
      typeof value === "number" ? { numReplicas: value } : pruneEmpty2({ numReplicas: value.count ?? value.replicas, stackerAssignment: value.stacker }) ?? {}
    ])
  );
}
function normalizeNetworking(config) {
  const customDomains = config.domains ? Object.fromEntries(config.domains.map((domain) => typeof domain === "string" ? [domain, { port: 8080 }] : [domain.domain, { port: domain.port ?? 8080 }])) : void 0;
  const tcpProxies = config.tcp ? Object.fromEntries(config.tcp.map((port) => [String(port), {}])) : config.tcpProxies ? Object.fromEntries(config.tcpProxies.map((port) => [port, {}])) : void 0;
  return pruneEmpty2({ ...config.networking, customDomains, tcpProxies });
}
function normalizeVariables(variables) {
  return Object.fromEntries(
    Object.entries(variables).map(([key, value]) => {
      if (typeof value === "string") return [key, { type: "literal", value }];
      if ("type" in value) return [key, value];
      return [key, { type: "raw", value }];
    })
  );
}
function normalizeVolumeMounts(volumeMounts) {
  if (!volumeMounts) return {};
  const rawMounts = {};
  const attachments = {};
  for (const [key, value] of Object.entries(volumeMounts)) {
    if (value && typeof value === "object" && "type" in value && value.type === "volume") {
      attachments[value.name] = { volume: value.address, mountPath: key };
      continue;
    }
    rawMounts[key] = value;
  }
  return pruneEmpty2({ volumeMounts: rawMounts, volumeAttachments: attachments }) ?? {};
}
function pruneEmpty2(value) {
  if (value == null || typeof value !== "object") return value;
  const entries = Object.entries(value).filter(([, child]) => child != null);
  if (entries.length === 0) return void 0;
  return Object.fromEntries(entries);
}

// src/iac/evaluator.ts
async function evaluateRailwayFile(filePath, options = {}) {
  const absolutePath = path2__default.default.resolve(filePath);
  const mod = await importRailwayFile(absolutePath);
  const exported = mod.default ?? mod;
  const definition = await resolveDefinition(exported, options.context);
  const graph = projectDefinitionToGraph(definition);
  const desiredConfig = graphToEnvironmentConfig(graph, options);
  return { graph, desiredConfig };
}
async function importRailwayFile(absolutePath) {
  const url$1 = `${url.pathToFileURL(absolutePath).toString()}?t=${Date.now()}`;
  if (/\.[cm]?tsx?$/.test(absolutePath)) {
    const { tsImport } = await import('tsx/esm/api');
    return tsImport(url$1, (typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('index.cjs', document.baseURI).href)));
  }
  return import(url$1);
}
async function resolveDefinition(exported, context = {}) {
  if (typeof exported === "function") return exported(createRailwayContext(context), project);
  return exported;
}
async function evaluateRailwayProject(options = {}) {
  const file = options.file ? path2__default.default.resolve(options.file) : findRailwayFile(options.cwd);
  const { graph, desiredConfig } = await evaluateRailwayFile(file, options);
  return new EvaluatedRailwayProject({ file, graph, desiredConfig });
}
function findRailwayFile(cwd = process.cwd()) {
  let current = path2__default.default.resolve(cwd);
  while (true) {
    const candidate = path2__default.default.join(current, ".railway", "railway.ts");
    if (fs__default.default.existsSync(candidate)) return candidate;
    const parent = path2__default.default.dirname(current);
    if (parent === current) {
      throw new Error(
        `Could not find .railway/railway.ts from ${path2__default.default.resolve(cwd)}. Pass { file } explicitly.`
      );
    }
    current = parent;
  }
}
var EvaluatedRailwayProject = class {
  file;
  graph;
  desiredConfig;
  constructor(snapshot) {
    this.file = snapshot.file;
    this.graph = snapshot.graph;
    this.desiredConfig = snapshot.desiredConfig;
  }
  get name() {
    return this.graph.project.name;
  }
  get resources() {
    return this.graph.resources;
  }
  get services() {
    return this.graph.resources.filter(
      (resource) => resource.type === "service" || resource.type === "database"
    );
  }
  resource(name) {
    const resource = this.graph.resources.find((candidate) => candidate.name === name);
    if (!resource) throw new Error(`Railway resource not found: ${name}`);
    return resource;
  }
  service(name) {
    const resource = this.graph.resources.find(
      (candidate) => candidate.name === name && (candidate.type === "service" || candidate.type === "database")
    );
    if (!resource) throw new Error(`Railway service not found: ${name}`);
    return resource;
  }
  database(name) {
    const resource = this.graph.resources.find(
      (candidate) => candidate.name === name && candidate.type === "database"
    );
    if (!resource) throw new Error(`Railway database not found: ${name}`);
    return resource;
  }
  bucket(name) {
    const resource = this.graph.resources.find(
      (candidate) => candidate.name === name && candidate.type === "bucket"
    );
    if (!resource) throw new Error(`Railway bucket not found: ${name}`);
    return resource;
  }
  volume(name) {
    const resource = this.graph.resources.find(
      (candidate) => candidate.name === name && candidate.type === "volume"
    );
    if (!resource) throw new Error(`Railway volume not found: ${name}`);
    return resource;
  }
  config(options = {}) {
    return graphToEnvironmentConfig(this.graph, options);
  }
  toJSON() {
    return {
      file: this.file,
      graph: this.graph,
      desiredConfig: this.desiredConfig
    };
  }
};

// src/iac/typegen.ts
function renderRailwayGraphTypes(graph) {
  const services = names(
    graph.resources.filter(
      (resource) => resource.type === "service" || resource.type === "database"
    )
  );
  const databases = names(graph.resources.filter((resource) => resource.type === "database"));
  const buckets = names(graph.resources.filter((resource) => resource.type === "bucket"));
  const volumes = names(graph.resources.filter((resource) => resource.type === "volume"));
  const resources = names(graph.resources);
  return `// Generated by Railway IaC. Do not edit by hand.
// Provides autocomplete for evaluated graph resource lookups.

import "railway/iac";

declare module "railway/iac" {
  interface RailwayIacServiceMap ${mapBody(services)}
  interface RailwayIacDatabaseMap ${mapBody(databases)}
  interface RailwayIacBucketMap ${mapBody(buckets)}
  interface RailwayIacVolumeMap ${mapBody(volumes)}
  interface RailwayIacResourceMap ${mapBody(resources)}
}
`;
}
function names(resources) {
  return [...new Set(resources.map((resource) => resource.name))].sort();
}
function mapBody(names2) {
  if (names2.length === 0) return "{}";
  return `{
${names2.map((name) => `    ${JSON.stringify(name)}: true;`).join("\n")}
  }`;
}

// src/iac/runner.ts
async function runRailwayIac(request = {}) {
  const command = request.command ?? "evaluate";
  const cwd = request.cwd ?? process.cwd();
  const file = request.file ?? await findRailwayFile(cwd);
  try {
    const evaluated = await evaluateRailwayProject({ file, context: runnerContext(request, command) });
    const input = {
      request,
      file: evaluated.file,
      graph: evaluated.graph,
      diagnostics: graphDiagnostics(evaluated.graph)
    };
    return await runEvaluatedCommand(command, input);
  } catch (error) {
    return errorResponse(command, file, error);
  }
}
async function runEvaluatedCommand(command, input) {
  switch (command) {
    case "typegen":
      return typegenRailwayIac(input);
    case "evaluate":
      return evaluateRailwayIac(input);
    case "current":
      return currentRailwayIac(input);
    case "plan":
      return planRailwayIac(input);
    case "stage":
      return stageRailwayIac(input);
    case "apply":
      return applyRailwayIac(input);
  }
}
function typegenRailwayIac({ file, graph, diagnostics }) {
  return {
    ok: hasNoErrors(diagnostics),
    command: "typegen",
    file,
    graphTypes: renderRailwayGraphTypes(graph),
    diagnostics
  };
}
function evaluateRailwayIac({ request, file, graph, diagnostics }) {
  return {
    ok: hasNoErrors(diagnostics),
    command: "evaluate",
    file,
    graph,
    ...request.includeTypes ? { graphTypes: renderRailwayGraphTypes(graph) } : {},
    diagnostics
  };
}
async function currentRailwayIac({ file, graph: desiredGraph, request, diagnostics }) {
  const context = requireBackboardContext(request.backboard, "plan");
  const client = new IacClient(clientConfig(context));
  const current = await readCurrentEnvironment(client, context);
  const currentGraph = graphFromCurrentEnvironment(current, desiredGraph);
  const { config: _config, ...currentEnvironment } = current;
  return {
    ok: hasNoErrors(diagnostics),
    command: "current",
    file,
    mode: "real",
    currentGraph,
    currentConfig: current.config,
    currentEnvironment,
    ...request.includeTypes ? { graphTypes: renderRailwayGraphTypes(currentGraph) } : {},
    diagnostics
  };
}
async function planRailwayIac({ file, graph: desiredGraph, request, diagnostics }) {
  const context = requireBackboardContext(request.backboard, "plan");
  const client = new IacClient(clientConfig(context));
  const current = await readCurrentEnvironment(client, context);
  const currentGraph = graphFromCurrentEnvironment(current, desiredGraph);
  const changeSet = diffGraphs({ current: currentGraph, desired: desiredGraph, revealValues: request.revealValues ?? false });
  const allDiagnostics = [...diagnostics, ...changeSetDiagnostics(changeSet)];
  const { config: _config, ...currentEnvironment } = current;
  const hasErrors = !hasNoErrors(allDiagnostics);
  const preview = !hasErrors && changeSet.changes.length > 0 ? await client.previewChangeSet({ environmentId: context.environmentId, changeSet }) : void 0;
  const previewedChangeSet = preview?.changeSet ?? changeSet;
  return {
    ok: !hasErrors,
    command: "plan",
    file,
    mode: "real",
    currentGraph,
    desiredGraph,
    currentConfig: current.config,
    currentEnvironment,
    changeSet: previewedChangeSet,
    ...preview ? { preview } : {},
    diff: renderChangeSet(previewedChangeSet),
    ...request.includeTypes ? { graphTypes: renderRailwayGraphTypes(desiredGraph) } : {},
    diagnostics: allDiagnostics
  };
}
async function stageRailwayIac(input) {
  const planned = await planRailwayIac(input);
  if (!planned.ok || !planned.changeSet || planned.changeSet.changes.length === 0) {
    return { ...planned, command: "stage" };
  }
  requireBackboardContext(input.request.backboard, "stage");
  return { ...planned, command: "stage" };
}
async function applyRailwayIac(input) {
  const planned = await planRailwayIac(input);
  if (!planned.ok || !planned.changeSet || planned.changeSet.changes.length === 0) {
    return { ...planned, command: "apply" };
  }
  const context = requireBackboardContext(input.request.backboard, "apply");
  const client = new IacClient(clientConfig(context));
  const applyResult = await client.applyChangeSet({
    environmentId: context.environmentId,
    changeSet: planned.changeSet,
    commitMessage: "Apply Railway configuration",
    // Base snapshot the plan was diffed against; backboard rejects a stale apply.
    ...planned.currentEnvironment?.configEtag ? { baseEtag: planned.currentEnvironment.configEtag } : {}
  });
  return {
    ...planned,
    command: "apply",
    applyResult,
    ...applyResult.deploymentId ? { deploymentId: applyResult.deploymentId } : {},
    ...applyResult.stagedPatchId ? { stagedPatchId: applyResult.stagedPatchId } : {}
  };
}
async function readCurrentEnvironment(client, context) {
  return client.getCurrentEnvironment(
    context.environmentId,
    context.decryptVariables === void 0 ? {} : { decryptVariables: context.decryptVariables }
  );
}
function graphFromCurrentEnvironment(current, desiredGraph) {
  return environmentConfigToGraph(current.config, {
    projectName: current.projectName ?? desiredGraph.project.name,
    serviceNamesById: current.serviceNamesById,
    volumeNamesById: current.volumeNamesById,
    bucketNamesById: current.bucketNamesById,
    customDomainsByServiceId: current.customDomainsByServiceId
  });
}
function changeSetDiagnostics(changeSet) {
  return changeSet.diagnostics.map((diagnostic) => ({
    severity: diagnostic.severity,
    path: diagnostic.path,
    message: diagnostic.message
  }));
}
function runnerContext(request, command) {
  const environment = request.context?.environment ?? request.context?.environmentName;
  const projectId = request.context?.projectId ?? request.backboard?.projectId;
  const environmentId = request.context?.environmentId ?? request.backboard?.environmentId;
  return {
    ...request.context,
    command,
    ...projectId ? { projectId } : {},
    ...environmentId ? { environmentId } : {},
    ...environment ? { environment, environmentName: environment } : {}
  };
}
function requireBackboardContext(context, command) {
  if (!context?.token) throw new Error(`Backboard token is required for ${command}.`);
  if (!context.environmentId) throw new Error(`Backboard environmentId is required for ${command}.`);
  return context;
}
function clientConfig(context) {
  return { token: context.token, ...context.authType ? { authType: context.authType } : {}, ...context.endpoint ? { graphqlEndpoint: context.endpoint } : {} };
}
function graphDiagnostics(graph) {
  return validateGraph(graph).map((message) => ({ severity: "error", path: "graph", message }));
}
function hasNoErrors(diagnostics) {
  return diagnostics.every((diagnostic) => diagnostic.severity !== "error");
}
function errorResponse(command, file, error) {
  return {
    ok: false,
    command,
    file,
    ...command === "plan" || command === "stage" || command === "apply" || command === "current" ? { mode: "real" } : {},
    diagnostics: [{ severity: "error", path: "", message: error instanceof Error ? error.message : String(error) }]
  };
}

// src/generated/graphql.ts
var RailwaySandboxDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "RailwaySandbox" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandbox" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "id" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxFields" } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "Sandbox" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "networkIsolation" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }, { "kind": "Field", "name": { "kind": "Name", "value": "region" } }, { "kind": "Field", "name": { "kind": "Name", "value": "idleTimeoutMinutes" } }, { "kind": "Field", "name": { "kind": "Name", "value": "createdAt" } }] } }] };
var RailwaySandboxesDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "RailwaySandboxes" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "first" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "Int" } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "after" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "before" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "last" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "Int" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxes" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "first" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "first" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "after" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "after" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "before" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "before" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "last" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "last" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "edges" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "node" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxFields" } }] } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "pageInfo" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "hasNextPage" } }, { "kind": "Field", "name": { "kind": "Name", "value": "endCursor" } }] } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "Sandbox" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "networkIsolation" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }, { "kind": "Field", "name": { "kind": "Name", "value": "region" } }, { "kind": "Field", "name": { "kind": "Name", "value": "idleTimeoutMinutes" } }, { "kind": "Field", "name": { "kind": "Name", "value": "createdAt" } }] } }] };
var RailwaySandboxCreateDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "mutation", "name": { "kind": "Name", "value": "RailwaySandboxCreate" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "input" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "SandboxCreateInput" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxCreate" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "input" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "input" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxFields" } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "Sandbox" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "networkIsolation" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }, { "kind": "Field", "name": { "kind": "Name", "value": "region" } }, { "kind": "Field", "name": { "kind": "Name", "value": "idleTimeoutMinutes" } }, { "kind": "Field", "name": { "kind": "Name", "value": "createdAt" } }] } }] };
var RailwaySandboxDestroyDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "mutation", "name": { "kind": "Name", "value": "RailwaySandboxDestroy" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxDestroy" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "id" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxFields" } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "Sandbox" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "networkIsolation" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }, { "kind": "Field", "name": { "kind": "Name", "value": "region" } }, { "kind": "Field", "name": { "kind": "Name", "value": "idleTimeoutMinutes" } }, { "kind": "Field", "name": { "kind": "Name", "value": "createdAt" } }] } }] };
var RailwaySandboxTemplateBuildDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "mutation", "name": { "kind": "Name", "value": "RailwaySandboxTemplateBuild" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "input" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "SandboxTemplateInput" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxTemplateBuild" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "input" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "input" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxTemplateBuildFields" } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxTemplateBuildFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "SandboxTemplateBuild" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }] } }] };
var RailwaySandboxTemplateBuildStatusDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "RailwaySandboxTemplateBuildStatus" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "ID" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxTemplateBuild" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "id" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxTemplateBuildFields" } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxTemplateBuildFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "SandboxTemplateBuild" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }] } }] };
var RailwaySandboxCheckpointsDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "RailwaySandboxCheckpoints" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxCheckpoints" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointFields" } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "SandboxCheckpoint" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "key" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }, { "kind": "Field", "name": { "kind": "Name", "value": "createdAt" } }] } }] };
var RailwaySandboxCheckpointCreateDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "mutation", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointCreate" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "name" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "sandboxId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxCheckpointCreate" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "name" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "name" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "sandboxId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "sandboxId" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointFields" } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "SandboxCheckpoint" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "key" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }, { "kind": "Field", "name": { "kind": "Name", "value": "createdAt" } }] } }] };
var RailwaySandboxCheckpointRenameDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "mutation", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointRename" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "ID" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "name" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxCheckpointRename" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "id" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "name" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "name" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "FragmentSpread", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointFields" } }] } }] } }, { "kind": "FragmentDefinition", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointFields" }, "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "SandboxCheckpoint" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "key" } }, { "kind": "Field", "name": { "kind": "Name", "value": "environmentId" } }, { "kind": "Field", "name": { "kind": "Name", "value": "createdAt" } }] } }] };
var RailwaySandboxCheckpointDeleteDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "mutation", "name": { "kind": "Name", "value": "RailwaySandboxCheckpointDelete" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "String" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "ID" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "sandboxCheckpointDelete" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "environmentId" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "environmentId" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "id" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "id" } } }] }] } }] };
var RailwayGenerateShellTokenDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "mutation", "name": { "kind": "Name", "value": "RailwayGenerateShellToken" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "input" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "ShellTokenInput" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "generateShellToken" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "input" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "input" } } }] }] } }] };

// src/sandbox/errors.ts
var SandboxNotFoundError = class extends RailwayError {
  id;
  environmentId;
  constructor(args) {
    super(
      `Sandbox "${args.id}" was not found in environment "${args.environmentId}".`
    );
    this.id = args.id;
    this.environmentId = args.environmentId;
  }
};
var SandboxFailedError = class extends RailwayError {
  id;
  status;
  constructor(args) {
    super(
      `Sandbox "${args.id}" entered terminal state "${args.status}" before becoming ready.`
    );
    this.id = args.id;
    this.status = args.status;
  }
};
var SandboxTemplateBuildError = class extends RailwayError {
  templateId;
  environmentId;
  constructor(args) {
    super(
      `Sandbox template "${args.templateId}" failed to build in environment "${args.environmentId}".`
    );
    this.templateId = args.templateId;
    this.environmentId = args.environmentId;
  }
};
var SandboxFilesError = class extends RailwayError {
  operation;
  path;
  constructor(args) {
    super(`File ${args.operation} failed for "${args.path}": ${args.message}`);
    this.operation = args.operation;
    this.path = args.path;
  }
};
var SandboxFileNotFoundError = class extends SandboxFilesError {
};
var SandboxTimeoutError = class extends RailwayError {
  resource;
  id;
  lastStatus;
  timeoutMs;
  constructor(args) {
    super(
      `Timed out after ${args.timeoutMs}ms waiting for ${args.resource} "${args.id}" to become ready (last status: ${args.lastStatus}).`
    );
    this.resource = args.resource;
    this.id = args.id;
    this.lastStatus = args.lastStatus;
    this.timeoutMs = args.timeoutMs;
  }
};

// src/core/exec-ws-client.ts
var STDOUT_FRAME = 1;
var STDERR_FRAME = 3;
var SHELL_SUBPROTOCOL = "railway-shell";
function connectExecWs(args) {
  const { config, jwt, command, cwd, env, sessionName, resumeFromLastRead, handlers } = args;
  const WS = resolveWebSocketImpl(config);
  return new Promise((resolve, reject) => {
    let opened = false;
    const socket = new WS(config.tcpProxyWsEndpoint, [
      SHELL_SUBPROTOCOL,
      jwt
    ]);
    socket.binaryType = "arraybuffer";
    socket.onopen = () => {
      opened = true;
      const data = { command };
      if (cwd) data.cwd = cwd;
      if (env && Object.keys(env).length > 0) data.env = env;
      if (sessionName) data.durable_session_name = sessionName;
      if (resumeFromLastRead) data.resume_from_last_read = true;
      socket.send(JSON.stringify({ type: "init_exec", data }));
      resolve({
        closeStdin: () => socket.send(JSON.stringify({ type: "stdin_close" })),
        signal: (name) => socket.send(JSON.stringify({ type: "signal", data: { signal: name } })),
        close: () => socket.close(1e3, "")
      });
    };
    socket.onmessage = (event) => {
      const { data } = event;
      if (data instanceof ArrayBuffer) handleBinaryFrame(data, handlers);
      else if (typeof data === "string") handleTextFrame(data, handlers);
    };
    socket.onclose = (event) => {
      if (!opened) {
        reject(
          new RailwayConnectionError({
            message: `tcp-proxy exec WebSocket closed before opening (code ${event.code}${event.reason ? `: ${event.reason}` : ""}).`,
            closeCode: event.code
          })
        );
        return;
      }
      handlers.onClose({ code: event.code, reason: event.reason });
    };
    socket.onerror = (event) => {
      if (opened) return;
      reject(
        new RailwayConnectionError({
          message: "tcp-proxy exec WebSocket connection failed.",
          cause: event
        })
      );
    };
  });
}
function handleBinaryFrame(buffer, handlers) {
  const view = new Uint8Array(buffer);
  if (view.length <= 1) return;
  if (view[0] === STDOUT_FRAME) handlers.onStdout(view.subarray(1));
  else if (view[0] === STDERR_FRAME) handlers.onStderr(view.subarray(1));
}
function handleTextFrame(text, handlers) {
  let frame;
  try {
    frame = JSON.parse(text);
  } catch {
    return;
  }
  if (frame.type === "exit") emitExit(frame.data, handlers);
  else if (frame.type === "durable_session") {
    emitDurableSession(frame.data, handlers);
  }
}
function emitExit(data, handlers) {
  handlers.onExit(data?.exit_code ?? 0, data?.reason ?? "");
}
function emitDurableSession(data, handlers) {
  if (data?.durable_session_name) {
    handlers.onDurableSession?.(data.durable_session_name);
  }
}

// src/sandbox/exec.ts
var decoder = () => new TextDecoder();
var REATTACH_PLACEHOLDER_COMMAND = ":";
var constructHandle;
var ExecHandle = class _ExecHandle {
  /** Durable session name for this exec; reattach via `exec({ sessionName })`. */
  sessionName;
  [Symbol.toStringTag] = "ExecHandle";
  #result;
  #kill;
  #detach;
  /** Constructed by `Sandbox.exec`; not constructible from outside the SDK. */
  constructor(args) {
    this.sessionName = args.sessionName;
    this.#result = args.result;
    this.#kill = args.kill;
    this.#detach = args.detach;
    this.sessionName.catch(() => {
    });
    this.#result.catch(() => {
    });
  }
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  then(onfulfilled, onrejected) {
    return this.#result.then(onfulfilled, onrejected);
  }
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  catch(onrejected) {
    return this.#result.catch(onrejected);
  }
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  finally(onfinally) {
    return this.#result.finally(onfinally);
  }
  /** The final result as a plain promise; identical to awaiting the handle. */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  result() {
    return this.#result;
  }
  /**
   * Terminate the running command with a signal (default `TERM`) delivered to
   * its process group — a real kill, regardless of durable sessions. The handle
   * then settles with the command's exit (a signalled process reports exit code
   * `-1`). To stop streaming WITHOUT ending the command, use `detach()`.
   */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  kill(signal = "TERM") {
    return this.#kill(signal);
  }
  /**
   * Stop streaming and close the WebSocket without ending the command — the
   * durable session keeps running server-side. Resolves with its `sessionName`
   * so you can reattach later via `exec({ sessionName })`; rejects if the server
   * assigned no durable session (reattach is impossible). The handle itself
   * settles with the output captured up to the detach.
   */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  detach() {
    return this.#detach();
  }
  static {
    constructHandle = (args) => new _ExecHandle(args);
  }
};
function startExec(context, target, options) {
  if (typeof target !== "string" && (options.cwd !== void 0 || options.env && Object.keys(options.env).length > 0)) {
    throw new RailwayError(
      "cwd/env apply only to fresh execs; a reattached session is already running."
    );
  }
  let resolveSessionName;
  let rejectSessionName;
  const sessionName = new Promise((resolve, reject) => {
    resolveSessionName = resolve;
    rejectSessionName = reject;
  });
  const control = { detached: false };
  const kill = async (signal) => {
    if (!control.connection) {
      control.pendingSignal = signal;
      return true;
    }
    try {
      control.connection.signal(signal);
      return true;
    } catch {
      return false;
    }
  };
  const detach = async () => {
    control.detached = true;
    control.connection?.close();
    return sessionName;
  };
  const result = runExec(
    context,
    target,
    options,
    resolveSessionName,
    rejectSessionName,
    control
  ).catch((error) => {
    rejectSessionName(error);
    throw error;
  });
  return constructHandle({ sessionName, result, kill, detach });
}
async function runExec(context, target, options, onSessionName, onNoSessionName, control) {
  const reattach = typeof target !== "string";
  const command = reattach ? REATTACH_PLACEHOLDER_COMMAND : target;
  const sessionName = reattach ? target.sessionName : void 0;
  let sessionNameResolved = false;
  const resolveSessionNameOnce = (name) => {
    if (sessionNameResolved) return;
    sessionNameResolved = true;
    onSessionName(name);
  };
  if (reattach) resolveSessionNameOnce(sessionName);
  const input = {
    environmentId: context.environmentId,
    instanceId: context.sandboxId,
    kind: "sandbox",
    scope: "shell"
  };
  const tokenData = await requestGraphQL(context.config, RailwayGenerateShellTokenDocument, { input });
  const jwt = tokenData.generateShellToken;
  let stdout = "";
  let stderr = "";
  let exitCode = null;
  let timedOut = false;
  let settled = false;
  let timer;
  const stdoutDecoder = decoder();
  const stderrDecoder = decoder();
  let resolveResult;
  let rejectResult;
  const done = new Promise((resolve, reject) => {
    resolveResult = resolve;
    rejectResult = reject;
  });
  const settle = (outcome) => {
    if (settled) return;
    settled = true;
    if (!sessionNameResolved) {
      sessionNameResolved = true;
      onNoSessionName(
        new RailwayError(
          "Server did not return a durable session for this exec."
        )
      );
    }
    if (timer) clearTimeout(timer);
    try {
      control.connection?.close();
    } catch {
    }
    if ("error" in outcome) rejectResult(outcome.error);
    else resolveResult(outcome);
  };
  const connection = await connectExecWs({
    config: context.config,
    jwt,
    command,
    ...reattach ? {} : { cwd: options.cwd, env: options.env },
    ...sessionName ? {
      sessionName,
      resumeFromLastRead: options.resumeFromLastRead ?? false
    } : {},
    handlers: {
      onDurableSession: (name) => resolveSessionNameOnce(name),
      onStdout: (bytes) => {
        if (settled) return;
        try {
          const chunk = stdoutDecoder.decode(bytes, { stream: true });
          stdout += chunk;
          options.onStdout?.(chunk);
        } catch (error) {
          settle({ error });
        }
      },
      onStderr: (bytes) => {
        if (settled) return;
        try {
          const chunk = stderrDecoder.decode(bytes, { stream: true });
          stderr += chunk;
          options.onStderr?.(chunk);
        } catch (error) {
          settle({ error });
        }
      },
      onExit: (code) => {
        exitCode = code;
        settle({ exitCode, stdout, stderr, truncated: false, timedOut });
      },
      onClose: () => settle({ exitCode, stdout, stderr, truncated: false, timedOut })
    }
  });
  control.connection = connection;
  if (control.detached) {
    connection.close();
    return done;
  }
  if (control.pendingSignal) connection.signal(control.pendingSignal);
  connection.closeStdin();
  if (options.timeoutSec !== void 0) {
    timer = setTimeout(() => {
      timedOut = true;
      connection.close();
    }, options.timeoutSec * 1e3);
  }
  return done;
}

// src/core/files-ws-client.ts
var BINARY_HEADER_BYTES = 12;
var FRAME_READ_END = 2;
var FRAME_WRITE_CHUNK = 3;
var FRAME_WRITE_END = 4;
var SHELL_SUBPROTOCOL2 = "railway-shell";
var FILES_CHUNK_BYTES = 64 * 1024;
var SEND_HIGH_WATER_BYTES = 128 * 1024;
var SEND_POLL_MS = 10;
var INACTIVITY_TIMEOUT_MS = 3e4;
var WRITE_HEARTBEAT_MS = 1e4;
var FilesRemoteError = class extends Error {
  /**
   * For uploads: whether the failure happened at `write_start` (before any
   * content was consumed, so a retry is safe) or while streaming.
   */
  stage;
  constructor(message, stage) {
    super(message);
    if (stage) this.stage = stage;
  }
};
var sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
function connectFilesWs(args) {
  const { config, jwt, endpoint } = args;
  const WS = resolveWebSocketImpl(config);
  return new Promise((resolve, reject) => {
    let opened = false;
    let closed = false;
    let nextId = 1;
    const pending = /* @__PURE__ */ new Map();
    const socket = new WS(endpoint, [
      SHELL_SUBPROTOCOL2,
      jwt
    ]);
    socket.binaryType = "arraybuffer";
    const failAll = (error) => {
      const ops = [...pending.values()];
      pending.clear();
      for (const op of ops) op.fail(error);
    };
    const closeSocket = () => {
      try {
        socket.close(1e3, "");
      } catch {
      }
    };
    const register = (op) => {
      const id = String(nextId);
      nextId = nextId >= 65535 ? 1 : nextId + 1;
      let timer;
      const disarm = () => {
        if (timer) clearTimeout(timer);
        timer = void 0;
      };
      const arm = () => {
        disarm();
        timer = setTimeout(() => {
          pending.delete(id);
          op.fail(
            new RailwayConnectionError({
              message: `tcp-proxy files operation timed out after ${INACTIVITY_TIMEOUT_MS}ms of inactivity.`
            })
          );
          closeSocket();
        }, INACTIVITY_TIMEOUT_MS);
      };
      pending.set(id, op);
      return { id, arm, disarm };
    };
    const sendJson = (type, id, data) => {
      socket.send(JSON.stringify({ type, id, data }));
    };
    const awaitDrain = async () => {
      const start = Date.now();
      while (!closed && (socket.bufferedAmount ?? 0) > SEND_HIGH_WATER_BYTES) {
        if (Date.now() - start > INACTIVITY_TIMEOUT_MS) {
          closeSocket();
          throw new RailwayConnectionError({
            message: `tcp-proxy files send buffer stalled for ${INACTIVITY_TIMEOUT_MS}ms.`
          });
        }
        await sleep(SEND_POLL_MS);
      }
    };
    const connection = {
      call(type, data) {
        return new Promise((resolveCall, rejectCall) => {
          const { id, arm, disarm } = register({
            onReply: (replyType, replyData) => {
              disarm();
              pending.delete(id);
              if (replyType === "error") {
                rejectCall(new FilesRemoteError(errorMessage(replyData)));
              } else {
                resolveCall(replyData);
              }
            },
            fail: (error) => {
              disarm();
              rejectCall(error);
            }
          });
          sendJson(type, id, data);
          arm();
        });
      },
      read(data, onChunk) {
        return new Promise((resolveRead, rejectRead) => {
          const { id, arm, disarm } = register({
            onReply: (replyType, replyData) => {
              if (replyType !== "error") return;
              disarm();
              pending.delete(id);
              rejectRead(new FilesRemoteError(errorMessage(replyData)));
            },
            onBinary: (frameType, payload) => {
              arm();
              try {
                if (payload.length > 0) onChunk(payload);
              } catch (error) {
                disarm();
                pending.delete(id);
                rejectRead(error);
                closeSocket();
                return;
              }
              if (frameType === FRAME_READ_END) {
                disarm();
                pending.delete(id);
                resolveRead();
              }
            },
            fail: (error) => {
              disarm();
              rejectRead(error);
            }
          });
          sendJson("read", id, data);
          arm();
        });
      },
      async write(start, chunks) {
        let settled = false;
        let notifySettled;
        const settledSignal = new Promise((resolve2) => {
          notifySettled = () => resolve2("settled");
        });
        let settle;
        const done = new Promise((resolveWrite, rejectWrite) => {
          settle = {
            resolve: () => {
              settled = true;
              notifySettled();
              resolveWrite();
            },
            reject: (error) => {
              settled = true;
              notifySettled();
              rejectWrite(error);
            }
          };
        });
        done.catch(() => {
        });
        let ready;
        const writeReady = new Promise((resolveReady, rejectReady) => {
          ready = { resolve: resolveReady, reject: rejectReady };
        });
        let awaitingReady = true;
        const { id, arm, disarm } = register({
          onReply: (replyType, replyData) => {
            if (replyType === "error") {
              disarm();
              pending.delete(id);
              const error = new FilesRemoteError(
                errorMessage(replyData),
                awaitingReady ? "start" : "stream"
              );
              if (awaitingReady) ready.reject(error);
              settle.reject(error);
              return;
            }
            if (awaitingReady && replyType === "write_ready") {
              awaitingReady = false;
              disarm();
              ready.resolve();
              return;
            }
            if (replyType === "ok") {
              disarm();
              pending.delete(id);
              if (awaitingReady) {
                const error = new FilesRemoteError(
                  "protocol error: ok before write_ready",
                  "start"
                );
                ready.reject(error);
                settle.reject(error);
                return;
              }
              settle.resolve();
            }
          },
          fail: (error) => {
            disarm();
            if (awaitingReady) ready.reject(error);
            settle.reject(error);
          }
        });
        sendJson("write_start", id, start);
        arm();
        await writeReady;
        const reqId = Number(id);
        let seq = 0;
        const sendContent = (payload, last) => {
          socket.send(
            binaryFrame(
              reqId,
              last ? FRAME_WRITE_END : FRAME_WRITE_CHUNK,
              seq++,
              payload
            )
          );
        };
        const heartbeat = setInterval(() => {
          if (settled || closed) return;
          try {
            sendContent(new Uint8Array(0), false);
          } catch {
          }
        }, WRITE_HEARTBEAT_MS);
        let completed;
        try {
          completed = await streamUpload({
            iterator: rechunk(chunks)[Symbol.asyncIterator](),
            settledSignal,
            isSettled: () => settled,
            awaitDrain,
            send: sendContent
          });
        } catch (error) {
          pending.delete(id);
          disarm();
          closeSocket();
          throw error;
        } finally {
          clearInterval(heartbeat);
        }
        if (completed) arm();
        return done;
      },
      close: closeSocket
    };
    socket.onopen = () => {
      opened = true;
      if (socket.bufferedAmount === void 0) {
        config.log(
          "files: webSocketImpl exposes no bufferedAmount; upload pacing is disabled"
        );
      }
      resolve(connection);
    };
    socket.onmessage = (event) => {
      const { data } = event;
      if (data instanceof ArrayBuffer) {
        handleBinaryFrame2(data);
      } else if (typeof data === "string") {
        handleTextFrame2(data);
      }
    };
    socket.onclose = (event) => {
      closed = true;
      if (!opened) {
        reject(
          new RailwayConnectionError({
            message: `tcp-proxy files WebSocket closed before opening (code ${event.code}${event.reason ? `: ${event.reason}` : ""}).`,
            closeCode: event.code
          })
        );
        return;
      }
      failAll(
        new RailwayConnectionError({
          message: `tcp-proxy files WebSocket closed (code ${event.code}${event.reason ? `: ${event.reason}` : ""}).`,
          closeCode: event.code
        })
      );
    };
    socket.onerror = (event) => {
      if (opened) return;
      reject(
        new RailwayConnectionError({
          message: "tcp-proxy files WebSocket connection failed.",
          cause: event
        })
      );
    };
    function handleTextFrame2(text) {
      let frame;
      try {
        frame = JSON.parse(text);
      } catch {
        return;
      }
      if (!frame.type || !frame.id) return;
      pending.get(frame.id)?.onReply(frame.type, frame.data);
    }
    function handleBinaryFrame2(buffer) {
      if (buffer.byteLength < BINARY_HEADER_BYTES) return;
      const view = new DataView(buffer);
      const reqId = view.getUint16(0);
      const frameType = view.getUint16(2);
      const payloadLen = view.getUint32(8);
      const available = buffer.byteLength - BINARY_HEADER_BYTES;
      const payload = new Uint8Array(
        buffer,
        BINARY_HEADER_BYTES,
        Math.min(payloadLen, available)
      );
      pending.get(String(reqId))?.onBinary?.(frameType, payload);
    }
  });
}
function errorMessage(data) {
  const message = data?.message;
  return message || "unknown error";
}
function binaryFrame(reqId, frameType, seq, payload) {
  const frame = new Uint8Array(BINARY_HEADER_BYTES + payload.length);
  const view = new DataView(frame.buffer);
  view.setUint16(0, reqId);
  view.setUint16(2, frameType);
  view.setUint32(4, seq);
  view.setUint32(8, payload.length);
  frame.set(payload, BINARY_HEADER_BYTES);
  return frame;
}
async function* rechunk(chunks) {
  for await (const chunk of chunks) {
    for (let offset = 0; offset < chunk.length; offset += FILES_CHUNK_BYTES) {
      yield chunk.subarray(
        offset,
        Math.min(offset + FILES_CHUNK_BYTES, chunk.length)
      );
    }
  }
}
async function streamUpload(args) {
  const { isSettled, awaitDrain, send } = args;
  let held;
  for (; ; ) {
    const next = await nextChunkOrSettled(args);
    if (next === null) return false;
    if (next.done) break;
    if (held) {
      await awaitDrain();
      if (isSettled()) return false;
      send(held, false);
    }
    held = next.value;
  }
  if (isSettled()) return false;
  await awaitDrain();
  if (isSettled()) return false;
  send(held ?? new Uint8Array(0), true);
  return true;
}
async function nextChunkOrSettled(args) {
  const { iterator, settledSignal, isSettled } = args;
  const nextChunk = iterator.next();
  void Promise.resolve(nextChunk).catch(() => {
  });
  const next = await Promise.race([nextChunk, settledSignal]);
  if (next !== "settled" && !isSettled()) return next;
  void Promise.resolve(iterator.return?.(void 0)).catch(() => {
  });
  return null;
}

// src/sandbox/files-transfer.ts
var READ_SEGMENT_TARGET_MS = 5e3;
var READ_SEGMENT_MIN_BYTES = 256 * 1024;
var READ_SEGMENT_MAX_BYTES = 32 * 1024 * 1024;
var READ_SEGMENT_INITIAL_BYTES = 2 * 1024 * 1024;
var TRANSFER_MAX_RETRIES = 3;
var TRANSFER_RETRY_DELAY_MS = 300;
var sleep2 = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
function isNotFound(message) {
  return /file does not exist|no such file|not found/i.test(message);
}
function isTransientTransfer(error) {
  return error instanceof RailwayConnectionError || error instanceof FilesRemoteError && /connection lost/i.test(error.message);
}
function canRetryWrite(error, source, retries) {
  return isTransientTransfer(error) && (source.replayable || !source.consumed()) && retries <= TRANSFER_MAX_RETRIES;
}
async function retryDelay() {
  await sleep2(TRANSFER_RETRY_DELAY_MS);
}
function startPull(args) {
  const { connect, log, path: path3, options, onChunk } = args;
  let connection;
  let aborted = false;
  const done = (async () => {
    let range;
    let retries = 0;
    for (; ; ) {
      if (aborted) return;
      try {
        connection = await connect();
        if (aborted) return;
        range ??= await resolveRange(connection, path3, options);
        await pullSegments({
          connection,
          path: path3,
          range,
          isAborted: () => aborted,
          onChunk: (bytes) => {
            retries = 0;
            onChunk(bytes);
          },
          ...args.waitForCapacity && { waitForCapacity: args.waitForCapacity }
        });
        return;
      } catch (error) {
        if (aborted) return;
        if (!canResumePull(error, ++retries)) throw error;
        log(`files read ${path3} resuming after connection loss`);
        await retryDelay();
      } finally {
        connection?.close();
        connection = void 0;
      }
    }
  })();
  return {
    done,
    abort: () => {
      aborted = true;
      connection?.close();
    }
  };
}
function canResumePull(error, retries) {
  return isTransientTransfer(error) && retries <= TRANSFER_MAX_RETRIES;
}
async function resolveRange(connection, path3, options) {
  if (options.fromEnd) {
    const info = await connection.call("stat", { path: path3 });
    return { position: Math.max(0, info.size - options.length), end: info.size };
  }
  const start = options.offset ?? 0;
  return {
    position: start,
    end: options.length !== void 0 ? start + options.length : Infinity
  };
}
async function pullSegments(args) {
  const { connection, path: path3, range, isAborted, onChunk } = args;
  let segmentBytes = READ_SEGMENT_INITIAL_BYTES;
  while (range.position < range.end) {
    await args.waitForCapacity?.();
    if (isAborted()) return;
    const length = Math.min(segmentBytes, range.end - range.position);
    const started = Date.now();
    let received = 0;
    await connection.read({ path: path3, offset: range.position, length }, (bytes) => {
      received += bytes.length;
      range.position += bytes.length;
      onChunk(bytes);
    });
    if (received < length) return;
    segmentBytes = nextSegmentSize(segmentBytes, Date.now() - started);
  }
}
function nextSegmentSize(current, elapsedMs) {
  const scaled = Math.round(
    current * READ_SEGMENT_TARGET_MS / Math.max(elapsedMs, 1)
  );
  return Math.min(
    READ_SEGMENT_MAX_BYTES,
    current * 4,
    Math.max(READ_SEGMENT_MIN_BYTES, scaled)
  );
}
function uploadSource(data) {
  const source = normalizeSource2(data);
  let consumed = false;
  return {
    ...source,
    chunks: () => markOnFirstPull(source.chunks(), () => {
      consumed = true;
    }),
    consumed: () => consumed
  };
}
function markOnFirstPull(iterable, mark) {
  return {
    [Symbol.asyncIterator]() {
      const inner = iterable[Symbol.asyncIterator]();
      const iterator = {
        next: (...args) => {
          mark();
          return inner.next(...args);
        }
      };
      if (inner.return) iterator.return = inner.return.bind(inner);
      if (inner.throw) iterator.throw = inner.throw.bind(inner);
      return iterator;
    }
  };
}
function normalizeSource2(data) {
  if (typeof data === "string") {
    const bytes = new TextEncoder().encode(data);
    return { chunks: () => single(bytes), size: bytes.length, replayable: true };
  }
  if (data instanceof Uint8Array) {
    return { chunks: () => single(data), size: data.length, replayable: true };
  }
  if (data instanceof ArrayBuffer) {
    const bytes = new Uint8Array(data);
    return { chunks: () => single(bytes), size: bytes.length, replayable: true };
  }
  const streaming = streamingSource(data);
  if (streaming) return streaming;
  throw new TypeError(
    "Unsupported write data: pass a string, Uint8Array, ArrayBuffer, Blob, ReadableStream, AsyncIterable of Uint8Array, or a function returning a stream or iterable."
  );
}
function streamingSource(data) {
  if (typeof Blob !== "undefined" && data instanceof Blob) {
    return {
      chunks: () => streamChunks(data.stream()),
      size: data.size,
      replayable: true
    };
  }
  if (typeof ReadableStream !== "undefined" && data instanceof ReadableStream) {
    const chunks = streamChunks(data);
    return { chunks: () => chunks, size: 0, replayable: false };
  }
  if (isAsyncIterable(data)) {
    return { chunks: () => data, size: 0, replayable: false };
  }
  if (typeof data === "function") {
    return { chunks: () => producedChunks(data()), size: 0, replayable: true };
  }
  return void 0;
}
async function* single(bytes) {
  if (bytes.length > 0) yield bytes;
}
async function* streamChunks(stream) {
  const reader = stream.getReader();
  try {
    for (; ; ) {
      const { done, value } = await reader.read();
      if (done) return;
      if (value && value.length > 0) yield value;
    }
  } finally {
    try {
      await reader.cancel();
    } catch {
    }
    reader.releaseLock();
  }
}
function producedChunks(produced) {
  return produced instanceof ReadableStream ? streamChunks(produced) : produced;
}
function isAsyncIterable(value) {
  return typeof value === "object" && value !== null && Symbol.asyncIterator in value && typeof value[Symbol.asyncIterator] === "function";
}

// src/sandbox/files.ts
var constructFiles;
var SandboxFiles = class _SandboxFiles {
  #context;
  /** Constructed by `Sandbox#files`; not constructible from outside the SDK. */
  constructor(context) {
    this.#context = context;
  }
  async read(path3, options = {}) {
    const target = validatePath(path3);
    validateReadOptions(options);
    this.#log(`files read ${target} format=${options.format ?? "text"}`);
    if (options.format === "stream") {
      return this.#readAsStream(target, options);
    }
    const chunks = [];
    try {
      await this.#startPull(target, options, (bytes2) => chunks.push(bytes2)).done;
    } catch (error) {
      throw this.#wrapError("read", target, error);
    }
    const bytes = concat(chunks);
    this.#log(`files read ${target} done (${bytes.length} bytes)`);
    return options.format === "bytes" ? bytes : new TextDecoder().decode(bytes);
  }
  /**
   * Write a file, creating it (and truncating any existing content). Missing
   * parent directories are created automatically. Strings, bytes, blobs, and
   * factory sources upload with retry if the session drops mid-transfer;
   * bare streams and async iterables upload unbuffered but are one-shot, so
   * a session dropped mid-stream surfaces as `RailwayConnectionError` (and a
   * partial file may remain at `path`).
   *
   * Files are created `0644`; pass `mode` to set permissions (applied right
   * after the upload completes).
   */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  async write(path3, data, options = {}) {
    const target = validatePath(path3);
    const mode = validateMode(options.mode);
    const source = uploadSource(data);
    const start = { path: target, mode: mode ?? 0, size: source.size };
    this.#log(
      `files write ${target} size=${source.size > 0 ? source.size : "stream"}`
    );
    let retries = 0;
    for (; ; ) {
      try {
        await this.#writeOnce(target, start, source);
        break;
      } catch (error) {
        if (!canRetryWrite(error, source, ++retries)) {
          throw this.#wrapError("write", target, error);
        }
        this.#log(`files write ${target} retrying after connection loss`);
        await retryDelay();
      }
    }
    if (mode !== void 0) await this.#applyMode(target, mode);
    this.#log(`files write ${target} done`);
  }
  /** List a directory (entries carry name, size, mode, and modify time). */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  async list(path3) {
    const target = validatePath(path3);
    const data = await this.#call("ls", target, { path: target });
    return data?.entries ?? [];
  }
  /** Stat a path (follows symlinks). Throws `SandboxFileNotFoundError` if absent. */
  async stat(path3) {
    const target = validatePath(path3);
    const data = await this.#call("stat", target, { path: target });
    return data;
  }
  /** Whether a path exists (file or directory). */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  async exists(path3) {
    try {
      await this.stat(path3);
      return true;
    } catch (error) {
      if (error instanceof SandboxFileNotFoundError) return false;
      throw error;
    }
  }
  /** Create a directory, including missing parents (like `mkdir -p`). */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  async mkdir(path3) {
    const target = validatePath(path3);
    await this.#call("mkdir", target, { path: target });
  }
  /**
   * Remove a file or empty directory. For recursive deletes use
   * `sandbox.exec("rm -rf ...")`.
   */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  async remove(path3) {
    const target = validatePath(path3);
    await this.#call("rm", target, { path: target });
  }
  /** Rename (move) a file or directory within the sandbox. */
  // Public SDK API; consumed by library users, not in-repo code.
  // fallow-ignore-next-line unused-class-member
  async rename(oldPath, newPath) {
    const from = validatePath(oldPath);
    const to = validatePath(newPath);
    await this.#call("rename", from, { old: from, new: to });
  }
  /** One upload attempt on its own connection, creating missing parents. */
  async #writeOnce(target, start, source) {
    const connection = await this.#connect("files:read files:write");
    try {
      try {
        await connection.write(start, source.chunks());
      } catch (error) {
        const parent = parentDir(target);
        if (!isMissingParent(error) || !parent) throw error;
        await connection.call("mkdir", { path: parent });
        await connection.write(start, source.chunks());
      }
    } finally {
      connection.close();
    }
  }
  /** Sets `path`'s permissions with a chmod run through the exec primitive. */
  async #applyMode(path3, mode) {
    const octal = mode.toString(8).padStart(3, "0");
    const result = await startExec(
      this.#context,
      `chmod ${octal} -- ${shellQuote(path3)}`,
      { timeoutSec: 30 }
    );
    if (result.exitCode !== 0) {
      throw new SandboxFilesError({
        operation: "write",
        path: path3,
        message: result.stderr.trim() || `chmod ${octal} exited with code ${result.exitCode}`
      });
    }
  }
  /** One-shot request/reply op on its own connection, with error mapping. */
  async #call(type, path3, payload) {
    this.#log(`files ${type} ${path3}`);
    const scope = type === "ls" || type === "stat" ? "files:read" : "files:read files:write";
    const connection = await this.#connect(scope);
    try {
      return await connection.call(type, payload);
    } catch (error) {
      throw this.#wrapError(type, path3, error);
    } finally {
      connection.close();
    }
  }
  async #readAsStream(path3, options) {
    let controller;
    let pull;
    let notifyPull;
    const stream = new ReadableStream({
      start: (c) => {
        controller = c;
      },
      pull: () => notifyPull?.(),
      cancel: () => {
        pull.abort();
        notifyPull?.();
      }
    });
    const waitForCapacity = () => {
      if ((controller.desiredSize ?? 0) > 0) return Promise.resolve();
      return new Promise((resolve) => {
        notifyPull = () => {
          notifyPull = void 0;
          resolve();
        };
      });
    };
    pull = this.#startPull(
      path3,
      options,
      (bytes) => controller.enqueue(bytes),
      waitForCapacity
    );
    pull.done.then(() => {
      try {
        controller.close();
      } catch {
      }
    }).catch((error) => {
      try {
        controller.error(this.#wrapError("read", path3, error));
      } catch {
      }
    });
    return stream;
  }
  #startPull(path3, options, onChunk, waitForCapacity) {
    return startPull({
      connect: () => this.#connect("files:read"),
      log: (message) => this.#log(message),
      path: path3,
      options,
      onChunk,
      ...waitForCapacity && { waitForCapacity }
    });
  }
  /** Mints a files-scoped token and opens a `/ws/files` session with it. */
  async #connect(scope) {
    const { config, environmentId, sandboxId } = this.#context;
    const input = {
      environmentId,
      instanceId: sandboxId,
      kind: "sandbox",
      scope
    };
    const tokenData = await requestGraphQL(config, RailwayGenerateShellTokenDocument, { input });
    return connectFilesWs({
      config,
      jwt: tokenData.generateShellToken,
      endpoint: deriveFilesWsEndpoint(config.tcpProxyWsEndpoint)
    });
  }
  #wrapError(operation, path3, error) {
    if (error instanceof FilesRemoteError) {
      const args = { operation, path: path3, message: error.message };
      return isNotFound(error.message) ? new SandboxFileNotFoundError(args) : new SandboxFilesError(args);
    }
    return error;
  }
  #log(message) {
    this.#context.config.log(`${message} sandbox=${this.#context.sandboxId}`);
  }
  static {
    constructFiles = (context) => new _SandboxFiles(context);
  }
};
function createSandboxFiles(context) {
  return constructFiles(context);
}
function validatePath(path3) {
  if (typeof path3 !== "string" || path3.trim() === "") {
    throw new TypeError("Path must be a non-empty string.");
  }
  return path3;
}
function validateReadOptions(options) {
  if (options.fromEnd && options.offset !== void 0) {
    throw new TypeError("`offset` and `fromEnd` are mutually exclusive.");
  }
  if (options.fromEnd && options.length === void 0) {
    throw new TypeError("`fromEnd` requires `length` (the tail size to read).");
  }
  requireIntegerAtLeast("offset", options.offset, 0);
  requireIntegerAtLeast("length", options.length, 1);
}
function requireIntegerAtLeast(name, value, min) {
  if (value === void 0) return;
  if (!Number.isInteger(value) || value < min) {
    throw new TypeError(`\`${name}\` must be an integer >= ${min}.`);
  }
}
function validateMode(mode) {
  if (mode === void 0) return void 0;
  if (!Number.isInteger(mode) || mode < 0 || mode > 4095) {
    throw new TypeError("`mode` must be POSIX permission bits (e.g. 0o755).");
  }
  return mode;
}
function shellQuote(path3) {
  return `'${path3.replaceAll("'", `'\\''`)}'`;
}
function isMissingParent(error) {
  return error instanceof FilesRemoteError && error.stage === "start" && isNotFound(error.message);
}
function parentDir(path3) {
  const trimmed = path3.replace(/\/+$/, "");
  const index = trimmed.lastIndexOf("/");
  return index > 0 ? trimmed.slice(0, index) : void 0;
}
function concat(chunks) {
  if (chunks.length === 1) return chunks[0];
  const total = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
  const merged = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    merged.set(chunk, offset);
    offset += chunk.length;
  }
  return merged;
}

// src/sandbox/engine.ts
var READINESS_TIMEOUT_MS = 5 * 6e4;
var POLL_INITIAL_DELAY_MS = 500;
var POLL_MAX_DELAY_MS = 5e3;
var sleep3 = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
var SandboxEngine = class {
  #config;
  constructor(config) {
    this.#config = config;
  }
  get environmentId() {
    return this.#config.environmentId;
  }
  async create(options = {}, template2) {
    const input = {
      environmentId: this.#config.environmentId
    };
    if (template2 !== void 0) {
      input.template = toTemplateInput(template2);
    }
    return this.#createAndWait(input, options);
  }
  async fork(sourceSandboxId, options = {}) {
    return this.#createAndWait(
      { environmentId: this.#config.environmentId, sourceSandboxId },
      options
    );
  }
  /** Applies the shared creation knobs, runs the create mutation, and waits for RUNNING. */
  async #createAndWait(input, options) {
    if (options.idleTimeoutMinutes !== void 0) {
      input.idleTimeoutMinutes = options.idleTimeoutMinutes;
    }
    if (options.networkIsolation !== void 0) {
      input.networkIsolation = options.networkIsolation;
    }
    if (options.env !== void 0) {
      input.variables = options.env;
    }
    this.#config.log(creationLine(input));
    const data = await requestGraphQL(this.#config, RailwaySandboxCreateDocument, { input });
    this.#config.log(
      `created sandbox ${data.sandboxCreate.id} status=${data.sandboxCreate.status}`
    );
    return this.#waitForRunning(data.sandboxCreate);
  }
  async buildTemplate(template2) {
    const variables = {
      environmentId: this.#config.environmentId,
      input: toTemplateInput(template2)
    };
    const data = await requestGraphQL(this.#config, RailwaySandboxTemplateBuildDocument, variables);
    return data.sandboxTemplateBuild;
  }
  async getTemplateBuild(id) {
    const variables = {
      id,
      environmentId: this.#config.environmentId
    };
    const data = await requestGraphQL(this.#config, RailwaySandboxTemplateBuildStatusDocument, variables);
    return data.sandboxTemplateBuild;
  }
  async buildTemplateUntilReady(template2) {
    const varCount = template2.variables ? Object.keys(template2.variables).length : 0;
    this.#config.log(
      `build template (${template2.instructions.length} steps, vars=${varCount})`
    );
    return this.#waitForBuildReady(await this.buildTemplate(template2));
  }
  /**
   * Capture a running sandbox's disk into a named checkpoint. Synchronous on
   * the backend: the checkpoint is bootable when this resolves.
   */
  async checkpoint(sandboxId, name) {
    this.#config.log(`checkpoint sandbox ${sandboxId} -> "${name}"`);
    const variables = {
      environmentId: this.#config.environmentId,
      name,
      sandboxId
    };
    const data = await requestGraphQL(this.#config, RailwaySandboxCheckpointCreateDocument, variables);
    this.#config.log(`checkpoint ${data.sandboxCheckpointCreate.id} ready`);
    return data.sandboxCheckpointCreate;
  }
  async listCheckpoints() {
    const data = await requestGraphQL(this.#config, RailwaySandboxCheckpointsDocument, {
      environmentId: this.#config.environmentId
    });
    return data.sandboxCheckpoints;
  }
  async renameCheckpoint(id, name) {
    this.#config.log(`rename checkpoint ${id} -> "${name}"`);
    const data = await requestGraphQL(this.#config, RailwaySandboxCheckpointRenameDocument, {
      environmentId: this.#config.environmentId,
      id,
      name
    });
    return data.sandboxCheckpointRename;
  }
  async deleteCheckpoint(id) {
    this.#config.log(`delete checkpoint ${id}`);
    await requestGraphQL(this.#config, RailwaySandboxCheckpointDeleteDocument, {
      environmentId: this.#config.environmentId,
      id
    });
  }
  /**
   * Resolves once the build is READY (its checkpoint exists). PENDING (no
   * workflow yet) and BUILDING keep polling; only FAILED is terminal.
   */
  async #waitForBuildReady(build) {
    if (build.status === "READY") {
      this.#config.log(`template build ${build.id} ready (cached)`);
      return build;
    }
    if (build.status === "FAILED") {
      throw new SandboxTemplateBuildError({
        templateId: build.id,
        environmentId: this.environmentId
      });
    }
    return this.#pollUntilReady({
      poll: () => this.getTemplateBuild(build.id),
      isReady: (build2) => build2.status === "READY",
      isTerminal: (build2) => build2.status === "FAILED",
      describe: (build2) => `template build ${build2.id} status=${build2.status}`,
      onTerminal: (build2) => {
        throw new SandboxTemplateBuildError({
          templateId: build2.id,
          environmentId: this.environmentId
        });
      },
      onTimeout: (build2) => {
        throw new SandboxTimeoutError({
          resource: "template",
          id: build2.id,
          lastStatus: build2.status,
          timeoutMs: READINESS_TIMEOUT_MS
        });
      }
    });
  }
  async #waitForRunning(created) {
    if (created.status === "RUNNING") return created;
    if (isSandboxTerminal(created.status)) {
      throw new SandboxFailedError({ id: created.id, status: created.status });
    }
    return this.#pollUntilReady({
      poll: () => this.#getOrThrow(created.id),
      isReady: (info) => info.status === "RUNNING",
      isTerminal: (info) => isSandboxTerminal(info.status),
      describe: (info) => `sandbox ${info.id} status=${info.status}`,
      onTerminal: (info) => {
        throw new SandboxFailedError({ id: info.id, status: info.status });
      },
      onTimeout: (info) => {
        throw new SandboxTimeoutError({
          resource: "sandbox",
          id: info.id,
          lastStatus: info.status,
          timeoutMs: READINESS_TIMEOUT_MS
        });
      }
    });
  }
  async #getOrThrow(id) {
    const info = await this.get(id);
    if (!info) {
      throw new SandboxNotFoundError({ id, environmentId: this.environmentId });
    }
    return info;
  }
  async #pollUntilReady(args) {
    const start = Date.now();
    const describe = args.describe;
    let delay = POLL_INITIAL_DELAY_MS;
    let last;
    do {
      await sleep3(delay);
      last = await args.poll();
      const elapsedSec = ((Date.now() - start) / 1e3).toFixed(1);
      if (args.isReady(last)) {
        if (describe) this.#config.log(`${describe(last)} ready after ${elapsedSec}s`);
        return last;
      }
      if (args.isTerminal(last)) {
        if (describe) this.#config.log(`${describe(last)} hit terminal state`);
        return args.onTerminal(last);
      }
      delay = Math.min(delay * 2, POLL_MAX_DELAY_MS);
      if (describe) {
        this.#config.log(
          `waiting on ${describe(last)}, retry in ${delay}ms (elapsed ${elapsedSec}s)`
        );
      }
    } while (Date.now() - start < READINESS_TIMEOUT_MS);
    if (describe) {
      const elapsedSec = ((Date.now() - start) / 1e3).toFixed(1);
      this.#config.log(`${describe(last)} timed out after ${elapsedSec}s`);
    }
    return args.onTimeout(last);
  }
  exec(id, target, options = {}) {
    const context = {
      config: this.#config,
      environmentId: this.#config.environmentId,
      sandboxId: id
    };
    return startExec(context, target, options);
  }
  files(id) {
    return createSandboxFiles({
      config: this.#config,
      environmentId: this.#config.environmentId,
      sandboxId: id
    });
  }
  async destroy(id) {
    const variables = {
      id,
      environmentId: this.#config.environmentId
    };
    this.#config.log(`destroy sandbox ${id}`);
    await requestGraphQL(this.#config, RailwaySandboxDestroyDocument, variables);
  }
  async get(id) {
    const variables = {
      id,
      environmentId: this.#config.environmentId
    };
    const data = await requestGraphQL(this.#config, RailwaySandboxDocument, variables);
    const info = data.sandbox ?? null;
    if (!info) {
      this.#config.log(`sandbox ${id} not found (env=${this.environmentId})`);
    }
    return info;
  }
  async list(options = {}) {
    const variables = {
      environmentId: this.#config.environmentId
    };
    if (options.first !== void 0) variables.first = options.first;
    if (options.after !== void 0) variables.after = options.after;
    const data = await requestGraphQL(this.#config, RailwaySandboxesDocument, variables);
    return data.sandboxes.edges.map((edge) => edge.node);
  }
};
function isSandboxTerminal(status) {
  return status === "FAILED" || status === "DESTROYED" || status === "DESTROYING";
}
function toTemplateInput(template2) {
  if ("name" in template2) {
    return { name: template2.name };
  }
  return {
    instructions: [...template2.instructions],
    ...template2.variables && { variables: template2.variables }
  };
}
function creationKind(input) {
  if (input.sourceSandboxId) return "fork";
  if (input.template?.name) return "create-from-checkpoint";
  if (input.template) return "create-from-template";
  return "create";
}
function creationLine(input) {
  const kind = creationKind(input);
  const envKeys = input.variables ? Object.keys(input.variables) : [];
  const parts = [
    kind,
    `env=${input.environmentId}`,
    `idleTimeout=${input.idleTimeoutMinutes ?? "none"}`,
    `network=${input.networkIsolation ?? "default"}`,
    `envKeys=[${envKeys.join(",")}]`
  ];
  if (input.sourceSandboxId) parts.push(`source=${input.sourceSandboxId}`);
  if (input.template?.name) parts.push(`checkpoint="${input.template.name}"`);
  return parts.join(" ");
}
function engineFromOptions(options = {}) {
  const base = normalizeRailwayClientConfig(options);
  const environmentId = resolveEnvironmentId(options.environmentId);
  return new SandboxEngine({ ...base, environmentId });
}

// src/sandbox/template.ts
var EMPTY_STATE = { instructions: [], env: [] };
function createSandboxTemplate() {
  return new SandboxTemplateRecipe(EMPTY_STATE);
}
function isSandboxTemplate(value) {
  return value instanceof SandboxTemplateRecipe;
}
function compileSandboxTemplate(template2) {
  if (!(template2 instanceof SandboxTemplateRecipe)) {
    throw new TypeError("Expected a SandboxTemplate returned by Sandbox.template().");
  }
  return template2.compile();
}
var SandboxTemplateRecipe = class _SandboxTemplateRecipe {
  #state;
  constructor(state) {
    this.#state = state;
  }
  run(command) {
    return this.#append(command);
  }
  withPackages(...packages) {
    if (packages.length === 0) return this;
    return this.#append(
      `apt-get update && apt-get install -y --no-install-recommends ${packages.join(" ")}`
    );
  }
  withEnv(vars) {
    let env = this.#state.env;
    for (const [key, value] of Object.entries(vars)) {
      env = upsertEnv(env, key, value);
    }
    return new _SandboxTemplateRecipe({ ...this.#state, env });
  }
  workdir(dir) {
    return new _SandboxTemplateRecipe({ ...this.#state, workdir: dir });
  }
  async build(options = {}) {
    const compiled = this.compile();
    if (compiled.instructions.length > 0) {
      await engineFromOptions(options).buildTemplateUntilReady(compiled);
    }
    return this;
  }
  compile() {
    const instructions = [...this.#state.instructions];
    if (this.#state.env.length === 0) return { instructions };
    return { instructions, variables: Object.fromEntries(this.#state.env) };
  }
  #append(command) {
    const instruction = compileInstruction(command, this.#state.workdir);
    return new _SandboxTemplateRecipe({
      ...this.#state,
      instructions: [...this.#state.instructions, instruction]
    });
  }
};
function compileInstruction(command, workdir) {
  if (workdir === void 0) return command;
  const quoted = shellQuote2(workdir);
  return `mkdir -p ${quoted} && cd ${quoted} && ${command}`;
}
function upsertEnv(env, key, value) {
  const index = env.findIndex(([k]) => k === key);
  if (index === -1) return [...env, [key, value]];
  const next = env.slice();
  next[index] = [key, value];
  return next;
}
function shellQuote2(value) {
  return `'${value.replaceAll("'", "'\\''")}'`;
}

// src/sandbox/sandbox.ts
var Sandbox = class _Sandbox {
  #engine;
  #info;
  #files;
  constructor(engine, info) {
    this.#engine = engine;
    this.#info = info;
  }
  get id() {
    return this.#info.id;
  }
  get status() {
    return this.#info.status;
  }
  /** Network access mode: `ISOLATED` (NAT egress only) or `PRIVATE` (joins the environment private network). */
  get networkIsolation() {
    return this.#info.networkIsolation;
  }
  get environmentId() {
    return this.#info.environmentId;
  }
  get region() {
    return this.#info.region;
  }
  get idleTimeoutMinutes() {
    return this.#info.idleTimeoutMinutes ?? null;
  }
  get createdAt() {
    return this.#info.createdAt;
  }
  /** Return a new immutable sandbox template. */
  static template() {
    return createSandboxTemplate();
  }
  static async create(sourceOrOptions = {}, maybeOptions = {}) {
    if (typeof sourceOrOptions === "string") {
      const name = sourceOrOptions.trim();
      if (name === "") {
        throw new TypeError("Checkpoint name must be a non-empty string.");
      }
      const engine2 = engineFromOptions(maybeOptions);
      const info2 = await engine2.create(maybeOptions, { name });
      return new _Sandbox(engine2, info2);
    }
    if (sourceOrOptions instanceof _Sandbox) {
      return sourceOrOptions.fork(maybeOptions);
    }
    if (isSandboxTemplate(sourceOrOptions)) {
      const engine2 = engineFromOptions(maybeOptions);
      const compiled = compileSandboxTemplate(sourceOrOptions);
      if (compiled.instructions.length === 0) {
        return new _Sandbox(engine2, await engine2.create(maybeOptions));
      }
      await engine2.buildTemplateUntilReady(compiled);
      const info2 = await engine2.create(maybeOptions, compiled);
      return new _Sandbox(engine2, info2);
    }
    const engine = engineFromOptions(sourceOrOptions);
    const info = await engine.create(sourceOrOptions);
    return new _Sandbox(engine, info);
  }
  static async connect(id, options = {}) {
    const engine = engineFromOptions(options);
    const info = await engine.get(id);
    if (!info) {
      throw new SandboxNotFoundError({ id, environmentId: engine.environmentId });
    }
    return new _Sandbox(engine, info);
  }
  static async list(options = {}) {
    const engine = engineFromOptions(options);
    return engine.list(options);
  }
  /** List the environment's named checkpoints (newest first). */
  static async checkpoints(options = {}) {
    return engineFromOptions(options).listCheckpoints();
  }
  /** Rename a checkpoint by id; new sandboxes boot from it via the new name. */
  static async renameCheckpoint(id, name, options = {}) {
    const trimmed = name.trim();
    if (trimmed === "") {
      throw new TypeError("Checkpoint name must be a non-empty string.");
    }
    return engineFromOptions(options).renameCheckpoint(id, trimmed);
  }
  /** Delete a checkpoint by id (find it by `key` via `Sandbox.checkpoints`). */
  static async deleteCheckpoint(id, options = {}) {
    await engineFromOptions(options).deleteCheckpoint(id);
  }
  exec(target, options = {}) {
    return this.#engine.exec(this.id, target, options);
  }
  /**
   * File operations on this sandbox's filesystem: `read`, `write`, `list`,
   * `stat`, `exists`, `mkdir`, `remove`, `rename`. Reads and writes stream,
   * so files larger than memory transfer safely.
   */
  get files() {
    this.#files ??= this.#engine.files(this.id);
    return this.#files;
  }
  /**
   * Fork this running sandbox into a new, independent one. Clones the
   * filesystem (a fresh boot, not live processes) into the same environment.
   */
  async fork(options) {
    const info = await this.#engine.fork(this.id, options);
    return new _Sandbox(this.#engine, info);
  }
  /**
   * Capture this sandbox's current disk into a reusable named checkpoint,
   * bootable as soon as this resolves. The sandbox must be running and the
   * name unused. Boot new sandboxes from it with `Sandbox.create(name)`.
   */
  async checkpoint(name) {
    const trimmed = name.trim();
    if (trimmed === "") {
      throw new TypeError("Checkpoint name must be a non-empty string.");
    }
    return this.#engine.checkpoint(this.id, trimmed);
  }
  async destroy() {
    await this.#engine.destroy(this.id);
  }
  /** Re-reads the sandbox to refresh `status` and other fields in place. */
  async refresh() {
    const info = await this.#engine.get(this.id);
    if (!info) {
      throw new SandboxNotFoundError({
        id: this.id,
        environmentId: this.environmentId
      });
    }
    this.#info = info;
    return this;
  }
  async [Symbol.asyncDispose]() {
    await this.destroy();
  }
  toJSON() {
    return { ...this.#info };
  }
};

exports.DEFAULT_RAILWAY_GRAPHQL_ENDPOINT = DEFAULT_RAILWAY_GRAPHQL_ENDPOINT;
exports.ExecHandle = ExecHandle;
exports.IacClient = IacClient;
exports.RAILWAY_GRAPH_VERSION = RAILWAY_GRAPH_VERSION;
exports.RailwayAuthError = RailwayAuthError;
exports.RailwayConnectionError = RailwayConnectionError;
exports.RailwayError = RailwayError;
exports.RailwayGraphQLError = RailwayGraphQLError;
exports.STALE_ENVIRONMENT_BASE_CODE = STALE_ENVIRONMENT_BASE_CODE;
exports.Sandbox = Sandbox;
exports.SandboxFailedError = SandboxFailedError;
exports.SandboxFileNotFoundError = SandboxFileNotFoundError;
exports.SandboxFiles = SandboxFiles;
exports.SandboxFilesError = SandboxFilesError;
exports.SandboxNotFoundError = SandboxNotFoundError;
exports.SandboxTemplateBuildError = SandboxTemplateBuildError;
exports.SandboxTimeoutError = SandboxTimeoutError;
exports.StaleEnvironmentError = StaleEnvironmentError;
exports.bucket = bucket;
exports.changeSetToEnvironmentPatch = changeSetToEnvironmentPatch;
exports.changeSetToGraph = changeSetToGraph;
exports.createRailwayContext = createRailwayContext;
exports.database = database;
exports.define = define;
exports.defineRailway = defineRailway;
exports.diffGraphs = diffGraphs;
exports.empty = empty;
exports.environmentConfigToGraph = environmentConfigToGraph;
exports.evaluateRailwayFile = evaluateRailwayFile;
exports.evaluateRailwayProject = evaluateRailwayProject;
exports.findRailwayFile = findRailwayFile;
exports.fn = fn;
exports.github = github;
exports.graphToEnvironmentConfig = graphToEnvironmentConfig;
exports.group = group;
exports.iac = iac_exports;
exports.image = image;
exports.indexGraph = indexGraph;
exports.mongo = mongo;
exports.mysql = mysql;
exports.postgres = postgres;
exports.project = project;
exports.redis = redis;
exports.ref = ref;
exports.renderChangeSet = renderChangeSet;
exports.renderPatchDiff = renderPatchDiff;
exports.renderRailwayGraphTypes = renderRailwayGraphTypes;
exports.resourceAddress = resourceAddress;
exports.service = service;
exports.template = template;
exports.validateChangeSet = validateChangeSet;
exports.validateGraph = validateGraph;
exports.volume = volume;
